// @babel/core/lib/transformation/index.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var marked0$0 = [run, transformFile].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.run = run;
function _traverse() {
  var data = require("@babel/traverse");
  _traverse = function () {
    return data;
  };
  return data;
}
var _pluginPass = require("./plugin-pass.js");
var _blockHoistPlugin = require("./block-hoist-plugin.js");
var _normalizeOpts = require("./normalize-opts.js");
var _normalizeFile = require("./normalize-file.js");
var _generate = require("./file/generate.js");
var _deepArray = require("../config/helpers/deep-array.js");
var _async = require("../gensync-utils/async.js");
function run(config, code, ast) {
  var file, opts, _opts$filename, outputCode, outputMap, _ref, _opts$filename2;

  return regeneratorRuntime.wrap(function run$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield((0, _normalizeFile["default"])(config.passes, (0, _normalizeOpts["default"])(config), code, ast), "t0", 1);

      case 1:
        file = context$1$0.t0;
        opts = file.opts;
        context$1$0.prev = 3;
        return context$1$0.delegateYield(transformFile(file, config.passes), "t1", 5);

      case 5:
        context$1$0.next = 12;
        break;

      case 7:
        context$1$0.prev = 7;
        context$1$0.t2 = context$1$0["catch"](3);

        context$1$0.t2.message = ((_opts$filename = opts.filename) != null ? _opts$filename : "unknown file") + ": " + context$1$0.t2.message;
        if (!context$1$0.t2.code) {
          context$1$0.t2.code = "BABEL_TRANSFORM_ERROR";
        }
        throw context$1$0.t2;

      case 12:
        outputCode = undefined, outputMap = undefined;
        context$1$0.prev = 13;

        if (opts.code !== false) {
          _ref = (0, _generate["default"])(config.passes, file);
          outputCode = _ref.outputCode;
          outputMap = _ref.outputMap;
        }
        context$1$0.next = 22;
        break;

      case 17:
        context$1$0.prev = 17;
        context$1$0.t3 = context$1$0["catch"](13);

        context$1$0.t3.message = ((_opts$filename2 = opts.filename) != null ? _opts$filename2 : "unknown file") + ": " + context$1$0.t3.message;
        if (!context$1$0.t3.code) {
          context$1$0.t3.code = "BABEL_GENERATE_ERROR";
        }
        throw context$1$0.t3;

      case 22:
        return context$1$0.abrupt("return", {
          metadata: file.metadata,
          options: opts,
          ast: opts.ast === true ? file.ast : null,
          code: outputCode === undefined ? null : outputCode,
          map: outputMap === undefined ? null : outputMap,
          sourceType: file.ast.program.sourceType,
          externalDependencies: (0, _deepArray.flattenToSet)(config.externalDependencies)
        });

      case 23:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this, [[3, 7], [13, 17]]);
}
function transformFile(file, pluginPasses) {
  var async, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, pluginPairs, passPairs, passes, visitors, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, plugin, pass, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, _step3$value, fn, visitor, _iteratorNormalCompletion4, _didIteratorError4, _iteratorError4, _iterator4, _step4, _step4$value;

  return regeneratorRuntime.wrap(function transformFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield((0, _async.isAsync)(), "t0", 1);

      case 1:
        async = context$1$0.t0;
        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 5;
        _iterator = pluginPasses[Symbol.iterator]();

      case 7:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 94;
          break;
        }

        pluginPairs = _step.value;
        passPairs = [];
        passes = [];
        visitors = [];
        _iteratorNormalCompletion2 = true;
        _didIteratorError2 = false;
        _iteratorError2 = undefined;
        context$1$0.prev = 15;

        for (_iterator2 = pluginPairs.concat([(0, _blockHoistPlugin["default"])()])[Symbol.iterator](); !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          plugin = _step2.value;
          pass = new _pluginPass["default"](file, plugin.key, plugin.options, async);

          passPairs.push([plugin, pass]);
          passes.push(pass);
          visitors.push(plugin.visitor);
        }
        context$1$0.next = 23;
        break;

      case 19:
        context$1$0.prev = 19;
        context$1$0.t1 = context$1$0["catch"](15);
        _didIteratorError2 = true;
        _iteratorError2 = context$1$0.t1;

      case 23:
        context$1$0.prev = 23;
        context$1$0.prev = 24;

        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }

      case 26:
        context$1$0.prev = 26;

        if (!_didIteratorError2) {
          context$1$0.next = 29;
          break;
        }

        throw _iteratorError2;

      case 29:
        return context$1$0.finish(26);

      case 30:
        return context$1$0.finish(23);

      case 31:
        _iteratorNormalCompletion3 = true;
        _didIteratorError3 = false;
        _iteratorError3 = undefined;
        context$1$0.prev = 34;
        _iterator3 = passPairs[Symbol.iterator]();

      case 36:
        if (_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done) {
          context$1$0.next = 46;
          break;
        }

        _step3$value = _slicedToArray(_step3.value, 2);
        plugin = _step3$value[0];
        pass = _step3$value[1];

        if (!plugin.pre) {
          context$1$0.next = 43;
          break;
        }

        fn = (0, _async.maybeAsync)(plugin.pre, "You appear to be using an async plugin/preset, but Babel has been called synchronously");
        return context$1$0.delegateYield(fn.call(pass, file), "t2", 43);

      case 43:
        _iteratorNormalCompletion3 = true;
        context$1$0.next = 36;
        break;

      case 46:
        context$1$0.next = 52;
        break;

      case 48:
        context$1$0.prev = 48;
        context$1$0.t3 = context$1$0["catch"](34);
        _didIteratorError3 = true;
        _iteratorError3 = context$1$0.t3;

      case 52:
        context$1$0.prev = 52;
        context$1$0.prev = 53;

        if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
          _iterator3["return"]();
        }

      case 55:
        context$1$0.prev = 55;

        if (!_didIteratorError3) {
          context$1$0.next = 58;
          break;
        }

        throw _iteratorError3;

      case 58:
        return context$1$0.finish(55);

      case 59:
        return context$1$0.finish(52);

      case 60:
        visitor = _traverse()["default"].visitors.merge(visitors, passes, file.opts.wrapPluginVisitorMethod);

        (0, _traverse()["default"])(file.ast, visitor, file.scope);
        _iteratorNormalCompletion4 = true;
        _didIteratorError4 = false;
        _iteratorError4 = undefined;
        context$1$0.prev = 65;
        _iterator4 = passPairs[Symbol.iterator]();

      case 67:
        if (_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done) {
          context$1$0.next = 77;
          break;
        }

        _step4$value = _slicedToArray(_step4.value, 2);
        plugin = _step4$value[0];
        pass = _step4$value[1];

        if (!plugin.post) {
          context$1$0.next = 74;
          break;
        }

        fn = (0, _async.maybeAsync)(plugin.post, "You appear to be using an async plugin/preset, but Babel has been called synchronously");
        return context$1$0.delegateYield(fn.call(pass, file), "t4", 74);

      case 74:
        _iteratorNormalCompletion4 = true;
        context$1$0.next = 67;
        break;

      case 77:
        context$1$0.next = 83;
        break;

      case 79:
        context$1$0.prev = 79;
        context$1$0.t5 = context$1$0["catch"](65);
        _didIteratorError4 = true;
        _iteratorError4 = context$1$0.t5;

      case 83:
        context$1$0.prev = 83;
        context$1$0.prev = 84;

        if (!_iteratorNormalCompletion4 && _iterator4["return"]) {
          _iterator4["return"]();
        }

      case 86:
        context$1$0.prev = 86;

        if (!_didIteratorError4) {
          context$1$0.next = 89;
          break;
        }

        throw _iteratorError4;

      case 89:
        return context$1$0.finish(86);

      case 90:
        return context$1$0.finish(83);

      case 91:
        _iteratorNormalCompletion = true;
        context$1$0.next = 7;
        break;

      case 94:
        context$1$0.next = 100;
        break;

      case 96:
        context$1$0.prev = 96;
        context$1$0.t6 = context$1$0["catch"](5);
        _didIteratorError = true;
        _iteratorError = context$1$0.t6;

      case 100:
        context$1$0.prev = 100;
        context$1$0.prev = 101;

        if (!_iteratorNormalCompletion && _iterator["return"]) {
          _iterator["return"]();
        }

      case 103:
        context$1$0.prev = 103;

        if (!_didIteratorError) {
          context$1$0.next = 106;
          break;
        }

        throw _iteratorError;

      case 106:
        return context$1$0.finish(103);

      case 107:
        return context$1$0.finish(100);

      case 108:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this, [[5, 96, 100, 108], [15, 19, 23, 31], [24,, 26, 30], [34, 48, 52, 60], [53,, 55, 59], [65, 79, 83, 91], [84,, 86, 90], [101,, 103, 107]]);
}
0 && 0;
