// @babel/core/lib/transformation/normalize-opts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = normalizeOptions;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function normalizeOptions(config) {
  var _config$options = config.options;
  var filename = _config$options.filename;
  var cwd = _config$options.cwd;
  var _config$options$filenameRelative = _config$options.filenameRelative;
  var filenameRelative = _config$options$filenameRelative === undefined ? typeof filename === "string" ? _path().relative(cwd, filename) : "unknown" : _config$options$filenameRelative;
  var _config$options$sourceType = _config$options.sourceType;
  var sourceType = _config$options$sourceType === undefined ? "module" : _config$options$sourceType;
  var inputSourceMap = _config$options.inputSourceMap;
  var _config$options$sourceMaps = _config$options.sourceMaps;
  var sourceMaps = _config$options$sourceMaps === undefined ? !!inputSourceMap : _config$options$sourceMaps;
  var _config$options$sourceRoot = _config$options.sourceRoot;
  var sourceRoot = _config$options$sourceRoot === undefined ? config.options.moduleRoot : _config$options$sourceRoot;
  var _config$options$sourceFileName = _config$options.sourceFileName;
  var sourceFileName = _config$options$sourceFileName === undefined ? _path().basename(filenameRelative) : _config$options$sourceFileName;
  var _config$options$comments = _config$options.comments;
  var comments = _config$options$comments === undefined ? true : _config$options$comments;
  var _config$options$compact = _config$options.compact;
  var compact = _config$options$compact === undefined ? "auto" : _config$options$compact;

  var opts = config.options;
  var options = Object.assign({}, opts, {
    parserOpts: Object.assign({
      sourceType: _path().extname(filenameRelative) === ".mjs" ? "module" : sourceType,
      sourceFileName: filename,
      plugins: []
    }, opts.parserOpts),
    generatorOpts: Object.assign({
      filename: filename,
      auxiliaryCommentBefore: opts.auxiliaryCommentBefore,
      auxiliaryCommentAfter: opts.auxiliaryCommentAfter,
      retainLines: opts.retainLines,
      comments: comments,
      shouldPrintComment: opts.shouldPrintComment,
      compact: compact,
      minified: opts.minified,
      sourceMaps: sourceMaps,
      sourceRoot: sourceRoot,
      sourceFileName: sourceFileName
    }, opts.generatorOpts)
  });
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = config.passes[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var plugins = _step.value;
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = plugins[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var plugin = _step2.value;

          if (plugin.manipulateOptions) {
            plugin.manipulateOptions(options, options.parserOpts);
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return options;
}
0 && 0;
