// @babel/core/lib/transformation/normalize-file.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var marked0$0 = [normalizeFile].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = normalizeFile;
function _fs() {
  var data = require("fs");
  _fs = function () {
    return data;
  };
  return data;
}
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _debug() {
  var data = require("debug");
  _debug = function () {
    return data;
  };
  return data;
}
function _t() {
  var data = require("@babel/types");
  _t = function () {
    return data;
  };
  return data;
}
function _convertSourceMap() {
  var data = require("convert-source-map");
  _convertSourceMap = function () {
    return data;
  };
  return data;
}
var _file = require("./file/file.js");
var _index = require("../parser/index.js");
var _cloneDeep = require("./util/clone-deep.js");

var _t2 = _t();

var file = _t2.file;
var traverseFast = _t2.traverseFast;

var debug = _debug()("babel:transform:file");
var INLINE_SOURCEMAP_REGEX = /^[@#]\s+sourceMappingURL=data:(?:application|text)\/json;(?:charset[:=]\S+?;)?base64,.*$/;
var EXTERNAL_SOURCEMAP_REGEX = /^[@#][ \t]+sourceMappingURL=([^\s'"`]+)[ \t]*$/;
function normalizeFile(pluginPasses, options, code, ast) {
  var inputMap, lastComment, match, inputMapContent;
  return regeneratorRuntime.wrap(function normalizeFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        code = "" + (code || "");

        if (!ast) {
          context$1$0.next = 11;
          break;
        }

        if (!(ast.type === "Program")) {
          context$1$0.next = 6;
          break;
        }

        ast = file(ast, [], []);
        context$1$0.next = 8;
        break;

      case 6:
        if (!(ast.type !== "File")) {
          context$1$0.next = 8;
          break;
        }

        throw new Error("AST root must be a Program or File node");

      case 8:
        if (options.cloneInputAst) {
          ast = (0, _cloneDeep["default"])(ast);
        }
        context$1$0.next = 13;
        break;

      case 11:
        return context$1$0.delegateYield((0, _index["default"])(pluginPasses, options, code), "t0", 12);

      case 12:
        ast = context$1$0.t0;

      case 13:
        inputMap = null;

        if (options.inputSourceMap !== false) {
          if (typeof options.inputSourceMap === "object") {
            inputMap = _convertSourceMap().fromObject(options.inputSourceMap);
          }
          if (!inputMap) {
            lastComment = extractComments(INLINE_SOURCEMAP_REGEX, ast);

            if (lastComment) {
              try {
                inputMap = _convertSourceMap().fromComment("//" + lastComment);
              } catch (err) {
                {
                  debug("discarding unknown inline input sourcemap");
                }
              }
            }
          }
          if (!inputMap) {
            lastComment = extractComments(EXTERNAL_SOURCEMAP_REGEX, ast);

            if (typeof options.filename === "string" && lastComment) {
              try {
                match = EXTERNAL_SOURCEMAP_REGEX.exec(lastComment);
                inputMapContent = _fs().readFileSync(_path().resolve(_path().dirname(options.filename), match[1]), "utf8");

                inputMap = _convertSourceMap().fromJSON(inputMapContent);
              } catch (err) {
                debug("discarding unknown file input sourcemap", err);
              }
            } else if (lastComment) {
              debug("discarding un-loadable file input sourcemap");
            }
          }
        }
        return context$1$0.abrupt("return", new _file["default"](options, {
          code: code,
          ast: ast,
          inputMap: inputMap
        }));

      case 16:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function extractCommentsFromList(regex, comments, lastComment) {
  if (comments) {
    comments = comments.filter(function (_ref) {
      var value = _ref.value;

      if (regex.test(value)) {
        lastComment = value;
        return false;
      }
      return true;
    });
  }
  return [comments, lastComment];
}
function extractComments(regex, ast) {
  var lastComment = null;
  traverseFast(ast, function (node) {
    var _extractCommentsFromList = extractCommentsFromList(regex, node.leadingComments, lastComment);

    var _extractCommentsFromList2 = _slicedToArray(_extractCommentsFromList, 2);

    node.leadingComments = _extractCommentsFromList2[0];
    lastComment = _extractCommentsFromList2[1];

    var _extractCommentsFromList3 = extractCommentsFromList(regex, node.innerComments, lastComment);

    var _extractCommentsFromList32 = _slicedToArray(_extractCommentsFromList3, 2);

    node.innerComments = _extractCommentsFromList32[0];
    lastComment = _extractCommentsFromList32[1];

    var _extractCommentsFromList4 = extractCommentsFromList(regex, node.trailingComments, lastComment);

    var _extractCommentsFromList42 = _slicedToArray(_extractCommentsFromList4, 2);

    node.trailingComments = _extractCommentsFromList42[0];
    lastComment = _extractCommentsFromList42[1];
  });
  return lastComment;
}
0 && 0;
