// @babel/helper-define-polyfill-provider/lib/visitors/usage.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;
var _utils = require("../utils");
function isRemoved(_x) {
  var _again = true;

  _function: while (_again) {
    var path = _x;
    _again = false;

    if (path.removed) return true;
    if (!path.parentPath) return false;
    if (path.listKey) {
      var _path$parentPath$node;
      if (!((_path$parentPath$node = path.parentPath.node) != null && (_path$parentPath$node = _path$parentPath$node[path.listKey]) != null && _path$parentPath$node.includes(path.node))) return true;
    } else {
      if (path.parentPath.node[path.key] !== path.node) return true;
    }
    _x = path.parentPath;
    _again = true;
    _path$parentPath$node = undefined;
    continue _function;
  }
}
var _default = function _default(callProvider) {
  function property(object, key, placement, path) {
    return callProvider({
      kind: "property",
      object: object,
      key: key,
      placement: placement
    }, path);
  }
  function handleReferencedIdentifier(path) {
    var name = path.node.name;
    var scope = path.scope;

    if (scope.getBindingIdentifier(name)) return;
    callProvider({
      kind: "global",
      name: name
    }, path);
  }
  function analyzeMemberExpression(path) {
    var key = (0, _utils.resolveKey)(path.get("property"), path.node.computed);
    return {
      key: key,
      handleAsMemberExpression: !!key && key !== "prototype"
    };
  }
  return {
    // Symbol(), new Promise
    ReferencedIdentifier: function ReferencedIdentifier(path) {
      var parentPath = path.parentPath;

      if (parentPath.isMemberExpression({
        object: path.node
      }) && analyzeMemberExpression(parentPath).handleAsMemberExpression) {
        return;
      }
      handleReferencedIdentifier(path);
    },
    "MemberExpression|OptionalMemberExpression": function MemberExpressionOptionalMemberExpression(path) {
      var _analyzeMemberExpression = analyzeMemberExpression(path);

      var key = _analyzeMemberExpression.key;
      var handleAsMemberExpression = _analyzeMemberExpression.handleAsMemberExpression;

      if (!handleAsMemberExpression) return;
      var object = path.get("object");
      var objectIsGlobalIdentifier = object.isIdentifier();
      if (objectIsGlobalIdentifier) {
        var binding = object.scope.getBinding(object.node.name);
        if (binding) {
          if (binding.path.isImportNamespaceSpecifier()) return;
          objectIsGlobalIdentifier = false;
        }
      }
      var source = (0, _utils.resolveSource)(object);
      var skipObject = property(source.id, key, source.placement, path);
      skipObject || (skipObject = !objectIsGlobalIdentifier || path.shouldSkip || object.shouldSkip || isRemoved(object));
      if (!skipObject) handleReferencedIdentifier(object);
    },
    ObjectPattern: function ObjectPattern(path) {
      var parentPath = path.parentPath;
      var parent = path.parent;

      var obj = undefined;

      // const { keys, values } = Object
      if (parentPath.isVariableDeclarator()) {
        obj = parentPath.get("init");
        // ({ keys, values } = Object)
      } else if (parentPath.isAssignmentExpression()) {
          obj = parentPath.get("right");
          // !function ({ keys, values }) {...} (Object)
          // resolution does not work after properties transform :-(
        } else if (parentPath.isFunction()) {
            var grand = parentPath.parentPath;
            if (grand.isCallExpression() || grand.isNewExpression()) {
              if (grand.node.callee === parent) {
                obj = grand.get("arguments")[path.key];
              }
            }
          }
      var id = null;
      var placement = null;
      if (obj) {
        ;

        var _ref = (0, _utils.resolveSource)(obj);

        id = _ref.id;
        placement = _ref.placement;
      }var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = path.get("properties")[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var prop = _step.value;

          if (prop.isObjectProperty()) {
            var key = (0, _utils.resolveKey)(prop.get("key"));
            if (key) property(id, key, placement, prop);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"]) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    },
    BinaryExpression: function BinaryExpression(path) {
      if (path.node.operator !== "in") return;
      var source = (0, _utils.resolveSource)(path.get("right"));
      var key = (0, _utils.resolveKey)(path.get("left"), true);
      if (!key) return;
      callProvider({
        kind: "in",
        object: source.id,
        key: key,
        placement: source.placement
      }, path);
    }
  };
};
exports["default"] = _default;
