// @babel/plugin-bugfix-firefox-class-in-computed-class-key/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _traverse = require("@babel/traverse");
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _default = exports["default"] = (0, _helperPluginUtils.declare)(function (_ref) {
  var t = _ref.types;
  var assertVersion = _ref.assertVersion;

  assertVersion(7);
  var containsClassExpressionVisitor = {
    ClassExpression: function ClassExpression(path, state) {
      state.found = true;
      path.stop();
    },
    Function: function Function(path) {
      path.skip();
    }
  };
  var containsYieldOrAwaitVisitor = _traverse.visitors.environmentVisitor({
    YieldExpression: function YieldExpression(path, state) {
      state["yield"] = true;
      if (state.await) path.stop();
    },
    AwaitExpression: function AwaitExpression(path, state) {
      state.await = true;
      if (state["yield"]) path.stop();
    }
  });
  function containsClassExpression(path) {
    if (t.isClassExpression(path.node)) return true;
    if (t.isFunction(path.node)) return false;
    var state = {
      found: false
    };
    path.traverse(containsClassExpressionVisitor, state);
    return state.found;
  }
  function wrap(path) {
    var context = {
      "yield": t.isYieldExpression(path.node),
      await: t.isAwaitExpression(path.node)
    };
    path.traverse(containsYieldOrAwaitVisitor, context);
    var replacement = undefined;
    if (context["yield"]) {
      var fn = t.functionExpression(null, [], t.blockStatement([t.returnStatement(path.node)]), true, context.await);
      replacement = t.yieldExpression(t.callExpression(t.memberExpression(fn, t.identifier("call")), [t.thisExpression(), t.identifier("arguments")]), true);
    } else {
      var fn = t.arrowFunctionExpression([], path.node, context.await);
      replacement = t.callExpression(fn, []);
      if (context.await) replacement = t.awaitExpression(replacement);
    }
    path.replaceWith(replacement);
  }
  return {
    name: "bugfix-firefox-class-in-computed-class-key",
    visitor: {
      Class: function Class(path) {
        var hasPrivateElement = path.node.body.body.some(function (node) {
          return t.isPrivate(node);
        });
        if (!hasPrivateElement) return;
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = path.get("body.body")[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var elem = _step.value;

            if ("computed" in elem.node && elem.node.computed && containsClassExpression(elem.get("key"))) {
              wrap(elem.get("key"));
            }
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"]) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
    }
  };
});
