// @babel/generator/lib/node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TokenContext = void 0;
exports.isLastChild = isLastChild;
exports.needsParens = needsParens;
exports.needsWhitespace = needsWhitespace;
exports.needsWhitespaceAfter = needsWhitespaceAfter;
exports.needsWhitespaceBefore = needsWhitespaceBefore;
var whitespace = require("./whitespace.js");
var parens = require("./parentheses.js");
var _t = require("@babel/types");
var FLIPPED_ALIAS_KEYS = _t.FLIPPED_ALIAS_KEYS;
var VISITOR_KEYS = _t.VISITOR_KEYS;
var isCallExpression = _t.isCallExpression;
var isDecorator = _t.isDecorator;
var isExpressionStatement = _t.isExpressionStatement;
var isMemberExpression = _t.isMemberExpression;
var isNewExpression = _t.isNewExpression;
var isParenthesizedExpression = _t.isParenthesizedExpression;

var TokenContext = exports.TokenContext = {
  expressionStatement: 1,
  arrowBody: 2,
  exportDefault: 4,
  forHead: 8,
  forInHead: 16,
  forOfHead: 32,
  arrowFlowReturnType: 64
};
function expandAliases(obj) {
  var map = new Map();
  function add(type, func) {
    var fn = map.get(type);
    map.set(type, fn ? function (node, parent, stack, inForInit, getRawIdentifier) {
      var _fn;
      return (_fn = fn(node, parent, stack, inForInit, getRawIdentifier)) != null ? _fn : func(node, parent, stack, inForInit, getRawIdentifier);
    } : func);
  }
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = Object.keys(obj)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var type = _step.value;

      var aliases = FLIPPED_ALIAS_KEYS[type];
      if (aliases) {
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = aliases[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var alias = _step2.value;

            add(alias, obj[type]);
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
              _iterator2["return"]();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }
      } else {
        add(type, obj[type]);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return map;
}
var expandedParens = expandAliases(parens);
var expandedWhitespaceNodes = expandAliases(whitespace.nodes);
function isOrHasCallExpression(_x) {
  var _left;

  var _again = true;

  _function: while (_again) {
    var node = _x;
    _again = false;

    if (isCallExpression(node)) {
      return true;
    }

    if (!(_left = isMemberExpression(node))) {
      return _left;
    }

    _x = node.object;
    _again = true;
    continue _function;
  }
}
function needsWhitespace(node, parent, type) {
  var _expandedWhitespaceNo;
  if (!node) return false;
  if (isExpressionStatement(node)) {
    node = node.expression;
  }
  var flag = (_expandedWhitespaceNo = expandedWhitespaceNodes.get(node.type)) == null ? void 0 : _expandedWhitespaceNo(node, parent);
  if (typeof flag === "number") {
    return (flag & type) !== 0;
  }
  return false;
}
function needsWhitespaceBefore(node, parent) {
  return needsWhitespace(node, parent, 1);
}
function needsWhitespaceAfter(node, parent) {
  return needsWhitespace(node, parent, 2);
}
function needsParens(node, parent, tokenContext, inForInit, getRawIdentifier) {
  var _expandedParens$get;
  if (!parent) return false;
  if (isNewExpression(parent) && parent.callee === node) {
    if (isOrHasCallExpression(node)) return true;
  }
  if (isDecorator(parent)) {
    return !isDecoratorMemberExpression(node) && !(isCallExpression(node) && isDecoratorMemberExpression(node.callee)) && !isParenthesizedExpression(node);
  }
  return (_expandedParens$get = expandedParens.get(node.type)) == null ? void 0 : _expandedParens$get(node, parent, tokenContext, inForInit, getRawIdentifier);
}
function isDecoratorMemberExpression(_x2) {
  var _left2;

  var _again2 = true;

  _function2: while (_again2) {
    var node = _x2;
    _again2 = false;

    switch (node.type) {
      case "Identifier":
        return true;
      case "MemberExpression":
        if (!(_left2 = !node.computed && node.property.type === "Identifier")) {
          return _left2;
        }

        _x2 = node.object;
        _again2 = true;
        continue _function2;

      default:
        return false;
    }
  }
}
function isLastChild(parent, child) {
  var visitorKeys = VISITOR_KEYS[parent.type];
  for (var i = visitorKeys.length - 1; i >= 0; i--) {
    var val = parent[visitorKeys[i]];
    if (val === child) {
      return true;
    } else if (Array.isArray(val)) {
      var j = val.length - 1;
      while (j >= 0 && val[j] === null) j--;
      return j >= 0 && val[j] === child;
    } else if (val) {
      return false;
    }
  }
  return false;
}
