// @babel/helper-validator-option/lib/validator.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OptionValidator = void 0;
var _findSuggestion = require("./find-suggestion.js");

var OptionValidator = (function () {
  function OptionValidator(descriptor) {
    _classCallCheck(this, OptionValidator);

    this.descriptor = descriptor;
  }

  _createClass(OptionValidator, [{
    key: "validateTopLevelOptions",
    value: function validateTopLevelOptions(options, TopLevelOptionShape) {
      var validOptionNames = Object.keys(TopLevelOptionShape);
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = Object.keys(options)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var option = _step.value;

          if (!validOptionNames.includes(option)) {
            throw new Error(this.formatMessage("'" + option + "' is not a valid top-level option.\n- Did you mean '" + (0, _findSuggestion.findSuggestion)(option, validOptionNames) + "'?"));
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"]) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "validateBooleanOption",
    value: function validateBooleanOption(name, value, defaultValue) {
      if (value === undefined) {
        return defaultValue;
      } else {
        this.invariant(typeof value === "boolean", "'" + name + "' option must be a boolean.");
      }
      return value;
    }
  }, {
    key: "validateStringOption",
    value: function validateStringOption(name, value, defaultValue) {
      if (value === undefined) {
        return defaultValue;
      } else {
        this.invariant(typeof value === "string", "'" + name + "' option must be a string.");
      }
      return value;
    }
  }, {
    key: "invariant",
    value: function invariant(condition, message) {
      if (!condition) {
        throw new Error(this.formatMessage(message));
      }
    }
  }, {
    key: "formatMessage",
    value: function formatMessage(message) {
      return this.descriptor + ": " + message;
    }
  }]);

  return OptionValidator;
})();

exports.OptionValidator = OptionValidator;
