// @babel/helper-define-polyfill-provider/lib/imports-injector.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

exports.__esModule = true;
exports["default"] = void 0;
var _babel = _interopRequireWildcard(require("@babel/core"));
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return { "default": obj };
  }var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {
    return cache.get(obj);
  }var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }newObj["default"] = obj;if (cache) {
    cache.set(obj, newObj);
  }return newObj;
}

var _ref = _babel["default"] || _babel;

var t = _ref.types;

var ImportsCachedInjector = (function () {
  function ImportsCachedInjector(resolver, getPreferredIndex) {
    _classCallCheck(this, ImportsCachedInjector);

    this._imports = new WeakMap();
    this._anonymousImports = new WeakMap();
    this._lastImports = new WeakMap();
    this._resolver = resolver;
    this._getPreferredIndex = getPreferredIndex;
  }

  _createClass(ImportsCachedInjector, [{
    key: "storeAnonymous",
    value: function storeAnonymous(programPath, url, moduleName, getVal) {
      var key = this._normalizeKey(programPath, url);
      var imports = this._ensure(this._anonymousImports, programPath, Set);
      if (imports.has(key)) return;
      var node = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)));
      imports.add(key);
      this._injectImport(programPath, node, moduleName);
    }
  }, {
    key: "storeNamed",
    value: function storeNamed(programPath, url, name, moduleName, getVal) {
      var key = this._normalizeKey(programPath, url, name);
      var imports = this._ensure(this._imports, programPath, Map);
      if (!imports.has(key)) {
        var _getVal = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)), t.identifier(name));

        var node = _getVal.node;
        var id = _getVal.name;

        imports.set(key, id);
        this._injectImport(programPath, node, moduleName);
      }
      return t.identifier(imports.get(key));
    }
  }, {
    key: "_injectImport",
    value: function _injectImport(programPath, node, moduleName) {
      var _this$_lastImports$ge;
      var newIndex = this._getPreferredIndex(moduleName);
      var lastImports = (_this$_lastImports$ge = this._lastImports.get(programPath)) != null ? _this$_lastImports$ge : [];
      var isPathStillValid = function isPathStillValid(path) {
        return path.node &&
        // Sometimes the AST is modified and the "last import"
        // we have has been replaced
        path.parent === programPath.node && path.container === programPath.node.body;
      };
      var last = undefined;
      if (newIndex === Infinity) {
        // Fast path: we can always just insert at the end if newIndex is `Infinity`
        if (lastImports.length > 0) {
          last = lastImports[lastImports.length - 1].path;
          if (!isPathStillValid(last)) last = undefined;
        }
      } else {
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = lastImports.entries()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var _step$value = _slicedToArray(_step.value, 2);

            var i = _step$value[0];
            var data = _step$value[1];
            var path = data.path;
            var index = data.index;

            if (isPathStillValid(path)) {
              if (newIndex < index) {
                var _path$insertBefore = path.insertBefore(node);

                var _path$insertBefore2 = _slicedToArray(_path$insertBefore, 1);

                var newPath = _path$insertBefore2[0];

                lastImports.splice(i, 0, {
                  path: newPath,
                  index: newIndex
                });
                return;
              }
              last = path;
            }
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"]) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
      if (last) {
        var _last$insertAfter = last.insertAfter(node);

        var _last$insertAfter2 = _slicedToArray(_last$insertAfter, 1);

        var newPath = _last$insertAfter2[0];

        lastImports.push({
          path: newPath,
          index: newIndex
        });
      } else {
        var _programPath$unshiftContainer = programPath.unshiftContainer("body", node);

        var _programPath$unshiftContainer2 = _slicedToArray(_programPath$unshiftContainer, 1);

        var newPath = _programPath$unshiftContainer2[0];

        this._lastImports.set(programPath, [{
          path: newPath,
          index: newIndex
        }]);
      }
    }
  }, {
    key: "_ensure",
    value: function _ensure(map, programPath, Collection) {
      var collection = map.get(programPath);
      if (!collection) {
        collection = new Collection();
        map.set(programPath, collection);
      }
      return collection;
    }
  }, {
    key: "_normalizeKey",
    value: function _normalizeKey(programPath, url) {
      var name = arguments.length <= 2 || arguments[2] === undefined ? "" : arguments[2];
      var sourceType = programPath.node.sourceType;

      // If we rely on the imported binding (the "name" parameter), we also need to cache
      // based on the sourceType. This is because the module transforms change the names
      // of the import variables.
      return (name && sourceType) + "::" + url + "::" + name;
    }
  }]);

  return ImportsCachedInjector;
})();

exports["default"] = ImportsCachedInjector;
