// @babel/helper-define-polyfill-provider/lib/index.js
"use strict";

exports.__esModule = true;
exports["default"] = definePolyfillProvider;
var _helperPluginUtils = require("@babel/helper-plugin-utils");
var _helperCompilationTargets = _interopRequireWildcard(require("@babel/helper-compilation-targets"));
var _utils = require("./utils");
var _importsInjector = _interopRequireDefault(require("./imports-injector"));
var _debugUtils = require("./debug-utils");
var _normalizeOptions = require("./normalize-options");
var v = _interopRequireWildcard(require("./visitors"));
var deps = _interopRequireWildcard(require("./node/dependencies"));
var _metaResolver = _interopRequireDefault(require("./meta-resolver"));
var _excluded = ["method", "targets", "ignoreBrowserslistConfig", "configPath", "debug", "shouldInjectPolyfill", "absoluteImports"];
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { "default": obj };
}
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return { "default": obj };
  }var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {
    return cache.get(obj);
  }var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }newObj["default"] = obj;if (cache) {
    cache.set(obj, newObj);
  }return newObj;
}
function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];
  }return target;
}
var getTargets = _helperCompilationTargets["default"]["default"] || _helperCompilationTargets["default"];
function resolveOptions(options, babelApi) {
  var method = options.method;
  var targetsOption = options.targets;
  var ignoreBrowserslistConfig = options.ignoreBrowserslistConfig;
  var configPath = options.configPath;
  var debug = options.debug;
  var shouldInjectPolyfill = options.shouldInjectPolyfill;
  var absoluteImports = options.absoluteImports;
  var providerOptions = _objectWithoutPropertiesLoose(options, _excluded);
  if (isEmpty(options)) {
    throw new Error("This plugin requires options, for example:\n    {\n      \"plugins\": [\n        [\"<plugin name>\", { method: \"usage-pure\" }]\n      ]\n    }\n\nSee more options at https://github.com/babel/babel-polyfills/blob/main/docs/usage.md");
  }
  var methodName = undefined;
  if (method === "usage-global") methodName = "usageGlobal";else if (method === "entry-global") methodName = "entryGlobal";else if (method === "usage-pure") methodName = "usagePure";else if (typeof method !== "string") {
    throw new Error(".method must be a string");
  } else {
    throw new Error(".method must be one of \"entry-global\", \"usage-global\"" + (" or \"usage-pure\" (received " + JSON.stringify(method) + ")"));
  }
  if (typeof shouldInjectPolyfill === "function") {
    if (options.include || options.exclude) {
      throw new Error(".include and .exclude are not supported when using the" + " .shouldInjectPolyfill function.");
    }
  } else if (shouldInjectPolyfill != null) {
    throw new Error(".shouldInjectPolyfill must be a function, or undefined" + (" (received " + JSON.stringify(shouldInjectPolyfill) + ")"));
  }
  if (absoluteImports != null && typeof absoluteImports !== "boolean" && typeof absoluteImports !== "string") {
    throw new Error(".absoluteImports must be a boolean, a string, or undefined" + (" (received " + JSON.stringify(absoluteImports) + ")"));
  }
  var targets = undefined;
  if (
  // If any browserslist-related option is specified, fallback to the old
  // behavior of not using the targets specified in the top-level options.
  targetsOption || configPath || ignoreBrowserslistConfig) {
    var targetsObj = typeof targetsOption === "string" || Array.isArray(targetsOption) ? {
      browsers: targetsOption
    } : targetsOption;
    targets = getTargets(targetsObj, {
      ignoreBrowserslistConfig: ignoreBrowserslistConfig,
      configPath: configPath
    });
  } else {
    targets = babelApi.targets();
  }
  return {
    method: method,
    methodName: methodName,
    targets: targets,
    absoluteImports: absoluteImports != null ? absoluteImports : false,
    shouldInjectPolyfill: shouldInjectPolyfill,
    debug: !!debug,
    providerOptions: providerOptions
  };
}
function instantiateProvider(factory, options, missingDependencies, dirname, debugLog, babelApi) {
  var _resolveOptions = resolveOptions(options, babelApi);

  var method = _resolveOptions.method;
  var methodName = _resolveOptions.methodName;
  var targets = _resolveOptions.targets;
  var _debug = _resolveOptions.debug;
  var _shouldInjectPolyfill = _resolveOptions.shouldInjectPolyfill;
  var providerOptions = _resolveOptions.providerOptions;
  var absoluteImports = _resolveOptions.absoluteImports;

  // eslint-disable-next-line prefer-const
  var include = undefined,
      exclude = undefined;
  var polyfillsSupport = undefined;
  var polyfillsNames = undefined;
  var filterPolyfills = undefined;
  var getUtils = (0, _utils.createUtilsGetter)(new _importsInjector["default"](function (moduleName) {
    return deps.resolve(dirname, moduleName, absoluteImports);
  }, function (name) {
    var _polyfillsNames$get, _polyfillsNames;
    return (_polyfillsNames$get = (_polyfillsNames = polyfillsNames) == null ? void 0 : _polyfillsNames.get(name)) != null ? _polyfillsNames$get : Infinity;
  }));
  var depsCache = new Map();
  var api = {
    babel: babelApi,
    getUtils: getUtils,
    method: options.method,
    targets: targets,
    createMetaResolver: _metaResolver["default"],
    shouldInjectPolyfill: function shouldInjectPolyfill(name) {
      if (polyfillsNames === undefined) {
        throw new Error("Internal error in the " + factory.name + " provider: " + "shouldInjectPolyfill() can't be called during initialization.");
      }
      if (!polyfillsNames.has(name)) {
        console.warn("Internal error in the " + providerName + " provider: " + ("unknown polyfill \"" + name + "\"."));
      }
      if (filterPolyfills && !filterPolyfills(name)) return false;
      var shouldInject = (0, _helperCompilationTargets.isRequired)(name, targets, {
        compatData: polyfillsSupport,
        includes: include,
        excludes: exclude
      });
      if (_shouldInjectPolyfill) {
        shouldInject = _shouldInjectPolyfill(name, shouldInject);
        if (typeof shouldInject !== "boolean") {
          throw new Error(".shouldInjectPolyfill must return a boolean.");
        }
      }
      return shouldInject;
    },
    debug: function debug(name) {
      var _debugLog, _debugLog$polyfillsSu;
      debugLog().found = true;
      if (!_debug || !name) return;
      if (debugLog().polyfills.has(providerName)) return;
      debugLog().polyfills.add(name);
      (_debugLog$polyfillsSu = (_debugLog = debugLog()).polyfillsSupport) != null ? _debugLog$polyfillsSu : _debugLog.polyfillsSupport = polyfillsSupport;
    },
    assertDependency: function assertDependency(name) {
      var version = arguments.length <= 1 || arguments[1] === undefined ? "*" : arguments[1];

      if (missingDependencies === false) return;
      if (absoluteImports) {
        // If absoluteImports is not false, we will try resolving
        // the dependency and throw if it's not possible. We can
        // skip the check here.
        return;
      }
      var dep = version === "*" ? name : name + "@^" + version;
      var found = missingDependencies.all ? false : mapGetOr(depsCache, name + " :: " + dirname, function () {
        return deps.has(dirname, name);
      });
      if (!found) {
        debugLog().missingDeps.add(dep);
      }
    }
  };
  var provider = factory(api, providerOptions, dirname);
  var providerName = provider.name || factory.name;
  if (typeof provider[methodName] !== "function") {
    throw new Error("The \"" + providerName + "\" provider doesn't support the \"" + method + "\" polyfilling method.");
  }
  if (Array.isArray(provider.polyfills)) {
    polyfillsNames = new Map(provider.polyfills.map(function (name, index) {
      return [name, index];
    }));
    filterPolyfills = provider.filterPolyfills;
  } else if (provider.polyfills) {
    polyfillsNames = new Map(Object.keys(provider.polyfills).map(function (name, index) {
      return [name, index];
    }));
    polyfillsSupport = provider.polyfills;
    filterPolyfills = provider.filterPolyfills;
  } else {
    polyfillsNames = new Map();
  }

  var _ref2 = (0, _normalizeOptions.validateIncludeExclude)(providerName, polyfillsNames, providerOptions.include || [], providerOptions.exclude || []);

  include = _ref2.include;
  exclude = _ref2.exclude;

  var callProvider = undefined;
  if (methodName === "usageGlobal") {
    callProvider = function (payload, path) {
      var _ref;
      var utils = getUtils(path);
      return (_ref = provider[methodName](payload, utils, path)) != null ? _ref : false;
    };
  } else {
    callProvider = function (payload, path) {
      var utils = getUtils(path);
      provider[methodName](payload, utils, path);
      return false;
    };
  }
  return {
    debug: _debug,
    method: method,
    targets: targets,
    provider: provider,
    providerName: providerName,
    callProvider: callProvider
  };
}
function definePolyfillProvider(factory) {
  return (0, _helperPluginUtils.declare)(function (babelApi, options, dirname) {
    babelApi.assertVersion("^7.0.0 || ^8.0.0-alpha.0");
    var traverse = babelApi.traverse;

    var debugLog = undefined;
    var missingDependencies = (0, _normalizeOptions.applyMissingDependenciesDefaults)(options, babelApi);

    var _instantiateProvider = instantiateProvider(factory, options, missingDependencies, dirname, function () {
      return debugLog;
    }, babelApi);

    var debug = _instantiateProvider.debug;
    var method = _instantiateProvider.method;
    var targets = _instantiateProvider.targets;
    var provider = _instantiateProvider.provider;
    var providerName = _instantiateProvider.providerName;
    var callProvider = _instantiateProvider.callProvider;

    var createVisitor = method === "entry-global" ? v.entry : v.usage;
    var visitor = provider.visitor ? traverse.visitors.merge([createVisitor(callProvider), provider.visitor]) : createVisitor(callProvider);
    if (debug && debug !== _debugUtils.presetEnvSilentDebugHeader) {
      console.log(providerName + ": `DEBUG` option");
      console.log("\nUsing targets: " + (0, _debugUtils.stringifyTargetsMultiline)(targets));
      console.log("\nUsing polyfills with `" + method + "` method:");
    }
    var runtimeName = provider.runtimeName;

    return {
      name: "inject-polyfills",
      visitor: visitor,
      pre: function pre(file) {
        var _provider$pre;
        if (runtimeName) {
          if (file.get("runtimeHelpersModuleName") && file.get("runtimeHelpersModuleName") !== runtimeName) {
            console.warn("Two different polyfill providers" + (" (" + file.get("runtimeHelpersModuleProvider")) + (" and " + providerName + ") are trying to define two") + " conflicting @babel/runtime alternatives:" + (" " + file.get("runtimeHelpersModuleName") + " and " + runtimeName + ".") + " The second one will be ignored.");
          } else {
            file.set("runtimeHelpersModuleName", runtimeName);
            file.set("runtimeHelpersModuleProvider", providerName);
          }
        }
        debugLog = {
          polyfills: new Set(),
          polyfillsSupport: undefined,
          found: false,
          providers: new Set(),
          missingDeps: new Set()
        };
        (_provider$pre = provider.pre) == null ? void 0 : _provider$pre.apply(this, arguments);
      },
      post: function post() {
        var _provider$post;
        (_provider$post = provider.post) == null ? void 0 : _provider$post.apply(this, arguments);
        if (missingDependencies !== false) {
          if (missingDependencies.log === "per-file") {
            deps.logMissing(debugLog.missingDeps);
          } else {
            deps.laterLogMissing(debugLog.missingDeps);
          }
        }
        if (!debug) return;
        if (this.filename) console.log("\n[" + this.filename + "]");
        if (debugLog.polyfills.size === 0) {
          console.log(method === "entry-global" ? debugLog.found ? "Based on your targets, the " + providerName + " polyfill did not add any polyfill." : "The entry point for the " + providerName + " polyfill has not been found." : "Based on your code and targets, the " + providerName + " polyfill did not add any polyfill.");
          return;
        }
        if (method === "entry-global") {
          console.log("The " + providerName + " polyfill entry has been replaced with " + "the following polyfills:");
        } else {
          console.log("The " + providerName + " polyfill added the following polyfills:");
        }
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = debugLog.polyfills[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var _name = _step.value;

            var _debugLog$polyfillsSu2;
            if ((_debugLog$polyfillsSu2 = debugLog.polyfillsSupport) != null && _debugLog$polyfillsSu2[_name]) {
              var filteredTargets = (0, _helperCompilationTargets.getInclusionReasons)(_name, targets, debugLog.polyfillsSupport);
              var formattedTargets = JSON.stringify(filteredTargets).replace(/,/g, ", ").replace(/^\{"/, '{ "').replace(/"\}$/, '" }');
              console.log("  " + _name + " " + formattedTargets);
            } else {
              console.log("  " + _name);
            }
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"]) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
    };
  });
}
function mapGetOr(map, key, getDefault) {
  var val = map.get(key);
  if (val === undefined) {
    val = getDefault();
    map.set(key, val);
  }
  return val;
}
function isEmpty(obj) {
  return Object.keys(obj).length === 0;
}
