// @babel/helper-define-polyfill-provider/lib/normalize-options.js
"use strict";

exports.__esModule = true;
exports.applyMissingDependenciesDefaults = applyMissingDependenciesDefaults;
exports.validateIncludeExclude = validateIncludeExclude;
var _utils = require("./utils");
function patternToRegExp(pattern) {
  if (pattern instanceof RegExp) return pattern;
  try {
    return new RegExp("^" + pattern + "$");
  } catch (_unused) {
    return null;
  }
}
function buildUnusedError(label, unused) {
  if (!unused.length) return "";
  return "  - The following \"" + label + "\" patterns didn't match any polyfill:\n" + unused.map(function (original) {
    return "    " + String(original) + "\n";
  }).join("");
}
function buldDuplicatesError(duplicates) {
  if (!duplicates.size) return "";
  return "  - The following polyfills were matched both by \"include\" and \"exclude\" patterns:\n" + Array.from(duplicates, function (name) {
    return "    " + name + "\n";
  }).join("");
}
function validateIncludeExclude(provider, polyfills, includePatterns, excludePatterns) {
  var current = undefined;
  var filter = function filter(pattern) {
    var regexp = patternToRegExp(pattern);
    if (!regexp) return false;
    var matched = false;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = polyfills.keys()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var polyfill = _step.value;

        if (regexp.test(polyfill)) {
          matched = true;
          current.add(polyfill);
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"]) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return !matched;
  };

  // prettier-ignore
  var include = current = new Set();
  var unusedInclude = Array.from(includePatterns).filter(filter);

  // prettier-ignore
  var exclude = current = new Set();
  var unusedExclude = Array.from(excludePatterns).filter(filter);
  var duplicates = (0, _utils.intersection)(include, exclude);
  if (duplicates.size > 0 || unusedInclude.length > 0 || unusedExclude.length > 0) {
    throw new Error("Error while validating the \"" + provider + "\" provider options:\n" + buildUnusedError("include", unusedInclude) + buildUnusedError("exclude", unusedExclude) + buldDuplicatesError(duplicates));
  }
  return {
    include: include,
    exclude: exclude
  };
}
function applyMissingDependenciesDefaults(options, babelApi) {
  var _options$missingDependencies = options.missingDependencies;
  var missingDependencies = _options$missingDependencies === undefined ? {} : _options$missingDependencies;

  if (missingDependencies === false) return false;
  var caller = babelApi.caller(function (caller) {
    return caller == null ? void 0 : caller.name;
  });
  var _missingDependencies$log = missingDependencies.log;
  var log = _missingDependencies$log === undefined ? "deferred" : _missingDependencies$log;
  var _missingDependencies$inject = missingDependencies.inject;
  var inject = _missingDependencies$inject === undefined ? caller === "rollup-plugin-babel" ? "throw" : "import" : _missingDependencies$inject;
  var _missingDependencies$all = missingDependencies.all;
  var all = _missingDependencies$all === undefined ? false : _missingDependencies$all;

  return {
    log: log,
    inject: inject,
    all: all
  };
}
