// @babel/helper-define-polyfill-provider/lib/utils.js
"use strict";

var _templateObject = _taggedTemplateLiteral(["require(", ")"], ["require(", ")"]),
    _templateObject2 = _taggedTemplateLiteral(["\n                  var ", " = require(", ").", "\n                "], ["\n                  var ", " = require(", ").", "\n                "]),
    _templateObject3 = _taggedTemplateLiteral(["var ", " = require(", ")"], ["var ", " = require(", ")"]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

exports.__esModule = true;
exports.createUtilsGetter = createUtilsGetter;
exports.getImportSource = getImportSource;
exports.getRequireSource = getRequireSource;
exports.has = has;
exports.intersection = intersection;
exports.resolveKey = resolveKey;
exports.resolveSource = resolveSource;
var _babel = _interopRequireWildcard(require("@babel/core"));
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function (nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
    return { "default": obj };
  }var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {
    return cache.get(obj);
  }var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }newObj["default"] = obj;if (cache) {
    cache.set(obj, newObj);
  }return newObj;
}

var _ref = _babel["default"] || _babel;

var t = _ref.types;
var template = _ref.template;

function intersection(a, b) {
  var result = new Set();
  a.forEach(function (v) {
    return b.has(v) && result.add(v);
  });
  return result;
}
function has(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
}
function resolve(_x3) {
  var _arguments = arguments;
  var _again = true;

  _function: while (_again) {
    var path = _x3;
    _again = false;
    var resolved = _arguments.length <= 1 || _arguments[1] === undefined ? new Set() : _arguments[1];

    if (resolved.has(path)) return;
    resolved.add(path);
    if (path.isVariableDeclarator()) {
      if (path.get("id").isIdentifier()) {
        _arguments = [_x3 = path.get("init"), resolved];
        _again = true;
        resolved = undefined;
        continue _function;
      }
    } else if (path.isReferencedIdentifier()) {
      var binding = path.scope.getBinding(path.node.name);
      if (!binding) return path;
      if (!binding.constant) return;
      _arguments = [_x3 = binding.path, resolved];
      _again = true;
      resolved = binding = undefined;
      continue _function;
    }
    return path;
  }
}
function resolveId(path) {
  if (path.isIdentifier() && !path.scope.hasBinding(path.node.name, /* noGlobals */true)) {
    return path.node.name;
  }
  var resolved = resolve(path);
  if (resolved != null && resolved.isIdentifier()) {
    return resolved.node.name;
  }
}
function resolveKey(path) {
  var computed = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];
  var scope = path.scope;

  if (path.isStringLiteral()) return path.node.value;
  var isIdentifier = path.isIdentifier();
  if (isIdentifier && !(computed || path.parent.computed)) {
    return path.node.name;
  }
  if (computed && path.isMemberExpression() && path.get("object").isIdentifier({
    name: "Symbol"
  }) && !scope.hasBinding("Symbol", /* noGlobals */true)) {
    var sym = resolveKey(path.get("property"), path.node.computed);
    if (sym) return "Symbol." + sym;
  }
  if (isIdentifier ? scope.hasBinding(path.node.name, /* noGlobals */true) : path.isPure()) {
    var _path$evaluate = path.evaluate();

    var value = _path$evaluate.value;

    if (typeof value === "string") return value;
  }
}
function resolveSource(obj) {
  if (obj.isMemberExpression() && obj.get("property").isIdentifier({
    name: "prototype"
  })) {
    var _id = resolveId(obj.get("object"));
    if (_id) {
      return {
        id: _id,
        placement: "prototype"
      };
    }
    return {
      id: null,
      placement: null
    };
  }
  var id = resolveId(obj);
  if (id) {
    return {
      id: id,
      placement: "static"
    };
  }
  var path = resolve(obj);
  switch (path == null ? void 0 : path.type) {
    case "RegExpLiteral":
      return {
        id: "RegExp",
        placement: "prototype"
      };
    case "FunctionExpression":
      return {
        id: "Function",
        placement: "prototype"
      };
    case "StringLiteral":
      return {
        id: "String",
        placement: "prototype"
      };
    case "NumberLiteral":
      return {
        id: "Number",
        placement: "prototype"
      };
    case "BooleanLiteral":
      return {
        id: "Boolean",
        placement: "prototype"
      };
    case "ObjectExpression":
      return {
        id: "Object",
        placement: "prototype"
      };
    case "ArrayExpression":
      return {
        id: "Array",
        placement: "prototype"
      };
  }
  return {
    id: null,
    placement: null
  };
}
function getImportSource(_ref2) {
  var node = _ref2.node;

  if (node.specifiers.length === 0) return node.source.value;
}
function getRequireSource(_ref3) {
  var node = _ref3.node;

  if (!t.isExpressionStatement(node)) return;
  var expression = node.expression;

  if (t.isCallExpression(expression) && t.isIdentifier(expression.callee) && expression.callee.name === "require" && expression.arguments.length === 1 && t.isStringLiteral(expression.arguments[0])) {
    return expression.arguments[0].value;
  }
}
function hoist(node) {
  // @ts-expect-error
  node._blockHoist = 3;
  return node;
}
function createUtilsGetter(cache) {
  return function (path) {
    var prog = path.findParent(function (p) {
      return p.isProgram();
    });
    return {
      injectGlobalImport: function injectGlobalImport(url, moduleName) {
        cache.storeAnonymous(prog, url, moduleName, function (isScript, source) {
          return isScript ? template.statement.ast(_templateObject, source) : t.importDeclaration([], source);
        });
      },
      injectNamedImport: function injectNamedImport(url, name, hint, moduleName) {
        if (hint === undefined) hint = name;
        return (function () {
          return cache.storeNamed(prog, url, name, moduleName, function (isScript, source, name) {
            var id = prog.scope.generateUidIdentifier(hint);
            return {
              node: isScript ? hoist(template.statement.ast(_templateObject2, id, source, name)) : t.importDeclaration([t.importSpecifier(id, name)], source),
              name: id.name
            };
          });
        })();
      },
      injectDefaultImport: function injectDefaultImport(url, hint, moduleName) {
        if (hint === undefined) hint = url;
        return (function () {
          return cache.storeNamed(prog, url, "default", moduleName, function (isScript, source) {
            var id = prog.scope.generateUidIdentifier(hint);
            return {
              node: isScript ? hoist(template.statement.ast(_templateObject3, id, source)) : t.importDeclaration([t.importDefaultSpecifier(id)], source),
              name: id.name
            };
          });
        })();
      }
    };
  };
}
