// @babel/helper-module-transforms/lib/index.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var _templateObject = _taggedTemplateLiteral(["var NAME = SOURCE;"], ["var NAME = SOURCE;"]),
    _templateObject2 = _taggedTemplateLiteral(["\n            Object.defineProperty(EXPORTS, \"NAME\", {\n              enumerable: true,\n              get: function() {\n                return NAMESPACE;\n              }\n            });\n          "], ["\n            Object.defineProperty(EXPORTS, \"NAME\", {\n              enumerable: true,\n              get: function() {\n                return NAMESPACE;\n              }\n            });\n          "]),
    _templateObject3 = _taggedTemplateLiteral(["EXPORTS.NAME = NAMESPACE;"], ["EXPORTS.NAME = NAMESPACE;"]),
    _templateObject4 = _taggedTemplateLiteral(["\n      ", ".", " = ", ";\n    "], ["\n      ", ".", " = ", ";\n    "]),
    _templateObject5 = _taggedTemplateLiteral(["\n      ", "[\"", "\"] = ", ";\n    "], ["\n      ", "[\"", "\"] = ", ";\n    "]),
    _templateObject6 = _taggedTemplateLiteral(["\n      Object.defineProperty(", ", \"", "\", {\n        enumerable: true,\n        get: function() {\n          return ", ";\n        },\n      });\n    "], ["\n      Object.defineProperty(", ", \"", "\", {\n        enumerable: true,\n        get: function() {\n          return ", ";\n        },\n      });\n    "]),
    _templateObject7 = _taggedTemplateLiteral(["\n        EXPORTS.__esModule = true;\n      "], ["\n        EXPORTS.__esModule = true;\n      "]),
    _templateObject8 = _taggedTemplateLiteral(["\n        Object.defineProperty(EXPORTS, \"__esModule\", {\n          value: true,\n        });\n      "], ["\n        Object.defineProperty(EXPORTS, \"__esModule\", {\n          value: true,\n        });\n      "]),
    _templateObject9 = _taggedTemplateLiteral(["\n        Object.keys(NAMESPACE).forEach(function(key) {\n          if (key === \"default\" || key === \"__esModule\") return;\n          VERIFY_NAME_LIST;\n          if (key in EXPORTS && EXPORTS[key] === NAMESPACE[key]) return;\n\n          EXPORTS[key] = NAMESPACE[key];\n        });\n      "], ["\n        Object.keys(NAMESPACE).forEach(function(key) {\n          if (key === \"default\" || key === \"__esModule\") return;\n          VERIFY_NAME_LIST;\n          if (key in EXPORTS && EXPORTS[key] === NAMESPACE[key]) return;\n\n          EXPORTS[key] = NAMESPACE[key];\n        });\n      "]),
    _templateObject10 = _taggedTemplateLiteral(["\n        Object.keys(NAMESPACE).forEach(function(key) {\n          if (key === \"default\" || key === \"__esModule\") return;\n          VERIFY_NAME_LIST;\n          if (key in EXPORTS && EXPORTS[key] === NAMESPACE[key]) return;\n\n          Object.defineProperty(EXPORTS, key, {\n            enumerable: true,\n            get: function() {\n              return NAMESPACE[key];\n            },\n          });\n        });\n    "], ["\n        Object.keys(NAMESPACE).forEach(function(key) {\n          if (key === \"default\" || key === \"__esModule\") return;\n          VERIFY_NAME_LIST;\n          if (key in EXPORTS && EXPORTS[key] === NAMESPACE[key]) return;\n\n          Object.defineProperty(EXPORTS, key, {\n            enumerable: true,\n            get: function() {\n              return NAMESPACE[key];\n            },\n          });\n        });\n    "]),
    _templateObject11 = _taggedTemplateLiteral(["\n            if (Object.prototype.hasOwnProperty.call(EXPORTS_LIST, key)) return;\n          "], ["\n            if (Object.prototype.hasOwnProperty.call(EXPORTS_LIST, key)) return;\n          "]),
    _templateObject12 = _taggedTemplateLiteral(["", "[\"", "\"] = ", ""], ["", "[\"", "\"] = ", ""]),
    _templateObject13 = _taggedTemplateLiteral(["", ".", " = ", ""], ["", ".", " = ", ""]),
    _templateObject14 = _taggedTemplateLiteral(["\n      Object.defineProperty(", ", \"", "\", {\n        enumerable: true,\n        value: void 0,\n        writable: true\n      })[\"", "\"] = ", ""], ["\n      Object.defineProperty(", ", \"", "\", {\n        enumerable: true,\n        value: void 0,\n        writable: true\n      })[\"", "\"] = ", ""]);

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "buildDynamicImport", {
  enumerable: true,
  get: function get() {
    return _dynamicImport.buildDynamicImport;
  }
});
exports.buildNamespaceInitStatements = buildNamespaceInitStatements;
exports.ensureStatementsHoisted = ensureStatementsHoisted;
Object.defineProperty(exports, "getModuleName", {
  enumerable: true,
  get: function get() {
    return _getModuleName["default"];
  }
});
Object.defineProperty(exports, "hasExports", {
  enumerable: true,
  get: function get() {
    return _normalizeAndLoadMetadata.hasExports;
  }
});
Object.defineProperty(exports, "isModule", {
  enumerable: true,
  get: function get() {
    return _helperModuleImports.isModule;
  }
});
Object.defineProperty(exports, "isSideEffectImport", {
  enumerable: true,
  get: function get() {
    return _normalizeAndLoadMetadata.isSideEffectImport;
  }
});
exports.rewriteModuleStatementsAndPrepareHeader = rewriteModuleStatementsAndPrepareHeader;
Object.defineProperty(exports, "rewriteThis", {
  enumerable: true,
  get: function get() {
    return _rewriteThis["default"];
  }
});
exports.wrapInterop = wrapInterop;
var _assert = require("assert");
var _core = require("@babel/core");
var _helperModuleImports = require("@babel/helper-module-imports");
var _rewriteThis = require("./rewrite-this.js");
var _rewriteLiveReferences = require("./rewrite-live-references.js");
var _normalizeAndLoadMetadata = require("./normalize-and-load-metadata.js");
var Lazy = require("./lazy-modules.js");
var _dynamicImport = require("./dynamic-import.js");
var _getModuleName = require("./get-module-name.js");
{
  exports.getDynamicImportSource = require("./dynamic-import").getDynamicImportSource;
}
function rewriteModuleStatementsAndPrepareHeader(path, _ref) {
  var exportName = _ref.exportName;
  var strict = _ref.strict;
  var allowTopLevelThis = _ref.allowTopLevelThis;
  var strictMode = _ref.strictMode;
  var noInterop = _ref.noInterop;
  var _ref$importInterop = _ref.importInterop;
  var importInterop = _ref$importInterop === undefined ? noInterop ? "none" : "babel" : _ref$importInterop;
  var lazy = _ref.lazy;
  var _ref$getWrapperPayload = _ref.getWrapperPayload;
  var getWrapperPayload = _ref$getWrapperPayload === undefined ? Lazy.toGetWrapperPayload(lazy != null ? lazy : false) : _ref$getWrapperPayload;
  var _ref$wrapReference = _ref.wrapReference;
  var wrapReference = _ref$wrapReference === undefined ? Lazy.wrapReference : _ref$wrapReference;
  var esNamespaceOnly = _ref.esNamespaceOnly;
  var filename = _ref.filename;
  var _ref$constantReexports = _ref.constantReexports;
  var constantReexports = _ref$constantReexports === undefined ? arguments[1].loose : _ref$constantReexports;
  var _ref$enumerableModuleMeta = _ref.enumerableModuleMeta;
  var enumerableModuleMeta = _ref$enumerableModuleMeta === undefined ? arguments[1].loose : _ref$enumerableModuleMeta;
  var noIncompleteNsImportDetection = _ref.noIncompleteNsImportDetection;
  return (function () {
    (0, _normalizeAndLoadMetadata.validateImportInteropOption)(importInterop);
    _assert((0, _helperModuleImports.isModule)(path), "Cannot process module statements in a script");
    path.node.sourceType = "script";
    var meta = (0, _normalizeAndLoadMetadata["default"])(path, exportName, {
      importInterop: importInterop,
      initializeReexports: constantReexports,
      getWrapperPayload: getWrapperPayload,
      esNamespaceOnly: esNamespaceOnly,
      filename: filename
    });
    if (!allowTopLevelThis) {
      (0, _rewriteThis["default"])(path);
    }
    (0, _rewriteLiveReferences["default"])(path, meta, wrapReference);
    if (strictMode !== false) {
      var hasStrict = path.node.directives.some(function (directive) {
        return directive.value.value === "use strict";
      });
      if (!hasStrict) {
        path.unshiftContainer("directives", _core.types.directive(_core.types.directiveLiteral("use strict")));
      }
    }
    var headers = [];
    if ((0, _normalizeAndLoadMetadata.hasExports)(meta) && !strict) {
      headers.push(buildESModuleHeader(meta, enumerableModuleMeta));
    }
    var nameList = buildExportNameListDeclaration(path, meta);
    if (nameList) {
      meta.exportNameListName = nameList.name;
      headers.push(nameList.statement);
    }
    headers.push.apply(headers, _toConsumableArray(buildExportInitializationStatements(path, meta, wrapReference, constantReexports, noIncompleteNsImportDetection)));
    return {
      meta: meta,
      headers: headers
    };
  }).apply(this, arguments);
}
function ensureStatementsHoisted(statements) {
  statements.forEach(function (header) {
    header._blockHoist = 3;
  });
}
function wrapInterop(programPath, expr, type) {
  if (type === "none") {
    return null;
  }
  if (type === "node-namespace") {
    return _core.types.callExpression(programPath.hub.addHelper("interopRequireWildcard"), [expr, _core.types.booleanLiteral(true)]);
  } else if (type === "node-default") {
    return null;
  }
  var helper = undefined;
  if (type === "default") {
    helper = "interopRequireDefault";
  } else if (type === "namespace") {
    helper = "interopRequireWildcard";
  } else {
    throw new Error("Unknown interop: " + type);
  }
  return _core.types.callExpression(programPath.hub.addHelper(helper), [expr]);
}
function buildNamespaceInitStatements(metadata, sourceMetadata) {
  var constantReexports = arguments.length <= 2 || arguments[2] === undefined ? false : arguments[2];
  var wrapReference = arguments.length <= 3 || arguments[3] === undefined ? Lazy.wrapReference : arguments[3];

  var _wrapReference;
  var statements = [];
  var srcNamespaceId = _core.types.identifier(sourceMetadata.name);
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = sourceMetadata.importsNamespace[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var localName = _step.value;

      if (localName === sourceMetadata.name) continue;
      statements.push(_core.template.statement(_templateObject)({
        NAME: localName,
        SOURCE: _core.types.cloneNode(srcNamespaceId)
      }));
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  var srcNamespace = (_wrapReference = wrapReference(srcNamespaceId, sourceMetadata.wrap)) != null ? _wrapReference : srcNamespaceId;
  if (constantReexports) {
    statements.push.apply(statements, _toConsumableArray(buildReexportsFromMeta(metadata, sourceMetadata, true, wrapReference)));
  }
  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = sourceMetadata.reexportNamespace[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var exportName = _step2.value;

      statements.push((!_core.types.isIdentifier(srcNamespace) ? _core.template.statement(_templateObject2) : _core.template.statement(_templateObject3))({
        EXPORTS: metadata.exportName,
        NAME: exportName,
        NAMESPACE: _core.types.cloneNode(srcNamespace)
      }));
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
        _iterator2["return"]();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  if (sourceMetadata.reexportAll) {
    var statement = buildNamespaceReexport(metadata, _core.types.cloneNode(srcNamespace), constantReexports);
    statement.loc = sourceMetadata.reexportAll.loc;
    statements.push(statement);
  }
  return statements;
}
var ReexportTemplate = {
  constant: function constant(_ref2) {
    var exports = _ref2.exports;
    var exportName = _ref2.exportName;
    var namespaceImport = _ref2.namespaceImport;
    return _core.template.statement.ast(_templateObject4, exports, exportName, namespaceImport);
  },
  constantComputed: function constantComputed(_ref3) {
    var exports = _ref3.exports;
    var exportName = _ref3.exportName;
    var namespaceImport = _ref3.namespaceImport;
    return _core.template.statement.ast(_templateObject5, exports, exportName, namespaceImport);
  },
  spec: function spec(_ref4) {
    var exports = _ref4.exports;
    var exportName = _ref4.exportName;
    var namespaceImport = _ref4.namespaceImport;
    return _core.template.statement.ast(_templateObject6, exports, exportName, namespaceImport);
  }
};
function buildReexportsFromMeta(meta, metadata, constantReexports, wrapReference) {
  var _wrapReference2;
  var namespace = _core.types.identifier(metadata.name);
  namespace = (_wrapReference2 = wrapReference(namespace, metadata.wrap)) != null ? _wrapReference2 : namespace;
  var stringSpecifiers = meta.stringSpecifiers;

  return Array.from(metadata.reexports, function (_ref5) {
    var _ref52 = _slicedToArray(_ref5, 2);

    var exportName = _ref52[0];
    var importName = _ref52[1];

    var namespaceImport = _core.types.cloneNode(namespace);
    if (importName === "default" && metadata.interop === "node-default") {} else if (stringSpecifiers.has(importName)) {
      namespaceImport = _core.types.memberExpression(namespaceImport, _core.types.stringLiteral(importName), true);
    } else {
      namespaceImport = _core.types.memberExpression(namespaceImport, _core.types.identifier(importName));
    }
    var astNodes = {
      exports: meta.exportName,
      exportName: exportName,
      namespaceImport: namespaceImport
    };
    if (constantReexports || _core.types.isIdentifier(namespaceImport)) {
      if (stringSpecifiers.has(exportName)) {
        return ReexportTemplate.constantComputed(astNodes);
      } else {
        return ReexportTemplate.constant(astNodes);
      }
    } else {
      return ReexportTemplate.spec(astNodes);
    }
  });
}
function buildESModuleHeader(metadata) {
  var enumerableModuleMeta = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];

  return (enumerableModuleMeta ? _core.template.statement(_templateObject7) : _core.template.statement(_templateObject8))({
    EXPORTS: metadata.exportName
  });
}
function buildNamespaceReexport(metadata, namespace, constantReexports) {
  return (constantReexports ? _core.template.statement(_templateObject9) : _core.template.statement(_templateObject10))({
    NAMESPACE: namespace,
    EXPORTS: metadata.exportName,
    VERIFY_NAME_LIST: metadata.exportNameListName ? (0, _core.template)(_templateObject11)({
      EXPORTS_LIST: metadata.exportNameListName
    }) : null
  });
}
function buildExportNameListDeclaration(programPath, metadata) {
  var exportedVars = Object.create(null);
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = metadata.local.values()[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var data = _step3.value;
      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        for (var _iterator5 = data.names[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
          var _name = _step5.value;

          exportedVars[_name] = true;
        }
      } catch (err) {
        _didIteratorError5 = true;
        _iteratorError5 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion5 && _iterator5["return"]) {
            _iterator5["return"]();
          }
        } finally {
          if (_didIteratorError5) {
            throw _iteratorError5;
          }
        }
      }
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  var hasReexport = false;
  var _iteratorNormalCompletion4 = true;
  var _didIteratorError4 = false;
  var _iteratorError4 = undefined;

  try {
    for (var _iterator4 = metadata.source.values()[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
      var data = _step4.value;
      var _iteratorNormalCompletion6 = true;
      var _didIteratorError6 = false;
      var _iteratorError6 = undefined;

      try {
        for (var _iterator6 = data.reexports.keys()[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
          var exportName = _step6.value;

          exportedVars[exportName] = true;
        }
      } catch (err) {
        _didIteratorError6 = true;
        _iteratorError6 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion6 && _iterator6["return"]) {
            _iterator6["return"]();
          }
        } finally {
          if (_didIteratorError6) {
            throw _iteratorError6;
          }
        }
      }

      var _iteratorNormalCompletion7 = true;
      var _didIteratorError7 = false;
      var _iteratorError7 = undefined;

      try {
        for (var _iterator7 = data.reexportNamespace[Symbol.iterator](), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
          var exportName = _step7.value;

          exportedVars[exportName] = true;
        }
      } catch (err) {
        _didIteratorError7 = true;
        _iteratorError7 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion7 && _iterator7["return"]) {
            _iterator7["return"]();
          }
        } finally {
          if (_didIteratorError7) {
            throw _iteratorError7;
          }
        }
      }

      hasReexport = hasReexport || !!data.reexportAll;
    }
  } catch (err) {
    _didIteratorError4 = true;
    _iteratorError4 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion4 && _iterator4["return"]) {
        _iterator4["return"]();
      }
    } finally {
      if (_didIteratorError4) {
        throw _iteratorError4;
      }
    }
  }

  if (!hasReexport || Object.keys(exportedVars).length === 0) return null;
  var name = programPath.scope.generateUidIdentifier("exportNames");
  delete exportedVars["default"];
  return {
    name: name.name,
    statement: _core.types.variableDeclaration("var", [_core.types.variableDeclarator(name, _core.types.valueToNode(exportedVars))])
  };
}
function buildExportInitializationStatements(programPath, metadata, wrapReference) {
  var constantReexports = arguments.length <= 3 || arguments[3] === undefined ? false : arguments[3];
  var noIncompleteNsImportDetection = arguments.length <= 4 || arguments[4] === undefined ? false : arguments[4];

  var initStatements = [];
  var _iteratorNormalCompletion8 = true;
  var _didIteratorError8 = false;
  var _iteratorError8 = undefined;

  try {
    for (var _iterator8 = metadata.local[Symbol.iterator](), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
      var _step8$value = _slicedToArray(_step8.value, 2);

      var localName = _step8$value[0];
      var data = _step8$value[1];

      if (data.kind === "import") {} else if (data.kind === "hoisted") {
        initStatements.push([data.names[0], buildInitStatement(metadata, data.names, _core.types.identifier(localName))]);
      } else if (!noIncompleteNsImportDetection) {
        var _iteratorNormalCompletion11 = true;
        var _didIteratorError11 = false;
        var _iteratorError11 = undefined;

        try {
          for (var _iterator11 = data.names[Symbol.iterator](), _step11; !(_iteratorNormalCompletion11 = (_step11 = _iterator11.next()).done); _iteratorNormalCompletion11 = true) {
            var exportName = _step11.value;

            initStatements.push([exportName, null]);
          }
        } catch (err) {
          _didIteratorError11 = true;
          _iteratorError11 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion11 && _iterator11["return"]) {
              _iterator11["return"]();
            }
          } finally {
            if (_didIteratorError11) {
              throw _iteratorError11;
            }
          }
        }
      }
    }
  } catch (err) {
    _didIteratorError8 = true;
    _iteratorError8 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion8 && _iterator8["return"]) {
        _iterator8["return"]();
      }
    } finally {
      if (_didIteratorError8) {
        throw _iteratorError8;
      }
    }
  }

  var _iteratorNormalCompletion9 = true;
  var _didIteratorError9 = false;
  var _iteratorError9 = undefined;

  try {
    for (var _iterator9 = metadata.source.values()[Symbol.iterator](), _step9; !(_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done); _iteratorNormalCompletion9 = true) {
      var data = _step9.value;

      if (!constantReexports) {
        var reexportsStatements = buildReexportsFromMeta(metadata, data, false, wrapReference);
        var reexports = [].concat(_toConsumableArray(data.reexports.keys()));
        for (var i = 0; i < reexportsStatements.length; i++) {
          initStatements.push([reexports[i], reexportsStatements[i]]);
        }
      }
      if (!noIncompleteNsImportDetection) {
        var _iteratorNormalCompletion12 = true;
        var _didIteratorError12 = false;
        var _iteratorError12 = undefined;

        try {
          for (var _iterator12 = data.reexportNamespace[Symbol.iterator](), _step12; !(_iteratorNormalCompletion12 = (_step12 = _iterator12.next()).done); _iteratorNormalCompletion12 = true) {
            var exportName = _step12.value;

            initStatements.push([exportName, null]);
          }
        } catch (err) {
          _didIteratorError12 = true;
          _iteratorError12 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion12 && _iterator12["return"]) {
              _iterator12["return"]();
            }
          } finally {
            if (_didIteratorError12) {
              throw _iteratorError12;
            }
          }
        }
      }
    }
  } catch (err) {
    _didIteratorError9 = true;
    _iteratorError9 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion9 && _iterator9["return"]) {
        _iterator9["return"]();
      }
    } finally {
      if (_didIteratorError9) {
        throw _iteratorError9;
      }
    }
  }

  initStatements.sort(function (_ref6, _ref7) {
    var _ref62 = _slicedToArray(_ref6, 1);

    var a = _ref62[0];

    var _ref72 = _slicedToArray(_ref7, 1);

    var b = _ref72[0];

    if (a < b) return -1;
    if (b < a) return 1;
    return 0;
  });
  var results = [];
  if (noIncompleteNsImportDetection) {
    var _iteratorNormalCompletion10 = true;
    var _didIteratorError10 = false;
    var _iteratorError10 = undefined;

    try {
      for (var _iterator10 = initStatements[Symbol.iterator](), _step10; !(_iteratorNormalCompletion10 = (_step10 = _iterator10.next()).done); _iteratorNormalCompletion10 = true) {
        var _step10$value = _slicedToArray(_step10.value, 2);

        var initStatement = _step10$value[1];

        results.push(initStatement);
      }
    } catch (err) {
      _didIteratorError10 = true;
      _iteratorError10 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion10 && _iterator10["return"]) {
          _iterator10["return"]();
        }
      } finally {
        if (_didIteratorError10) {
          throw _iteratorError10;
        }
      }
    }
  } else {
    var chunkSize = 100;
    for (var i = 0; i < initStatements.length; i += chunkSize) {
      var uninitializedExportNames = [];
      for (var j = 0; j < chunkSize && i + j < initStatements.length; j++) {
        var _initStatements = _slicedToArray(initStatements[i + j], 2);

        var exportName = _initStatements[0];
        var initStatement = _initStatements[1];

        if (initStatement !== null) {
          if (uninitializedExportNames.length > 0) {
            results.push(buildInitStatement(metadata, uninitializedExportNames, programPath.scope.buildUndefinedNode()));
            uninitializedExportNames = [];
          }
          results.push(initStatement);
        } else {
          uninitializedExportNames.push(exportName);
        }
      }
      if (uninitializedExportNames.length > 0) {
        results.push(buildInitStatement(metadata, uninitializedExportNames, programPath.scope.buildUndefinedNode()));
      }
    }
  }
  return results;
}
var InitTemplate = {
  computed: function computed(_ref8) {
    var exports = _ref8.exports;
    var name = _ref8.name;
    var value = _ref8.value;
    return _core.template.expression.ast(_templateObject12, exports, name, value);
  },
  "default": function _default(_ref9) {
    var exports = _ref9.exports;
    var name = _ref9.name;
    var value = _ref9.value;
    return _core.template.expression.ast(_templateObject13, exports, name, value);
  },
  define: function define(_ref10) {
    var exports = _ref10.exports;
    var name = _ref10.name;
    var value = _ref10.value;
    return _core.template.expression.ast(_templateObject14, exports, name, name, value);
  }
};
function buildInitStatement(metadata, exportNames, initExpr) {
  var stringSpecifiers = metadata.stringSpecifiers;
  var exports = metadata.exportName;

  return _core.types.expressionStatement(exportNames.reduce(function (value, name) {
    var params = {
      exports: exports,
      name: name,
      value: value
    };
    if (name === "__proto__") {
      return InitTemplate.define(params);
    }
    if (stringSpecifiers.has(name)) {
      return InitTemplate.computed(params);
    }
    return InitTemplate["default"](params);
  }, initExpr));
}
