// @babel/core/lib/parser/index.js
"use strict";

var marked0$0 = [parser].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = parser;
function _parser() {
  var data = require("@babel/parser");
  _parser = function () {
    return data;
  };
  return data;
}
function _codeFrame() {
  var data = require("@babel/code-frame");
  _codeFrame = function () {
    return data;
  };
  return data;
}
var _missingPluginHelper = require("./util/missing-plugin-helper.js");
function parser(pluginPasses, _ref, code) {
  var parserOpts = _ref.parserOpts;
  var _ref$highlightCode = _ref.highlightCode;
  var highlightCode = _ref$highlightCode === undefined ? true : _ref$highlightCode;
  var _ref$filename = _ref.filename;
  var filename = _ref$filename === undefined ? "unknown" : _ref$filename;

  var results, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, plugins, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, plugin, parserOverride, ast, loc, missingPlugin, codeFrame;

  return regeneratorRuntime.wrap(function parser$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.prev = 0;
        results = [];
        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 5;
        _iterator = pluginPasses[Symbol.iterator]();

      case 7:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 31;
          break;
        }

        plugins = _step.value;
        _iteratorNormalCompletion2 = true;
        _didIteratorError2 = false;
        _iteratorError2 = undefined;
        context$1$0.prev = 12;

        for (_iterator2 = plugins[Symbol.iterator](); !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          plugin = _step2.value;
          parserOverride = plugin.parserOverride;

          if (parserOverride) {
            ast = parserOverride(code, parserOpts, _parser().parse);

            if (ast !== undefined) results.push(ast);
          }
        }
        context$1$0.next = 20;
        break;

      case 16:
        context$1$0.prev = 16;
        context$1$0.t0 = context$1$0["catch"](12);
        _didIteratorError2 = true;
        _iteratorError2 = context$1$0.t0;

      case 20:
        context$1$0.prev = 20;
        context$1$0.prev = 21;

        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }

      case 23:
        context$1$0.prev = 23;

        if (!_didIteratorError2) {
          context$1$0.next = 26;
          break;
        }

        throw _iteratorError2;

      case 26:
        return context$1$0.finish(23);

      case 27:
        return context$1$0.finish(20);

      case 28:
        _iteratorNormalCompletion = true;
        context$1$0.next = 7;
        break;

      case 31:
        context$1$0.next = 37;
        break;

      case 33:
        context$1$0.prev = 33;
        context$1$0.t1 = context$1$0["catch"](5);
        _didIteratorError = true;
        _iteratorError = context$1$0.t1;

      case 37:
        context$1$0.prev = 37;
        context$1$0.prev = 38;

        if (!_iteratorNormalCompletion && _iterator["return"]) {
          _iterator["return"]();
        }

      case 40:
        context$1$0.prev = 40;

        if (!_didIteratorError) {
          context$1$0.next = 43;
          break;
        }

        throw _iteratorError;

      case 43:
        return context$1$0.finish(40);

      case 44:
        return context$1$0.finish(37);

      case 45:
        if (!(results.length === 0)) {
          context$1$0.next = 49;
          break;
        }

        return context$1$0.abrupt("return", (0, _parser().parse)(code, parserOpts));

      case 49:
        if (!(results.length === 1)) {
          context$1$0.next = 54;
          break;
        }

        return context$1$0.delegateYield([], "t2", 51);

      case 51:
        if (!(typeof results[0].then === "function")) {
          context$1$0.next = 53;
          break;
        }

        throw new Error("You appear to be using an async parser plugin, " + "which your current version of Babel does not support. " + "If you're using a published plugin, you may need to upgrade " + "your @babel/core version.");

      case 53:
        return context$1$0.abrupt("return", results[0]);

      case 54:
        throw new Error("More than one plugin attempted to override parsing.");

      case 57:
        context$1$0.prev = 57;
        context$1$0.t3 = context$1$0["catch"](0);

        if (context$1$0.t3.code === "BABEL_PARSER_SOURCETYPE_MODULE_REQUIRED") {
          context$1$0.t3.message += "\nConsider renaming the file to '.mjs', or setting sourceType:module " + "or sourceType:unambiguous in your Babel config for this file.";
        }
        loc = context$1$0.t3.loc;
        missingPlugin = context$1$0.t3.missingPlugin;

        if (loc) {
          codeFrame = (0, _codeFrame().codeFrameColumns)(code, {
            start: {
              line: loc.line,
              column: loc.column + 1
            }
          }, {
            highlightCode: highlightCode
          });

          if (missingPlugin) {
            context$1$0.t3.message = filename + ": " + (0, _missingPluginHelper["default"])(missingPlugin[0], loc, codeFrame, filename);
          } else {
            context$1$0.t3.message = filename + ": " + context$1$0.t3.message + "\n\n" + codeFrame;
          }
          context$1$0.t3.code = "BABEL_PARSE_ERROR";
        }
        throw context$1$0.t3;

      case 64:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this, [[0, 57], [5, 33, 37, 45], [12, 16, 20, 28], [21,, 23, 27], [38,, 40, 44]]);
}
0 && 0;
