// @babel/core/lib/tools/build-external-helpers.js
"use strict";

var _templateObject = _taggedTemplateLiteral(["\n    (function (root, factory) {\n      if (typeof define === \"function\" && define.amd) {\n        define(AMD_ARGUMENTS, factory);\n      } else if (typeof exports === \"object\") {\n        factory(COMMON_ARGUMENTS);\n      } else {\n        factory(BROWSER_ARGUMENTS);\n      }\n    })(UMD_ROOT, function (FACTORY_PARAMETERS) {\n      FACTORY_BODY\n    });\n  "], ["\n    (function (root, factory) {\n      if (typeof define === \"function\" && define.amd) {\n        define(AMD_ARGUMENTS, factory);\n      } else if (typeof exports === \"object\") {\n        factory(COMMON_ARGUMENTS);\n      } else {\n        factory(BROWSER_ARGUMENTS);\n      }\n    })(UMD_ROOT, function (FACTORY_PARAMETERS) {\n      FACTORY_BODY\n    });\n  "]);

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
function helpers() {
  var data = require("@babel/helpers");
  helpers = function () {
    return data;
  };
  return data;
}
function _generator() {
  var data = require("@babel/generator");
  _generator = function () {
    return data;
  };
  return data;
}
function _template() {
  var data = require("@babel/template");
  _template = function () {
    return data;
  };
  return data;
}
function _t() {
  var data = require("@babel/types");
  _t = function () {
    return data;
  };
  return data;
}

var _t2 = _t();

var arrayExpression = _t2.arrayExpression;
var assignmentExpression = _t2.assignmentExpression;
var binaryExpression = _t2.binaryExpression;
var blockStatement = _t2.blockStatement;
var callExpression = _t2.callExpression;
var cloneNode = _t2.cloneNode;
var conditionalExpression = _t2.conditionalExpression;
var exportNamedDeclaration = _t2.exportNamedDeclaration;
var exportSpecifier = _t2.exportSpecifier;
var expressionStatement = _t2.expressionStatement;
var functionExpression = _t2.functionExpression;
var identifier = _t2.identifier;
var memberExpression = _t2.memberExpression;
var objectExpression = _t2.objectExpression;
var program = _t2.program;
var stringLiteral = _t2.stringLiteral;
var unaryExpression = _t2.unaryExpression;
var variableDeclaration = _t2.variableDeclaration;
var variableDeclarator = _t2.variableDeclarator;

var buildUmdWrapper = function buildUmdWrapper(replacements) {
  return _template()["default"].statement(_templateObject)(replacements);
};
function buildGlobal(allowlist) {
  var namespace = identifier("babelHelpers");
  var body = [];
  var container = functionExpression(null, [identifier("global")], blockStatement(body));
  var tree = program([expressionStatement(callExpression(container, [conditionalExpression(binaryExpression("===", unaryExpression("typeof", identifier("global")), stringLiteral("undefined")), identifier("self"), identifier("global"))]))]);
  body.push(variableDeclaration("var", [variableDeclarator(namespace, assignmentExpression("=", memberExpression(identifier("global"), namespace), objectExpression([])))]));
  buildHelpers(body, namespace, allowlist);
  return tree;
}
function buildModule(allowlist) {
  var body = [];
  var refs = buildHelpers(body, null, allowlist);
  body.unshift(exportNamedDeclaration(null, Object.keys(refs).map(function (name) {
    return exportSpecifier(cloneNode(refs[name]), identifier(name));
  })));
  return program(body, [], "module");
}
function buildUmd(allowlist) {
  var namespace = identifier("babelHelpers");
  var body = [];
  body.push(variableDeclaration("var", [variableDeclarator(namespace, identifier("global"))]));
  buildHelpers(body, namespace, allowlist);
  return program([buildUmdWrapper({
    FACTORY_PARAMETERS: identifier("global"),
    BROWSER_ARGUMENTS: assignmentExpression("=", memberExpression(identifier("root"), namespace), objectExpression([])),
    COMMON_ARGUMENTS: identifier("exports"),
    AMD_ARGUMENTS: arrayExpression([stringLiteral("exports")]),
    FACTORY_BODY: body,
    UMD_ROOT: identifier("this")
  })]);
}
function buildVar(allowlist) {
  var namespace = identifier("babelHelpers");
  var body = [];
  body.push(variableDeclaration("var", [variableDeclarator(namespace, objectExpression([]))]));
  var tree = program(body);
  buildHelpers(body, namespace, allowlist);
  body.push(expressionStatement(namespace));
  return tree;
}
function buildHelpers(body, namespace, allowlist) {
  var getHelperReference = function getHelperReference(name) {
    return namespace ? memberExpression(namespace, identifier(name)) : identifier("_" + name);
  };
  var refs = {};
  helpers().list.forEach(function (name) {
    if (allowlist && !allowlist.includes(name)) return;
    var ref = refs[name] = getHelperReference(name);

    var _helpers$get = helpers().get(name, getHelperReference, namespace ? null : "_" + name, [], namespace ? function (ast, exportName, mapExportBindingAssignments) {
      mapExportBindingAssignments(function (node) {
        return assignmentExpression("=", ref, node);
      });
      ast.body.push(expressionStatement(assignmentExpression("=", ref, identifier(exportName))));
    } : null);

    var nodes = _helpers$get.nodes;

    body.push.apply(body, _toConsumableArray(nodes));
  });
  return refs;
}
function _default(allowlist) {
  var outputType = arguments.length <= 1 || arguments[1] === undefined ? "global" : arguments[1];

  var tree = undefined;
  var build = ({
    global: buildGlobal,
    module: buildModule,
    umd: buildUmd,
    "var": buildVar
  })[outputType];
  if (build) {
    tree = build(allowlist);
  } else {
    throw new Error("Unsupported output type " + outputType);
  }
  return (0, _generator()["default"])(tree).code;
}
0 && 0;
