// @babel/cli/lib/babel/dir.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
function _slash() {
  var data = require("slash");
  _slash = function () {
    return data;
  };
  return data;
}
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _fs() {
  var data = require("fs");
  _fs = function () {
    return data;
  };
  return data;
}
var util = require("./util.js");
var watcher = require("./watcher.js");
function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
        u = i.value;
  } catch (n) {
    return void e(n);
  }i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
        e = arguments;return new Promise(function (r, o) {
      var a = n.apply(t, e);function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }_next(void 0);
    });
  };
}
var FILE_TYPE = Object.freeze({
  NON_COMPILABLE: "NON_COMPILABLE",
  COMPILED: "COMPILED",
  IGNORED: "IGNORED",
  ERR_COMPILATION: "ERR_COMPILATION"
});
function outputFileSync(filePath, data) {
  ((function (v, w) {
    return v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1];
  })(process.versions.node, "10.12") ? _fs().mkdirSync : require("make-dir").sync)(_path().dirname(filePath), {
    recursive: true
  });
  _fs().writeFileSync(filePath, data);
}
function _default(_x) {
  return _ref.apply(this, arguments);
}
function _ref() {
  _ref = _asyncToGenerator(regeneratorRuntime.mark(function callee$1$0(_ref2) {
    var cliOptions = _ref2.cliOptions;
    var babelOptions = _ref2.babelOptions;

    var write, _write, getDest, handleFile, _handleFile, handle, _handle, compiledFiles, startTime, logSuccess, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, filename;

    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          _handle = function _handle() {
            _handle = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(filenameOrDir) {
              var stat, dirname, count, files, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, filename, written;

              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    if (_fs().existsSync(filenameOrDir)) {
                      context$4$0.next = 2;
                      break;
                    }

                    return context$4$0.abrupt("return", 0);

                  case 2:
                    stat = _fs().statSync(filenameOrDir);

                    if (!stat.isDirectory()) {
                      context$4$0.next = 38;
                      break;
                    }

                    dirname = filenameOrDir;
                    count = 0;
                    files = util.readdir(dirname, cliOptions.includeDotfiles);
                    _iteratorNormalCompletion = true;
                    _didIteratorError = false;
                    _iteratorError = undefined;
                    context$4$0.prev = 10;
                    _iterator = files[Symbol.iterator]();

                  case 12:
                    if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                      context$4$0.next = 21;
                      break;
                    }

                    filename = _step.value;
                    context$4$0.next = 16;
                    return handleFile(filename, dirname);

                  case 16:
                    written = context$4$0.sent;

                    if (written) count += 1;

                  case 18:
                    _iteratorNormalCompletion = true;
                    context$4$0.next = 12;
                    break;

                  case 21:
                    context$4$0.next = 27;
                    break;

                  case 23:
                    context$4$0.prev = 23;
                    context$4$0.t0 = context$4$0["catch"](10);
                    _didIteratorError = true;
                    _iteratorError = context$4$0.t0;

                  case 27:
                    context$4$0.prev = 27;
                    context$4$0.prev = 28;

                    if (!_iteratorNormalCompletion && _iterator["return"]) {
                      _iterator["return"]();
                    }

                  case 30:
                    context$4$0.prev = 30;

                    if (!_didIteratorError) {
                      context$4$0.next = 33;
                      break;
                    }

                    throw _iteratorError;

                  case 33:
                    return context$4$0.finish(30);

                  case 34:
                    return context$4$0.finish(27);

                  case 35:
                    return context$4$0.abrupt("return", count);

                  case 38:
                    filename = filenameOrDir;
                    context$4$0.next = 41;
                    return handleFile(filename, _path().dirname(filename));

                  case 41:
                    written = context$4$0.sent;
                    return context$4$0.abrupt("return", written ? 1 : 0);

                  case 43:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this, [[10, 23, 27, 35], [28,, 30, 34]]);
            }));
            return _handle.apply(this, arguments);
          };

          handle = function handle(_x6) {
            return _handle.apply(this, arguments);
          };

          _handleFile = function _handleFile() {
            _handleFile = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(src, base) {
              var written, filename, dest;
              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    context$4$0.next = 2;
                    return write(src, base);

                  case 2:
                    written = context$4$0.sent;

                    if (cliOptions.copyFiles && written === FILE_TYPE.NON_COMPILABLE || cliOptions.copyIgnored && written === FILE_TYPE.IGNORED) {
                      filename = _path().relative(base, src);
                      dest = getDest(filename, base);

                      outputFileSync(dest, _fs().readFileSync(src));
                      util.chmod(src, dest);
                    }
                    return context$4$0.abrupt("return", written === FILE_TYPE.COMPILED);

                  case 5:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this);
            }));
            return _handleFile.apply(this, arguments);
          };

          handleFile = function handleFile(_x4, _x5) {
            return _handleFile.apply(this, arguments);
          };

          getDest = function getDest(filename, base) {
            if (cliOptions.relative) {
              return _path().join(base, cliOptions.outDir, filename);
            }
            return _path().join(cliOptions.outDir, filename);
          };

          _write = function _write() {
            _write = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(src, base) {
              var relative, dest, res, outputMap, mapLoc;
              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    relative = _path().relative(base, src);

                    if (util.isCompilableExtension(relative, cliOptions.extensions)) {
                      context$4$0.next = 3;
                      break;
                    }

                    return context$4$0.abrupt("return", FILE_TYPE.NON_COMPILABLE);

                  case 3:
                    relative = util.withExtension(relative, cliOptions.keepFileExtension ? _path().extname(relative) : cliOptions.outFileExtension);
                    dest = getDest(relative, base);
                    context$4$0.prev = 5;
                    context$4$0.next = 8;
                    return util.compile(src, Object.assign({}, babelOptions, {
                      sourceFileName: _slash()(_path().relative(dest + "/..", src))
                    }));

                  case 8:
                    res = context$4$0.sent;

                    if (res) {
                      context$4$0.next = 11;
                      break;
                    }

                    return context$4$0.abrupt("return", FILE_TYPE.IGNORED);

                  case 11:
                    if (res.map) {
                      outputMap = false;

                      if (babelOptions.sourceMaps && babelOptions.sourceMaps !== "inline") {
                        outputMap = "external";
                      } else if (babelOptions.sourceMaps == null) {
                        outputMap = util.hasDataSourcemap(res.code) ? "external" : "both";
                      }
                      if (outputMap) {
                        mapLoc = dest + ".map";

                        if (outputMap === "external") {
                          res.code = util.addSourceMappingUrl(res.code, mapLoc);
                        }
                        res.map.file = _path().basename(relative);
                        outputFileSync(mapLoc, JSON.stringify(res.map));
                      }
                    }
                    outputFileSync(dest, res.code);
                    util.chmod(src, dest);
                    if (cliOptions.verbose) {
                      console.log(_path().relative(process.cwd(), src) + " -> " + dest);
                    }
                    return context$4$0.abrupt("return", FILE_TYPE.COMPILED);

                  case 18:
                    context$4$0.prev = 18;
                    context$4$0.t0 = context$4$0["catch"](5);

                    if (!cliOptions.watch) {
                      context$4$0.next = 23;
                      break;
                    }

                    console.error(context$4$0.t0);
                    return context$4$0.abrupt("return", FILE_TYPE.ERR_COMPILATION);

                  case 23:
                    throw context$4$0.t0;

                  case 24:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this, [[5, 18]]);
            }));
            return _write.apply(this, arguments);
          };

          write = function write(_x2, _x3) {
            return _write.apply(this, arguments);
          };

          compiledFiles = 0;
          startTime = null;
          logSuccess = util.debounce(function () {
            if (startTime === null) {
              return;
            }
            var diff = process.hrtime(startTime);
            console.log("Successfully compiled " + compiledFiles + " " + (compiledFiles !== 1 ? "files" : "file") + " with Babel (" + (diff[0] * 1e3 + Math.round(diff[1] / 1e6)) + "ms).");
            compiledFiles = 0;
            startTime = null;
          }, 100);

          if (cliOptions.watch) watcher.enable({
            enableGlobbing: true
          });

          if (cliOptions.skipInitialBuild) {
            context$2$0.next = 43;
            break;
          }

          if (cliOptions.deleteDirOnStart) {
            util.deleteDir(cliOptions.outDir);
          }
          ((function (v, w) {
            return v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1];
          })(process.versions.node, "10.12") ? _fs().mkdirSync : require("make-dir").sync)(cliOptions.outDir, {
            recursive: true
          });
          startTime = process.hrtime();
          _iteratorNormalCompletion2 = true;
          _didIteratorError2 = false;
          _iteratorError2 = undefined;
          context$2$0.prev = 18;
          _iterator2 = cliOptions.filenames[Symbol.iterator]();

        case 20:
          if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
            context$2$0.next = 28;
            break;
          }

          filename = _step2.value;
          context$2$0.next = 24;
          return handle(filename);

        case 24:
          compiledFiles += context$2$0.sent;

        case 25:
          _iteratorNormalCompletion2 = true;
          context$2$0.next = 20;
          break;

        case 28:
          context$2$0.next = 34;
          break;

        case 30:
          context$2$0.prev = 30;
          context$2$0.t0 = context$2$0["catch"](18);
          _didIteratorError2 = true;
          _iteratorError2 = context$2$0.t0;

        case 34:
          context$2$0.prev = 34;
          context$2$0.prev = 35;

          if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
            _iterator2["return"]();
          }

        case 37:
          context$2$0.prev = 37;

          if (!_didIteratorError2) {
            context$2$0.next = 40;
            break;
          }

          throw _iteratorError2;

        case 40:
          return context$2$0.finish(37);

        case 41:
          return context$2$0.finish(34);

        case 42:
          if (!cliOptions.quiet) {
            logSuccess();
            logSuccess.flush();
          }

        case 43:
          if (cliOptions.watch) {
            (function () {
              var processing = 0;
              var filenames = cliOptions.filenames;

              var getBase = undefined;
              if (filenames.length === 1) {
                (function () {
                  var base = filenames[0];
                  var absoluteBase = _path().resolve(base);
                  getBase = function (filename) {
                    return filename === absoluteBase ? _path().dirname(base) : base;
                  };
                })();
              } else {
                (function () {
                  var filenameToBaseMap = new Map(filenames.map(function (filename) {
                    var absoluteFilename = _path().resolve(filename);
                    return [absoluteFilename, _path().dirname(filename)];
                  }));
                  var absoluteFilenames = new Map(filenames.map(function (filename) {
                    var absoluteFilename = _path().resolve(filename);
                    return [absoluteFilename, filename];
                  }));

                  var _path2 = _path();

                  var sep = _path2.sep;

                  getBase = function (filename) {
                    var base = filenameToBaseMap.get(filename);
                    if (base !== undefined) {
                      return base;
                    }
                    var _iteratorNormalCompletion3 = true;
                    var _didIteratorError3 = false;
                    var _iteratorError3 = undefined;

                    try {
                      for (var _iterator3 = absoluteFilenames[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
                        var _step3$value = _slicedToArray(_step3.value, 2);

                        var absoluteFilenameOrDir = _step3$value[0];
                        var relative = _step3$value[1];

                        if (filename.startsWith(absoluteFilenameOrDir + sep)) {
                          filenameToBaseMap.set(filename, relative);
                          return relative;
                        }
                      }
                    } catch (err) {
                      _didIteratorError3 = true;
                      _iteratorError3 = err;
                    } finally {
                      try {
                        if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
                          _iterator3["return"]();
                        }
                      } finally {
                        if (_didIteratorError3) {
                          throw _iteratorError3;
                        }
                      }
                    }

                    return "";
                  };
                })();
              }
              filenames.forEach(function (filenameOrDir) {
                watcher.watch(filenameOrDir);
              });
              watcher.startWatcher();
              watcher.onFilesChange(_asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(filenames) {
                var written;
                return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                  while (1) switch (context$4$0.prev = context$4$0.next) {
                    case 0:
                      processing++;
                      if (startTime === null) startTime = process.hrtime();
                      context$4$0.prev = 2;
                      context$4$0.next = 5;
                      return Promise.all(filenames.map(function (filename) {
                        return handleFile(filename, getBase(filename));
                      }));

                    case 5:
                      written = context$4$0.sent;

                      compiledFiles += written.filter(Boolean).length;
                      context$4$0.next = 12;
                      break;

                    case 9:
                      context$4$0.prev = 9;
                      context$4$0.t0 = context$4$0["catch"](2);

                      console.error(context$4$0.t0);

                    case 12:
                      processing--;
                      if (processing === 0 && !cliOptions.quiet) logSuccess();

                    case 14:
                    case "end":
                      return context$4$0.stop();
                  }
                }, callee$3$0, this, [[2, 9]]);
              })));
            })();
          }

        case 44:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this, [[18, 30, 34, 42], [35,, 37, 41]]);
  }));
  return _ref.apply(this, arguments);
}
