// @babel/cli/lib/babel/file.js
"use strict";

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
function _convertSourceMap() {
  var data = require("convert-source-map");
  _convertSourceMap = function () {
    return data;
  };
  return data;
}
function _traceMapping() {
  var data = require("@jridgewell/trace-mapping");
  _traceMapping = function () {
    return data;
  };
  return data;
}
function _slash() {
  var data = require("slash");
  _slash = function () {
    return data;
  };
  return data;
}
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _fs() {
  var data = require("fs");
  _fs = function () {
    return data;
  };
  return data;
}
var util = require("./util.js");
var watcher = require("./watcher.js");
function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
        u = i.value;
  } catch (n) {
    return void e(n);
  }i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
        e = arguments;return new Promise(function (r, o) {
      var a = n.apply(t, e);function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }_next(void 0);
    });
  };
}
function _default(_x) {
  return _ref.apply(this, arguments);
}
function _ref() {
  _ref = _asyncToGenerator(regeneratorRuntime.mark(function callee$1$0(_ref2) {
    var cliOptions = _ref2.cliOptions;
    var babelOptions = _ref2.babelOptions;

    var buildResult, countNewlines, output, readStdin, stdin, _stdin, walk, _walk, files, _files;

    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          _files = function _files() {
            _files = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(filenames) {
              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    if (cliOptions.watch) {
                      watcher.enable({
                        enableGlobbing: false
                      });
                    }

                    if (cliOptions.skipInitialBuild) {
                      context$4$0.next = 4;
                      break;
                    }

                    context$4$0.next = 4;
                    return walk(filenames);

                  case 4:
                    if (cliOptions.watch) {
                      filenames.forEach(watcher.watch);
                      watcher.startWatcher();
                      watcher.onFilesChange(function (changes, event, cause) {
                        var actionableChange = changes.some(function (filename) {
                          return util.isCompilableExtension(filename, cliOptions.extensions) || filenames.includes(filename);
                        });
                        if (!actionableChange) return;
                        if (cliOptions.verbose) {
                          console.log(event + " " + cause);
                        }
                        walk(filenames)["catch"](function (err) {
                          console.error(err);
                        });
                      });
                    }

                  case 5:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this);
            }));
            return _files.apply(this, arguments);
          };

          files = function files(_x3) {
            return _files.apply(this, arguments);
          };

          _walk = function _walk() {
            _walk = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0(filenames) {
              var _filenames, results;

              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    _filenames = [];

                    filenames.forEach(function (filename) {
                      if (!_fs().existsSync(filename)) return;
                      var stat = _fs().statSync(filename);
                      if (stat.isDirectory()) {
                        _filenames.push.apply(_filenames, _toConsumableArray(util.readdirForCompilable(filename, cliOptions.includeDotfiles, cliOptions.extensions)));
                      } else {
                        _filenames.push(filename);
                      }
                    });
                    context$4$0.next = 4;
                    return Promise.all(_filenames.map(_asyncToGenerator(regeneratorRuntime.mark(function callee$4$0(filename) {
                      var sourceFilename;
                      return regeneratorRuntime.wrap(function callee$4$0$(context$5$0) {
                        while (1) switch (context$5$0.prev = context$5$0.next) {
                          case 0:
                            sourceFilename = filename;

                            if (cliOptions.outFile) {
                              sourceFilename = _path().relative(_path().dirname(cliOptions.outFile), sourceFilename);
                            }
                            sourceFilename = _slash()(sourceFilename);
                            context$5$0.prev = 3;
                            context$5$0.next = 6;
                            return util.compile(filename, Object.assign({}, babelOptions, {
                              sourceFileName: sourceFilename,
                              sourceMaps: babelOptions.sourceMaps === "inline" ? true : babelOptions.sourceMaps
                            }));

                          case 6:
                            return context$5$0.abrupt("return", context$5$0.sent);

                          case 9:
                            context$5$0.prev = 9;
                            context$5$0.t0 = context$5$0["catch"](3);

                            if (cliOptions.watch) {
                              context$5$0.next = 13;
                              break;
                            }

                            throw context$5$0.t0;

                          case 13:
                            console.error(context$5$0.t0);
                            return context$5$0.abrupt("return", null);

                          case 15:
                          case "end":
                            return context$5$0.stop();
                        }
                      }, callee$4$0, this, [[3, 9]]);
                    }))));

                  case 4:
                    results = context$4$0.sent;

                    output(results);

                  case 6:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this);
            }));
            return _walk.apply(this, arguments);
          };

          walk = function walk(_x2) {
            return _walk.apply(this, arguments);
          };

          _stdin = function _stdin() {
            _stdin = _asyncToGenerator(regeneratorRuntime.mark(function callee$3$0() {
              var code, res;
              return regeneratorRuntime.wrap(function callee$3$0$(context$4$0) {
                while (1) switch (context$4$0.prev = context$4$0.next) {
                  case 0:
                    context$4$0.next = 2;
                    return readStdin();

                  case 2:
                    code = context$4$0.sent;
                    context$4$0.next = 5;
                    return util.transformRepl(cliOptions.filename, code, Object.assign({}, babelOptions, {
                      sourceFileName: "stdin"
                    }));

                  case 5:
                    res = context$4$0.sent;

                    output([res]);

                  case 7:
                  case "end":
                    return context$4$0.stop();
                }
              }, callee$3$0, this);
            }));
            return _stdin.apply(this, arguments);
          };

          stdin = function stdin() {
            return _stdin.apply(this, arguments);
          };

          readStdin = function readStdin() {
            return new Promise(function (resolve, reject) {
              var code = "";
              process.stdin.setEncoding("utf8");
              process.stdin.on("readable", function () {
                var chunk = process.stdin.read();
                if (chunk !== null) code += chunk;
              });
              process.stdin.on("end", function () {
                resolve(code);
              });
              process.stdin.on("error", reject);
            });
          };

          output = function output(fileResults) {
            var result = buildResult(fileResults);
            if (cliOptions.outFile) {
              ((function (v, w) {
                return v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1];
              })(process.versions.node, "10.12") ? _fs().mkdirSync : require("make-dir").sync)(_path().dirname(cliOptions.outFile), {
                recursive: true
              });
              var outputMap = false;
              if (babelOptions.sourceMaps && babelOptions.sourceMaps !== "inline") {
                outputMap = "external";
              } else if (babelOptions.sourceMaps == null && result.hasRawMap) {
                outputMap = util.hasDataSourcemap(result.code) ? "external" : "both";
              }
              if (outputMap) {
                var mapLoc = cliOptions.outFile + ".map";
                if (outputMap === "external") {
                  result.code = util.addSourceMappingUrl(result.code, mapLoc);
                }
                _fs().writeFileSync(mapLoc, JSON.stringify((0, _traceMapping().encodedMap)(result.map)));
              }
              _fs().writeFileSync(cliOptions.outFile, result.code);
            } else {
              process.stdout.write(result.code + "\n");
            }
          };

          countNewlines = function countNewlines(code) {
            var count = 0;
            var index = -1;
            while ((index = code.indexOf("\n", index + 1)) !== -1) {
              count++;
            }
            return count;
          };

          buildResult = function buildResult(fileResults) {
            var mapSections = [];
            var code = "";
            var offset = 0;
            var hasRawMap = false;
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
              for (var _iterator = fileResults[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                var result = _step.value;

                if (!result) continue;
                hasRawMap = !!result.map;
                mapSections.push({
                  offset: {
                    line: offset,
                    column: 0
                  },
                  map: result.map || {
                    version: 3,
                    names: [],
                    sources: [],
                    mappings: []
                  }
                });
                code += result.code + "\n";
                offset += countNewlines(result.code) + 1;
              }
            } catch (err) {
              _didIteratorError = true;
              _iteratorError = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion && _iterator["return"]) {
                  _iterator["return"]();
                }
              } finally {
                if (_didIteratorError) {
                  throw _iteratorError;
                }
              }
            }

            var map = new (_traceMapping().AnyMap)({
              version: 3,
              file: cliOptions.sourceMapTarget || _path().basename(cliOptions.outFile || "") || "stdout",
              sections: mapSections
            });
            map.sourceRoot = babelOptions.sourceRoot;
            if (babelOptions.sourceMaps === "inline" || !cliOptions.outFile && babelOptions.sourceMaps) {
              code += "\n" + _convertSourceMap().fromObject((0, _traceMapping().encodedMap)(map)).toComment();
            }
            return {
              map: map,
              code: code,
              hasRawMap: hasRawMap
            };
          };

          if (!cliOptions.filenames.length) {
            context$2$0.next = 15;
            break;
          }

          context$2$0.next = 13;
          return files(cliOptions.filenames);

        case 13:
          context$2$0.next = 17;
          break;

        case 15:
          context$2$0.next = 17;
          return stdin();

        case 17:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this);
  }));
  return _ref.apply(this, arguments);
}
