// @babel/cli/lib/babel/watcher.js
"use strict";

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.enable = enable;
exports.onFilesChange = onFilesChange;
exports.startWatcher = startWatcher;
exports.updateExternalDependencies = updateExternalDependencies;
exports.watch = watch;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
var fileToDeps = new Map();
var depToFiles = new Map();
var isWatchMode = false;
var watcher = undefined;
var watchQueue = new Set();
var hasStarted = false;
function enable(_ref) {
  var enableGlobbing = _ref.enableGlobbing;

  isWatchMode = true;

  var _requireChokidar = requireChokidar();

  var FSWatcher = _requireChokidar.FSWatcher;

  var options = {
    disableGlobbing: !enableGlobbing,
    persistent: true,
    ignoreInitial: true,
    awaitWriteFinish: {
      stabilityThreshold: 50,
      pollInterval: 10
    }
  };
  watcher = new FSWatcher(options);
  watcher.on("unlink", unwatchFile);
}
function startWatcher() {
  hasStarted = true;
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = watchQueue[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var dep = _step.value;

      watcher.add(dep);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  watchQueue.clear();
  watcher.on("ready", function () {
    console.log("The watcher is ready.");
  });
}
function watch(filename) {
  if (!isWatchMode) {
    throw new Error("Internal Babel error: .watch called when not in watch mode.");
  }
  if (!hasStarted) {
    watchQueue.add(_path().resolve(filename));
  } else {
    watcher.add(_path().resolve(filename));
  }
}
function onFilesChange(callback) {
  if (!isWatchMode) {
    throw new Error("Internal Babel error: .onFilesChange called when not in watch mode.");
  }
  watcher.on("all", function (event, filename) {
    var _depToFiles$get;
    if (event !== "change" && event !== "add") return;
    var absoluteFile = _path().resolve(filename);
    callback([absoluteFile].concat(_toConsumableArray((_depToFiles$get = depToFiles.get(absoluteFile)) != null ? _depToFiles$get : [])), event, absoluteFile);
  });
}
function updateExternalDependencies(filename, dependencies) {
  if (!isWatchMode) return;
  var absFilename = _path().resolve(filename);
  var absDependencies = new Set(Array.from(dependencies, function (dep) {
    return _path().resolve(dep);
  }));
  var deps = fileToDeps.get(absFilename);
  if (deps) {
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = deps[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var dep = _step2.value;

        if (!absDependencies.has(dep)) {
          removeFileDependency(absFilename, dep);
        }
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }
  }
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = absDependencies[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var dep = _step3.value;

      var _deps = depToFiles.get(dep);
      if (!_deps) {
        depToFiles.set(dep, _deps = new Set());
        if (!hasStarted) {
          watchQueue.add(dep);
        } else {
          watcher.add(dep);
        }
      }
      _deps.add(absFilename);
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  fileToDeps.set(absFilename, absDependencies);
}
function removeFileDependency(filename, dep) {
  var deps = depToFiles.get(dep);
  deps["delete"](filename);
  if (deps.size === 0) {
    depToFiles["delete"](dep);
    if (!hasStarted) {
      watchQueue["delete"](dep);
    } else {
      watcher.unwatch(dep);
    }
  }
}
function unwatchFile(filename) {
  var deps = fileToDeps.get(filename);
  if (!deps) return;
  var _iteratorNormalCompletion4 = true;
  var _didIteratorError4 = false;
  var _iteratorError4 = undefined;

  try {
    for (var _iterator4 = deps[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
      var dep = _step4.value;

      removeFileDependency(filename, dep);
    }
  } catch (err) {
    _didIteratorError4 = true;
    _iteratorError4 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion4 && _iterator4["return"]) {
        _iterator4["return"]();
      }
    } finally {
      if (_didIteratorError4) {
        throw _iteratorError4;
      }
    }
  }

  fileToDeps["delete"](filename);
}
function requireChokidar() {
  try {
    return parseInt(process.versions.node) >= 8 ? require("chokidar") : require("@nicolo-ribaudo/chokidar-2");
  } catch (err) {
    console.error("The optional dependency chokidar failed to install and is required for " + "--watch. Chokidar is likely not supported on your platform.");
    throw err;
  }
}
