// @babel/generator/lib/token-map.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var marked0$0 = [childrenIterator].map(regeneratorRuntime.mark);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TokenMap = void 0;
var _t = require("@babel/types");
var traverseFast = _t.traverseFast;
var VISITOR_KEYS = _t.VISITOR_KEYS;

var TokenMap = (function () {
  function TokenMap(ast, tokens, source) {
    var _this = this;

    _classCallCheck(this, TokenMap);

    this._tokens = void 0;
    this._source = void 0;
    this._nodesToTokenIndexes = new Map();
    this._nodesOccurrencesCountCache = new Map();
    this._tokensCache = new Map();
    this._tokens = tokens;
    this._source = source;
    traverseFast(ast, function (node) {
      var indexes = _this._getTokensIndexesOfNode(node);
      if (indexes.length > 0) _this._nodesToTokenIndexes.set(node, indexes);
    });
    this._tokensCache = null;
  }

  _createClass(TokenMap, [{
    key: "has",
    value: function has(node) {
      return this._nodesToTokenIndexes.has(node);
    }
  }, {
    key: "getIndexes",
    value: function getIndexes(node) {
      return this._nodesToTokenIndexes.get(node);
    }
  }, {
    key: "find",
    value: function find(node, condition) {
      var indexes = this._nodesToTokenIndexes.get(node);
      if (indexes) {
        for (var k = 0; k < indexes.length; k++) {
          var index = indexes[k];
          var tok = this._tokens[index];
          if (condition(tok, index)) return tok;
        }
      }
      return null;
    }
  }, {
    key: "findLastIndex",
    value: function findLastIndex(node, condition) {
      var indexes = this._nodesToTokenIndexes.get(node);
      if (indexes) {
        for (var k = indexes.length - 1; k >= 0; k--) {
          var index = indexes[k];
          var tok = this._tokens[index];
          if (condition(tok, index)) return index;
        }
      }
      return -1;
    }
  }, {
    key: "findMatching",
    value: function findMatching(node, test) {
      var occurrenceCount = arguments.length <= 2 || arguments[2] === undefined ? 0 : arguments[2];

      var indexes = this._nodesToTokenIndexes.get(node);
      if (indexes) {
        var i = 0;
        var count = occurrenceCount;
        if (count > 1) {
          var cache = this._nodesOccurrencesCountCache.get(node);
          if (cache && cache.test === test && cache.count < count) {
            i = cache.i + 1;
            occurrenceCount -= cache.count + 1;
          }
        }
        for (; i < indexes.length; i++) {
          var tok = this._tokens[indexes[i]];
          if (this.matchesOriginal(tok, test)) {
            if (occurrenceCount === 0) {
              if (count > 0) {
                this._nodesOccurrencesCountCache.set(node, {
                  test: test,
                  count: count,
                  i: i
                });
              }
              return tok;
            }
            occurrenceCount--;
          }
        }
      }
      return null;
    }
  }, {
    key: "matchesOriginal",
    value: function matchesOriginal(token, test) {
      if (token.end - token.start !== test.length) return false;
      if (token.value != null) return token.value === test;
      return this._source.startsWith(test, token.start);
    }
  }, {
    key: "startMatches",
    value: function startMatches(node, test) {
      var indexes = this._nodesToTokenIndexes.get(node);
      if (!indexes) return false;
      var tok = this._tokens[indexes[0]];
      if (tok.start !== node.start) return false;
      return this.matchesOriginal(tok, test);
    }
  }, {
    key: "endMatches",
    value: function endMatches(node, test) {
      var indexes = this._nodesToTokenIndexes.get(node);
      if (!indexes) return false;
      var tok = this._tokens[indexes[indexes.length - 1]];
      if (tok.end !== node.end) return false;
      return this.matchesOriginal(tok, test);
    }
  }, {
    key: "_getTokensIndexesOfNode",
    value: function _getTokensIndexesOfNode(node) {
      if (node.start == null || node.end == null) return [];

      var _findTokensOfNode2 = this._findTokensOfNode(node, 0, this._tokens.length - 1);

      var first = _findTokensOfNode2.first;
      var last = _findTokensOfNode2.last;

      var low = first;
      var children = childrenIterator(node);
      if ((node.type === "ExportNamedDeclaration" || node.type === "ExportDefaultDeclaration") && node.declaration && node.declaration.type === "ClassDeclaration") {
        children.next();
      }
      var indexes = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = children[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var child = _step.value;

          if (child == null) continue;
          if (child.start == null || child.end == null) continue;
          var childTok = this._findTokensOfNode(child, low, last);
          var high = childTok.first;
          for (var k = low; k < high; k++) {
            indexes.push(k);
          }low = childTok.last + 1;
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"]) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      for (var k = low; k <= last; k++) {
        indexes.push(k);
      }return indexes;
    }
  }, {
    key: "_findTokensOfNode",
    value: function _findTokensOfNode(node, low, high) {
      var cached = this._tokensCache.get(node);
      if (cached) return cached;
      var first = this._findFirstTokenOfNode(node.start, low, high);
      var last = this._findLastTokenOfNode(node.end, first, high);
      this._tokensCache.set(node, {
        first: first,
        last: last
      });
      return {
        first: first,
        last: last
      };
    }
  }, {
    key: "_findFirstTokenOfNode",
    value: function _findFirstTokenOfNode(start, low, high) {
      while (low <= high) {
        var mid = high + low >> 1;
        if (start < this._tokens[mid].start) {
          high = mid - 1;
        } else if (start > this._tokens[mid].start) {
          low = mid + 1;
        } else {
          return mid;
        }
      }
      return low;
    }
  }, {
    key: "_findLastTokenOfNode",
    value: function _findLastTokenOfNode(end, low, high) {
      while (low <= high) {
        var mid = high + low >> 1;
        if (end < this._tokens[mid].end) {
          high = mid - 1;
        } else if (end > this._tokens[mid].end) {
          low = mid + 1;
        } else {
          return mid;
        }
      }
      return high;
    }
  }]);

  return TokenMap;
})();

exports.TokenMap = TokenMap;
function childrenIterator(node) {
  var i, keys, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, key, child;

  return regeneratorRuntime.wrap(function childrenIterator$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(node.type === "TemplateLiteral")) {
          context$1$0.next = 13;
          break;
        }

        context$1$0.next = 3;
        return node.quasis[0];

      case 3:
        i = 1;

      case 4:
        if (!(i < node.quasis.length)) {
          context$1$0.next = 12;
          break;
        }

        context$1$0.next = 7;
        return node.expressions[i - 1];

      case 7:
        context$1$0.next = 9;
        return node.quasis[i];

      case 9:
        i++;
        context$1$0.next = 4;
        break;

      case 12:
        return context$1$0.abrupt("return");

      case 13:
        keys = VISITOR_KEYS[node.type];
        _iteratorNormalCompletion2 = true;
        _didIteratorError2 = false;
        _iteratorError2 = undefined;
        context$1$0.prev = 17;
        _iterator2 = keys[Symbol.iterator]();

      case 19:
        if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
          context$1$0.next = 33;
          break;
        }

        key = _step2.value;
        child = node[key];

        if (child) {
          context$1$0.next = 24;
          break;
        }

        return context$1$0.abrupt("continue", 30);

      case 24:
        if (!Array.isArray(child)) {
          context$1$0.next = 28;
          break;
        }

        return context$1$0.delegateYield(child, "t0", 26);

      case 26:
        context$1$0.next = 30;
        break;

      case 28:
        context$1$0.next = 30;
        return child;

      case 30:
        _iteratorNormalCompletion2 = true;
        context$1$0.next = 19;
        break;

      case 33:
        context$1$0.next = 39;
        break;

      case 35:
        context$1$0.prev = 35;
        context$1$0.t1 = context$1$0["catch"](17);
        _didIteratorError2 = true;
        _iteratorError2 = context$1$0.t1;

      case 39:
        context$1$0.prev = 39;
        context$1$0.prev = 40;

        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }

      case 42:
        context$1$0.prev = 42;

        if (!_didIteratorError2) {
          context$1$0.next = 45;
          break;
        }

        throw _iteratorError2;

      case 45:
        return context$1$0.finish(42);

      case 46:
        return context$1$0.finish(39);

      case 47:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this, [[17, 35, 39, 47], [40,, 42, 46]]);
}
