// @babel/helper-create-regexp-features-plugin/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRegExpFeaturePlugin = createRegExpFeaturePlugin;
var _regexpuCore = require("regexpu-core");
var _core = require("@babel/core");
var _helperAnnotateAsPure = require("@babel/helper-annotate-as-pure");
var _semver = require("semver");
var _features = require("./features.js");
var _util = require("./util.js");
var versionKey = "@babel/plugin-regexp-features/version";
function createRegExpFeaturePlugin(_ref) {
  var name = _ref.name;
  var feature = _ref.feature;
  var _ref$options = _ref.options;
  var options = _ref$options === undefined ? {} : _ref$options;
  var _ref$manipulateOptions = _ref.manipulateOptions;
  var manipulateOptions = _ref$manipulateOptions === undefined ? function () {} : _ref$manipulateOptions;

  return {
    name: name,
    manipulateOptions: manipulateOptions,
    pre: function pre() {
      var _file$get;
      var file = this.file;

      var features = (_file$get = file.get(_features.featuresKey)) != null ? _file$get : 0;
      var newFeatures = (0, _features.enableFeature)(features, _features.FEATURES[feature]);
      var useUnicodeFlag = options.useUnicodeFlag;
      var runtime = options.runtime;

      if (useUnicodeFlag === false) {
        newFeatures = (0, _features.enableFeature)(newFeatures, _features.FEATURES.unicodeFlag);
      }
      if (newFeatures !== features) {
        file.set(_features.featuresKey, newFeatures);
      }
      if (runtime !== undefined) {
        if (file.has(_features.runtimeKey) && file.get(_features.runtimeKey) !== runtime && (0, _features.hasFeature)(newFeatures, _features.FEATURES.duplicateNamedCaptureGroups)) {
          throw new Error("The 'runtime' option must be the same for " + "'@babel/plugin-transform-named-capturing-groups-regex' and " + "'@babel/plugin-transform-duplicate-named-capturing-groups-regex'.");
        }
        if (feature === "namedCaptureGroups") {
          if (!runtime || !file.has(_features.runtimeKey)) file.set(_features.runtimeKey, runtime);
        } else {
          file.set(_features.runtimeKey, runtime);
        }
      }
      {
        if (typeof file.get(versionKey) === "number") {
          file.set(versionKey, "7.27.1");
          return;
        }
      }
      if (!file.get(versionKey) || _semver.lt(file.get(versionKey), "7.27.1")) {
        file.set(versionKey, "7.27.1");
      }
    },
    visitor: {
      RegExpLiteral: function RegExpLiteral(path) {
        var _file$get2;
        var node = path.node;
        var file = this.file;

        var features = file.get(_features.featuresKey);
        var runtime = (_file$get2 = file.get(_features.runtimeKey)) != null ? _file$get2 : true;
        var regexpuOptions = (0, _util.generateRegexpuOptions)(node.pattern, features);
        if ((0, _util.canSkipRegexpu)(node, regexpuOptions)) {
          return;
        }
        var namedCaptureGroups = {
          __proto__: null
        };
        if (regexpuOptions.namedGroups === "transform") {
          regexpuOptions.onNamedGroup = function (name, index) {
            var prev = namedCaptureGroups[name];
            if (typeof prev === "number") {
              namedCaptureGroups[name] = [prev, index];
            } else if (Array.isArray(prev)) {
              prev.push(index);
            } else {
              namedCaptureGroups[name] = index;
            }
          };
        }
        var newFlags = undefined;
        if (regexpuOptions.modifiers === "transform") {
          regexpuOptions.onNewFlags = function (flags) {
            newFlags = flags;
          };
        }
        node.pattern = _regexpuCore(node.pattern, node.flags, regexpuOptions);
        if (regexpuOptions.namedGroups === "transform" && Object.keys(namedCaptureGroups).length > 0 && runtime && !isRegExpTest(path)) {
          var call = _core.types.callExpression(this.addHelper("wrapRegExp"), [node, _core.types.valueToNode(namedCaptureGroups)]);
          (0, _helperAnnotateAsPure["default"])(call);
          path.replaceWith(call);
        }
        node.flags = (0, _util.transformFlags)(regexpuOptions, newFlags != null ? newFlags : node.flags);
      }
    }
  };
}
function isRegExpTest(path) {
  return path.parentPath.isMemberExpression({
    object: path.node,
    computed: false
  }) && path.parentPath.get("property").isIdentifier({
    name: "test"
  });
}
