// @babel/core/lib/gensync-utils/async.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.forwardAsync = forwardAsync;
exports.isAsync = void 0;
exports.isThenable = isThenable;
exports.maybeAsync = maybeAsync;
exports.waitFor = exports.onFirstPause = void 0;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
        u = i.value;
  } catch (n) {
    return void e(n);
  }i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
        e = arguments;return new Promise(function (r, o) {
      var a = n.apply(t, e);function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }_next(void 0);
    });
  };
}
var runGenerator = _gensync()(regeneratorRuntime.mark(function callee$0$0(item) {
  return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(item, "t0", 1);

      case 1:
        return context$1$0.abrupt("return", context$1$0.t0);

      case 2:
      case "end":
        return context$1$0.stop();
    }
  }, callee$0$0, this);
}));
var isAsync = exports.isAsync = _gensync()({
  sync: function sync() {
    return false;
  },
  errback: function errback(cb) {
    return cb(null, true);
  }
});
function maybeAsync(fn, message) {
  return _gensync()({
    sync: function sync() {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      var result = fn.apply(this, args);
      if (isThenable(result)) throw new Error(message);
      return result;
    },
    async: function async() {
      for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return Promise.resolve(fn.apply(this, args));
    }
  });
}
var withKind = _gensync()({
  sync: function sync(cb) {
    return cb("sync");
  },
  async: (function () {
    var _ref = _asyncToGenerator(regeneratorRuntime.mark(function callee$1$0(cb) {
      return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt("return", cb("async"));

          case 1:
          case "end":
            return context$2$0.stop();
        }
      }, callee$1$0, this);
    }));
    return function async(_x) {
      return _ref.apply(this, arguments);
    };
  })()
});
function forwardAsync(action, cb) {
  var g = _gensync()(action);
  return withKind(function (kind) {
    var adapted = g[kind];
    return cb(adapted);
  });
}
var onFirstPause = exports.onFirstPause = _gensync()({
  name: "onFirstPause",
  arity: 2,
  sync: function sync(item) {
    return runGenerator.sync(item);
  },
  errback: function errback(item, firstPause, cb) {
    var completed = false;
    runGenerator.errback(item, function (err, value) {
      completed = true;
      cb(err, value);
    });
    if (!completed) {
      firstPause();
    }
  }
});
var waitFor = exports.waitFor = _gensync()({
  sync: function sync(x) {
    return x;
  },
  async: (function () {
    var _ref2 = _asyncToGenerator(regeneratorRuntime.mark(function callee$1$0(x) {
      return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt("return", x);

          case 1:
          case "end":
            return context$2$0.stop();
        }
      }, callee$1$0, this);
    }));
    return function async(_x2) {
      return _ref2.apply(this, arguments);
    };
  })()
});
function isThenable(val) {
  return !!val && (typeof val === "object" || typeof val === "function") && !!val.then && typeof val.then === "function";
}
0 && 0;
