// @babel/helper-create-class-features-plugin/lib/decorators-2018-09.js
"use strict";

var _templateObject = _taggedTemplateLiteral(["return ", ""], ["return ", ""]),
    _templateObject2 = _taggedTemplateLiteral(["\n    ", "(\n      ", ",\n      function (", ", ", ") {\n        ", "\n        return { F: ", ", d: ", " };\n      },\n      ", "\n    )\n  "], ["\n    ", "(\n      ", ",\n      function (", ", ", ") {\n        ", "\n        return { F: ", ", d: ", " };\n      },\n      ", "\n    )\n  "]),
    _templateObject3 = _taggedTemplateLiteral(["let ", " = ", ""], ["let ", " = ", ""]),
    _templateObject4 = _taggedTemplateLiteral(["\n        ", "(this)\n      "], ["\n        ", "(this)\n      "]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildDecoratedClass = buildDecoratedClass;
var _core = require("@babel/core");
var _helperReplaceSupers = require("@babel/helper-replace-supers");
;
function prop(key, value) {
  if (!value) return null;
  return _core.types.objectProperty(_core.types.identifier(key), value);
}
function method(key, body) {
  return _core.types.objectMethod("method", _core.types.identifier(key), [], _core.types.blockStatement(body));
}
function takeDecorators(node) {
  var result = undefined;
  if (node.decorators && node.decorators.length > 0) {
    result = _core.types.arrayExpression(node.decorators.map(function (decorator) {
      return decorator.expression;
    }));
  }
  node.decorators = undefined;
  return result;
}
function getKey(node) {
  if (node.computed) {
    return node.key;
  } else if (_core.types.isIdentifier(node.key)) {
    return _core.types.stringLiteral(node.key.name);
  } else {
    return _core.types.stringLiteral(String(node.key.value));
  }
}
function extractElementDescriptor(file, classRef, superRef, path) {
  var isMethod = path.isClassMethod();
  if (path.isPrivate()) {
    throw path.buildCodeFrameError("Private " + (isMethod ? "methods" : "fields") + " in decorated classes are not supported yet.");
  }
  if (path.node.type === "ClassAccessorProperty") {
    throw path.buildCodeFrameError("Accessor properties are not supported in 2018-09 decorator transform, please specify { \"version\": \"2021-12\" } instead.");
  }
  if (path.node.type === "StaticBlock") {
    throw path.buildCodeFrameError("Static blocks are not supported in 2018-09 decorator transform, please specify { \"version\": \"2021-12\" } instead.");
  }
  var node = path.node;
  var scope = path.scope;

  if (!path.isTSDeclareMethod()) {
    new _helperReplaceSupers["default"]({
      methodPath: path,
      objectRef: classRef,
      superRef: superRef,
      file: file,
      refToPreserve: classRef
    }).replace();
  }
  var properties = [prop("kind", _core.types.stringLiteral(_core.types.isClassMethod(node) ? node.kind : "field")), prop("decorators", takeDecorators(node)), prop("static", node["static"] && _core.types.booleanLiteral(true)), prop("key", getKey(node))].filter(Boolean);
  if (isMethod) {
    {
      var _path$ensureFunctionN;
      (_path$ensureFunctionN = path.ensureFunctionName) != null ? _path$ensureFunctionN : path.ensureFunctionName = require("@babel/traverse").NodePath.prototype.ensureFunctionName;
    }
    path.ensureFunctionName(false);
    properties.push(prop("value", _core.types.toExpression(path.node)));
  } else if (_core.types.isClassProperty(node) && node.value) {
    properties.push(method("value", _core.template.statements.ast(_templateObject, node.value)));
  } else {
    properties.push(prop("value", scope.buildUndefinedNode()));
  }
  path.remove();
  return _core.types.objectExpression(properties);
}
function addDecorateHelper(file) {
  return file.addHelper("decorate");
}
function buildDecoratedClass(ref, path, elements, file) {
  var node = path.node;
  var scope = path.scope;

  var initializeId = scope.generateUidIdentifier("initialize");
  var isDeclaration = node.id && path.isDeclaration();
  var isStrict = path.isInStrictMode();
  var superClass = node.superClass;

  node.type = "ClassDeclaration";
  if (!node.id) node.id = _core.types.cloneNode(ref);
  var superId = undefined;
  if (superClass) {
    superId = scope.generateUidIdentifierBasedOnNode(node.superClass, "super");
    node.superClass = superId;
  }
  var classDecorators = takeDecorators(node);
  var definitions = _core.types.arrayExpression(elements.filter(function (element) {
    return !element.node.abstract && element.node.type !== "TSIndexSignature";
  }).map(function (path) {
    return extractElementDescriptor(file, node.id, superId, path);
  }));
  var wrapperCall = _core.template.expression.ast(_templateObject2, addDecorateHelper(file), classDecorators || _core.types.nullLiteral(), initializeId, superClass ? _core.types.cloneNode(superId) : null, node, _core.types.cloneNode(node.id), definitions, superClass);
  if (!isStrict) {
    wrapperCall.arguments[1].body.directives.push(_core.types.directive(_core.types.directiveLiteral("use strict")));
  }
  var replacement = wrapperCall;
  var classPathDesc = "arguments.1.body.body.0";
  if (isDeclaration) {
    replacement = _core.template.statement.ast(_templateObject3, ref, wrapperCall);
    classPathDesc = "declarations.0.init." + classPathDesc;
  }
  return {
    instanceNodes: [_core.template.statement.ast(_templateObject4, _core.types.cloneNode(initializeId))],
    wrapClass: function wrapClass(path) {
      path.replaceWith(replacement);
      return path.get(classPathDesc);
    }
  };
}
