// @babel/helper-create-class-features-plugin/lib/decorators.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var _templateObject = _taggedTemplateLiteral(["\n        function () {\n          return this.", ";\n        }\n      "], ["\n        function () {\n          return this.", ";\n        }\n      "]),
    _templateObject2 = _taggedTemplateLiteral(["\n        function (value) {\n          this.", " = value;\n        }\n      "], ["\n        function (value) {\n          this.", " = value;\n        }\n      "]),
    _templateObject3 = _taggedTemplateLiteral(["\n      o => o.", "\n    "], ["\n      o => o.", "\n    "]),
    _templateObject4 = _taggedTemplateLiteral(["\n      (o, v) => o.", " = v\n    "], ["\n      (o, v) => o.", " = v\n    "]),
    _templateObject5 = _taggedTemplateLiteral(["\n        class extends ", " {}\n      "], ["\n        class extends ", " {}\n      "]);

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;
exports.hasDecorators = hasDecorators;
exports.hasOwnDecorators = hasOwnDecorators;
var _core = require("@babel/core");
var _helperReplaceSupers = require("@babel/helper-replace-supers");
var _helperSkipTransparentExpressionWrappers = require("@babel/helper-skip-transparent-expression-wrappers");
var _fields = require("./fields.js");
var _misc = require("./misc.js");
function hasOwnDecorators(node) {
  var _node$decorators;
  return !!((_node$decorators = node.decorators) != null && _node$decorators.length);
}
function hasDecorators(node) {
  return hasOwnDecorators(node) || node.body.body.some(hasOwnDecorators);
}
function incrementId(id) {
  var idx = arguments.length <= 1 || arguments[1] === undefined ? id.length - 1 : arguments[1];
  return (function () {
    if (idx === -1) {
      id.unshift(65);
      return;
    }
    var current = id[idx];
    if (current === 90) {
      id[idx] = 97;
    } else if (current === 122) {
      id[idx] = 65;
      incrementId(id, idx - 1);
    } else {
      id[idx] = current + 1;
    }
  })();
}
function createPrivateUidGeneratorForClass(classPath) {
  var currentPrivateId = [];
  var privateNames = new Set();
  classPath.traverse({
    PrivateName: function PrivateName(path) {
      privateNames.add(path.node.id.name);
    }
  });
  return function () {
    var reifiedId = undefined;
    do {
      incrementId(currentPrivateId);
      reifiedId = String.fromCharCode.apply(String, currentPrivateId);
    } while (privateNames.has(reifiedId));
    return _core.types.privateName(_core.types.identifier(reifiedId));
  };
}
function createLazyPrivateUidGeneratorForClass(classPath) {
  var generator = undefined;
  return function () {
    if (!generator) {
      generator = createPrivateUidGeneratorForClass(classPath);
    }
    return generator();
  };
}
function replaceClassWithVar(path, className) {
  var id = path.node.id;
  var scope = path.scope;
  if (path.type === "ClassDeclaration") {
    var _className = id.name;
    var varId = scope.generateUidIdentifierBasedOnNode(id);
    var classId = _core.types.identifier(_className);
    scope.rename(_className, varId.name);
    path.get("id").replaceWith(classId);
    return {
      id: _core.types.cloneNode(varId),
      path: path
    };
  } else {
    var varId = undefined;
    if (id) {
      className = id.name;
      varId = generateLetUidIdentifier(scope.parent, className);
      scope.rename(className, varId.name);
    } else {
      varId = generateLetUidIdentifier(scope.parent, typeof className === "string" ? className : "decorated_class");
    }
    var newClassExpr = _core.types.classExpression(typeof className === "string" ? _core.types.identifier(className) : null, path.node.superClass, path.node.body);

    var _path$replaceWith = path.replaceWith(_core.types.sequenceExpression([newClassExpr, varId]));

    var _path$replaceWith2 = _slicedToArray(_path$replaceWith, 1);

    var newPath = _path$replaceWith2[0];

    return {
      id: _core.types.cloneNode(varId),
      path: newPath.get("expressions.0")
    };
  }
}
function generateClassProperty(key, value, isStatic) {
  if (key.type === "PrivateName") {
    return _core.types.classPrivateProperty(key, value, undefined, isStatic);
  } else {
    return _core.types.classProperty(key, value, undefined, undefined, isStatic);
  }
}
function assignIdForAnonymousClass(path, className) {
  if (!path.node.id) {
    path.node.id = typeof className === "string" ? _core.types.identifier(className) : path.scope.generateUidIdentifier("Class");
  }
}
function addProxyAccessorsFor(className, element, getterKey, setterKey, targetKey, isComputed, isStatic, version) {
  var thisArg = (version === "2023-11" || version === "2023-05") && isStatic ? className : _core.types.thisExpression();
  var getterBody = _core.types.blockStatement([_core.types.returnStatement(_core.types.memberExpression(_core.types.cloneNode(thisArg), _core.types.cloneNode(targetKey)))]);
  var setterBody = _core.types.blockStatement([_core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(_core.types.cloneNode(thisArg), _core.types.cloneNode(targetKey)), _core.types.identifier("v")))]);
  var getter = undefined,
      setter = undefined;
  if (getterKey.type === "PrivateName") {
    getter = _core.types.classPrivateMethod("get", getterKey, [], getterBody, isStatic);
    setter = _core.types.classPrivateMethod("set", setterKey, [_core.types.identifier("v")], setterBody, isStatic);
  } else {
    getter = _core.types.classMethod("get", getterKey, [], getterBody, isComputed, isStatic);
    setter = _core.types.classMethod("set", setterKey, [_core.types.identifier("v")], setterBody, isComputed, isStatic);
  }
  element.insertAfter(setter);
  element.insertAfter(getter);
}
function extractProxyAccessorsFor(targetKey, version) {
  if (version !== "2023-11" && version !== "2023-05" && version !== "2023-01") {
    return [_core.template.expression.ast(_templateObject, _core.types.cloneNode(targetKey)), _core.template.expression.ast(_templateObject2, _core.types.cloneNode(targetKey))];
  }
  return [_core.template.expression.ast(_templateObject3, _core.types.cloneNode(targetKey)), _core.template.expression.ast(_templateObject4, _core.types.cloneNode(targetKey))];
}
function getComputedKeyLastElement(_x2) {
  var _again = true;

  _function: while (_again) {
    var path = _x2;
    _again = false;

    path = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path);
    if (path.isSequenceExpression()) {
      var expressions = path.get("expressions");
      _x2 = expressions[expressions.length - 1];
      _again = true;
      expressions = undefined;
      continue _function;
    }
    return path;
  }
}
function getComputedKeyMemoiser(path) {
  var element = getComputedKeyLastElement(path);
  if (element.isConstantExpression()) {
    return _core.types.cloneNode(path.node);
  } else if (element.isIdentifier() && path.scope.hasUid(element.node.name)) {
    return _core.types.cloneNode(path.node);
  } else if (element.isAssignmentExpression() && element.get("left").isIdentifier()) {
    return _core.types.cloneNode(element.node.left);
  } else {
    throw new Error("Internal Error: the computed key " + path.toString() + " has not yet been memoised.");
  }
}
function prependExpressionsToComputedKey(expressions, fieldPath) {
  var key = fieldPath.get("key");
  if (key.isSequenceExpression()) {
    expressions.push.apply(expressions, _toConsumableArray(key.node.expressions));
  } else {
    expressions.push(key.node);
  }
  key.replaceWith(maybeSequenceExpression(expressions));
}
function appendExpressionsToComputedKey(expressions, fieldPath) {
  var key = fieldPath.get("key");
  var completion = getComputedKeyLastElement(key);
  if (completion.isConstantExpression()) {
    prependExpressionsToComputedKey(expressions, fieldPath);
  } else {
    var scopeParent = key.scope.parent;
    var maybeAssignment = (0, _misc.memoiseComputedKey)(completion.node, scopeParent, scopeParent.generateUid("computedKey"));
    if (!maybeAssignment) {
      prependExpressionsToComputedKey(expressions, fieldPath);
    } else {
      var expressionSequence = [].concat(_toConsumableArray(expressions), [_core.types.cloneNode(maybeAssignment.left)]);
      var completionParent = completion.parentPath;
      if (completionParent.isSequenceExpression()) {
        completionParent.pushContainer("expressions", expressionSequence);
      } else {
        completion.replaceWith(maybeSequenceExpression([_core.types.cloneNode(maybeAssignment)].concat(_toConsumableArray(expressionSequence))));
      }
    }
  }
}
function prependExpressionsToFieldInitializer(expressions, fieldPath) {
  var initializer = fieldPath.get("value");
  if (initializer.node) {
    expressions.push(initializer.node);
  } else if (expressions.length > 0) {
    expressions[expressions.length - 1] = _core.types.unaryExpression("void", expressions[expressions.length - 1]);
  }
  initializer.replaceWith(maybeSequenceExpression(expressions));
}
function prependExpressionsToStaticBlock(expressions, blockPath) {
  blockPath.unshiftContainer("body", _core.types.expressionStatement(maybeSequenceExpression(expressions)));
}
function prependExpressionsToConstructor(expressions, constructorPath) {
  constructorPath.node.body.body.unshift(_core.types.expressionStatement(maybeSequenceExpression(expressions)));
}
function isProtoInitCallExpression(expression, protoInitCall) {
  return _core.types.isCallExpression(expression) && _core.types.isIdentifier(expression.callee, {
    name: protoInitCall.name
  });
}
function optimizeSuperCallAndExpressions(expressions, protoInitLocal) {
  if (protoInitLocal) {
    if (expressions.length >= 2 && isProtoInitCallExpression(expressions[1], protoInitLocal)) {
      var mergedSuperCall = _core.types.callExpression(_core.types.cloneNode(protoInitLocal), [expressions[0]]);
      expressions.splice(0, 2, mergedSuperCall);
    }
    if (expressions.length >= 2 && _core.types.isThisExpression(expressions[expressions.length - 1]) && isProtoInitCallExpression(expressions[expressions.length - 2], protoInitLocal)) {
      expressions.splice(expressions.length - 1, 1);
    }
  }
  return maybeSequenceExpression(expressions);
}
function insertExpressionsAfterSuperCallAndOptimize(expressions, constructorPath, protoInitLocal) {
  constructorPath.traverse({
    CallExpression: {
      exit: function exit(path) {
        if (!path.get("callee").isSuper()) return;
        var newNodes = [path.node].concat(_toConsumableArray(expressions.map(function (expr) {
          return _core.types.cloneNode(expr);
        })));
        if (path.isCompletionRecord()) {
          newNodes.push(_core.types.thisExpression());
        }
        path.replaceWith(optimizeSuperCallAndExpressions(newNodes, protoInitLocal));
        path.skip();
      }
    },
    ClassMethod: function ClassMethod(path) {
      if (path.node.kind === "constructor") {
        path.skip();
      }
    }
  });
}
function createConstructorFromExpressions(expressions, isDerivedClass) {
  var body = [_core.types.expressionStatement(maybeSequenceExpression(expressions))];
  if (isDerivedClass) {
    body.unshift(_core.types.expressionStatement(_core.types.callExpression(_core.types["super"](), [_core.types.spreadElement(_core.types.identifier("args"))])));
  }
  return _core.types.classMethod("constructor", _core.types.identifier("constructor"), isDerivedClass ? [_core.types.restElement(_core.types.identifier("args"))] : [], _core.types.blockStatement(body));
}
function createStaticBlockFromExpressions(expressions) {
  return _core.types.staticBlock([_core.types.expressionStatement(maybeSequenceExpression(expressions))]);
}
var FIELD = 0;
var ACCESSOR = 1;
var METHOD = 2;
var GETTER = 3;
var SETTER = 4;
var STATIC_OLD_VERSION = 5;
var STATIC = 8;
var DECORATORS_HAVE_THIS = 16;
function getElementKind(element) {
  switch (element.node.type) {
    case "ClassProperty":
    case "ClassPrivateProperty":
      return FIELD;
    case "ClassAccessorProperty":
      return ACCESSOR;
    case "ClassMethod":
    case "ClassPrivateMethod":
      if (element.node.kind === "get") {
        return GETTER;
      } else if (element.node.kind === "set") {
        return SETTER;
      } else {
        return METHOD;
      }
  }
}
function toSortedDecoratorInfo(info) {
  return [].concat(_toConsumableArray(info.filter(function (el) {
    return el.isStatic && el.kind >= ACCESSOR && el.kind <= SETTER;
  })), _toConsumableArray(info.filter(function (el) {
    return !el.isStatic && el.kind >= ACCESSOR && el.kind <= SETTER;
  })), _toConsumableArray(info.filter(function (el) {
    return el.isStatic && el.kind === FIELD;
  })), _toConsumableArray(info.filter(function (el) {
    return !el.isStatic && el.kind === FIELD;
  })));
}
function generateDecorationList(decorators, decoratorsThis, version) {
  var decsCount = decorators.length;
  var haveOneThis = decoratorsThis.some(Boolean);
  var decs = [];
  for (var i = 0; i < decsCount; i++) {
    if ((version === "2023-11" || version === "2023-05") && haveOneThis) {
      decs.push(decoratorsThis[i] || _core.types.unaryExpression("void", _core.types.numericLiteral(0)));
    }
    decs.push(decorators[i].expression);
  }
  return {
    haveThis: haveOneThis,
    decs: decs
  };
}
function generateDecorationExprs(decorationInfo, version) {
  return _core.types.arrayExpression(decorationInfo.map(function (el) {
    var flag = el.kind;
    if (el.isStatic) {
      flag += version === "2023-11" || version === "2023-05" ? STATIC : STATIC_OLD_VERSION;
    }
    if (el.decoratorsHaveThis) flag += DECORATORS_HAVE_THIS;
    return _core.types.arrayExpression([el.decoratorsArray, _core.types.numericLiteral(flag), el.name].concat(_toConsumableArray(el.privateMethods || [])));
  }));
}
function extractElementLocalAssignments(decorationInfo) {
  var localIds = [];
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = decorationInfo[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var el = _step.value;
      var locals = el.locals;

      if (Array.isArray(locals)) {
        localIds.push.apply(localIds, _toConsumableArray(locals));
      } else if (locals !== undefined) {
        localIds.push(locals);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return localIds;
}
function addCallAccessorsFor(version, element, key, getId, setId, isStatic) {
  element.insertAfter(_core.types.classPrivateMethod("get", _core.types.cloneNode(key), [], _core.types.blockStatement([_core.types.returnStatement(_core.types.callExpression(_core.types.cloneNode(getId), version === "2023-11" && isStatic ? [] : [_core.types.thisExpression()]))]), isStatic));
  element.insertAfter(_core.types.classPrivateMethod("set", _core.types.cloneNode(key), [_core.types.identifier("v")], _core.types.blockStatement([_core.types.expressionStatement(_core.types.callExpression(_core.types.cloneNode(setId), version === "2023-11" && isStatic ? [_core.types.identifier("v")] : [_core.types.thisExpression(), _core.types.identifier("v")]))]), isStatic));
}
function movePrivateAccessor(element, key, methodLocalVar, isStatic) {
  var params = undefined;
  var block = undefined;
  if (element.node.kind === "set") {
    params = [_core.types.identifier("v")];
    block = [_core.types.expressionStatement(_core.types.callExpression(methodLocalVar, [_core.types.thisExpression(), _core.types.identifier("v")]))];
  } else {
    params = [];
    block = [_core.types.returnStatement(_core.types.callExpression(methodLocalVar, [_core.types.thisExpression()]))];
  }
  element.replaceWith(_core.types.classPrivateMethod(element.node.kind, _core.types.cloneNode(key), params, _core.types.blockStatement(block), isStatic));
}
function isClassDecoratableElementPath(path) {
  var type = path.type;

  return type !== "TSDeclareMethod" && type !== "TSIndexSignature" && type !== "StaticBlock";
}
function staticBlockToIIFE(block) {
  return _core.types.callExpression(_core.types.arrowFunctionExpression([], _core.types.blockStatement(block.body)), []);
}
function staticBlockToFunctionClosure(block) {
  return _core.types.functionExpression(null, [], _core.types.blockStatement(block.body));
}
function fieldInitializerToClosure(value) {
  return _core.types.functionExpression(null, [], _core.types.blockStatement([_core.types.returnStatement(value)]));
}
function maybeSequenceExpression(exprs) {
  if (exprs.length === 0) return _core.types.unaryExpression("void", _core.types.numericLiteral(0));
  if (exprs.length === 1) return exprs[0];
  return _core.types.sequenceExpression(exprs);
}
function createFunctionExpressionFromPrivateMethod(node) {
  var params = node.params;
  var body = node.body;
  var isGenerator = node.generator;
  var isAsync = node.async;

  return _core.types.functionExpression(undefined, params, body, isGenerator, isAsync);
}
function createSetFunctionNameCall(state, className) {
  return _core.types.callExpression(state.addHelper("setFunctionName"), [_core.types.thisExpression(), className]);
}
function createToPropertyKeyCall(state, propertyKey) {
  return _core.types.callExpression(state.addHelper("toPropertyKey"), [propertyKey]);
}
function createPrivateBrandCheckClosure(brandName) {
  return _core.types.arrowFunctionExpression([_core.types.identifier("_")], _core.types.binaryExpression("in", _core.types.cloneNode(brandName), _core.types.identifier("_")));
}
function usesPrivateField(expression) {
  {
    try {
      _core.types.traverseFast(expression, function (node) {
        if (_core.types.isPrivateName(node)) {
          throw null;
        }
      });
      return false;
    } catch (_unused) {
      return true;
    }
  }
}
function convertToComputedKey(path) {
  var node = path.node;

  node.computed = true;
  if (_core.types.isIdentifier(node.key)) {
    node.key = _core.types.stringLiteral(node.key.name);
  }
}
function hasInstancePrivateAccess(path, privateNames) {
  var containsInstancePrivateAccess = false;
  if (privateNames.length > 0) {
    var privateNameVisitor = (0, _fields.privateNameVisitorFactory)({
      PrivateName: function PrivateName(path, state) {
        if (state.privateNamesMap.has(path.node.id.name)) {
          containsInstancePrivateAccess = true;
          path.stop();
        }
      }
    });
    var privateNamesMap = new Map();
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = privateNames[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var _name = _step2.value;

        privateNamesMap.set(_name, null);
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }

    path.traverse(privateNameVisitor, {
      privateNamesMap: privateNamesMap
    });
  }
  return containsInstancePrivateAccess;
}
function checkPrivateMethodUpdateError(path, decoratedPrivateMethods) {
  var privateNameVisitor = (0, _fields.privateNameVisitorFactory)({
    PrivateName: function PrivateName(path, state) {
      if (!state.privateNamesMap.has(path.node.id.name)) return;
      var parentPath = path.parentPath;
      var parentParentPath = parentPath.parentPath;
      if (parentParentPath.node.type === "AssignmentExpression" && parentParentPath.node.left === parentPath.node || parentParentPath.node.type === "UpdateExpression" || parentParentPath.node.type === "RestElement" || parentParentPath.node.type === "ArrayPattern" || parentParentPath.node.type === "ObjectProperty" && parentParentPath.node.value === parentPath.node && parentParentPath.parentPath.type === "ObjectPattern" || parentParentPath.node.type === "ForOfStatement" && parentParentPath.node.left === parentPath.node) {
        throw path.buildCodeFrameError("Decorated private methods are read-only, but \"#" + path.node.id.name + "\" is updated via this expression.");
      }
    }
  });
  var privateNamesMap = new Map();
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = decoratedPrivateMethods[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var _name2 = _step3.value;

      privateNamesMap.set(_name2, null);
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  path.traverse(privateNameVisitor, {
    privateNamesMap: privateNamesMap
  });
}
function transformClass(path, state, constantSuper, ignoreFunctionLength, className, propertyVisitor, version) {
  var _path$node$id;
  var body = path.get("body.body");
  var classDecorators = path.node.decorators;
  var hasElementDecorators = false;
  var hasComputedKeysSideEffects = false;
  var elemDecsUseFnContext = false;
  var generateClassPrivateUid = createLazyPrivateUidGeneratorForClass(path);
  var classAssignments = [];
  var scopeParent = path.scope.parent;
  var memoiseExpression = function memoiseExpression(expression, hint, assignments) {
    var localEvaluatedId = generateLetUidIdentifier(scopeParent, hint);
    assignments.push(_core.types.assignmentExpression("=", localEvaluatedId, expression));
    return _core.types.cloneNode(localEvaluatedId);
  };
  var protoInitLocal = undefined;
  var staticInitLocal = undefined;
  var classIdName = (_path$node$id = path.node.id) == null ? void 0 : _path$node$id.name;
  var setClassName = typeof className === "object" ? className : undefined;
  var usesFunctionContextOrYieldAwait = function usesFunctionContextOrYieldAwait(decorator) {
    {
      try {
        _core.types.traverseFast(decorator, function (node) {
          if (_core.types.isThisExpression(node) || _core.types.isSuper(node) || _core.types.isYieldExpression(node) || _core.types.isAwaitExpression(node) || _core.types.isIdentifier(node, {
            name: "arguments"
          }) || classIdName && _core.types.isIdentifier(node, {
            name: classIdName
          }) || _core.types.isMetaProperty(node) && node.meta.name !== "import") {
            throw null;
          }
        });
        return false;
      } catch (_unused2) {
        return true;
      }
    }
  };
  var instancePrivateNames = [];
  var _iteratorNormalCompletion4 = true;
  var _didIteratorError4 = false;
  var _iteratorError4 = undefined;

  try {
    for (var _iterator4 = body[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
      var element = _step4.value;

      if (!isClassDecoratableElementPath(element)) {
        continue;
      }
      var elementNode = element.node;
      if (!elementNode["static"] && _core.types.isPrivateName(elementNode.key)) {
        instancePrivateNames.push(elementNode.key.id.name);
      }
      if (isDecorated(elementNode)) {
        switch (elementNode.type) {
          case "ClassProperty":
            propertyVisitor.ClassProperty(element, state);
            break;
          case "ClassPrivateProperty":
            propertyVisitor.ClassPrivateProperty(element, state);
            break;
          case "ClassAccessorProperty":
            propertyVisitor.ClassAccessorProperty(element, state);
            if (version === "2023-11") {
              break;
            }
          default:
            if (elementNode["static"]) {
              staticInitLocal != null ? staticInitLocal : staticInitLocal = generateLetUidIdentifier(scopeParent, "initStatic");
            } else {
              protoInitLocal != null ? protoInitLocal : protoInitLocal = generateLetUidIdentifier(scopeParent, "initProto");
            }
            break;
        }
        hasElementDecorators = true;
        elemDecsUseFnContext || (elemDecsUseFnContext = elementNode.decorators.some(usesFunctionContextOrYieldAwait));
      } else if (elementNode.type === "ClassAccessorProperty") {
        propertyVisitor.ClassAccessorProperty(element, state);
        var key = elementNode.key;
        var value = elementNode.value;
        var isStatic = elementNode["static"];
        var computed = elementNode.computed;

        var newId = generateClassPrivateUid();
        var newField = generateClassProperty(newId, value, isStatic);
        var keyPath = element.get("key");

        var _element$replaceWith = element.replaceWith(newField);

        var _element$replaceWith2 = _slicedToArray(_element$replaceWith, 1);

        var newPath = _element$replaceWith2[0];

        var getterKey = undefined,
            setterKey = undefined;
        if (computed && !keyPath.isConstantExpression()) {
          getterKey = (0, _misc.memoiseComputedKey)(createToPropertyKeyCall(state, key), scopeParent, scopeParent.generateUid("computedKey"));
          setterKey = _core.types.cloneNode(getterKey.left);
        } else {
          getterKey = _core.types.cloneNode(key);
          setterKey = _core.types.cloneNode(key);
        }
        assignIdForAnonymousClass(path, className);
        addProxyAccessorsFor(path.node.id, newPath, getterKey, setterKey, newId, computed, isStatic, version);
      }
      if ("computed" in element.node && element.node.computed) {
        hasComputedKeysSideEffects || (hasComputedKeysSideEffects = !scopeParent.isStatic(element.node.key));
      }
    }
  } catch (err) {
    _didIteratorError4 = true;
    _iteratorError4 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion4 && _iterator4["return"]) {
        _iterator4["return"]();
      }
    } finally {
      if (_didIteratorError4) {
        throw _iteratorError4;
      }
    }
  }

  if (!classDecorators && !hasElementDecorators) {
    if (!path.node.id && typeof className === "string") {
      path.node.id = _core.types.identifier(className);
    }
    if (setClassName) {
      path.node.body.body.unshift(createStaticBlockFromExpressions([createSetFunctionNameCall(state, setClassName)]));
    }
    return;
  }
  var elementDecoratorInfo = [];
  var constructorPath = undefined;
  var decoratedPrivateMethods = new Set();
  var classInitLocal = undefined,
      classIdLocal = undefined;
  var decoratorReceiverId = null;
  function handleDecorators(decorators) {
    var hasSideEffects = false;
    var usesFnContext = false;
    var decoratorsThis = [];
    var _iteratorNormalCompletion5 = true;
    var _didIteratorError5 = false;
    var _iteratorError5 = undefined;

    try {
      for (var _iterator5 = decorators[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
        var decorator = _step5.value;
        var expression = decorator.expression;

        var object = undefined;
        if ((version === "2023-11" || version === "2023-05") && _core.types.isMemberExpression(expression)) {
          if (_core.types.isSuper(expression.object)) {
            object = _core.types.thisExpression();
          } else if (scopeParent.isStatic(expression.object)) {
            object = _core.types.cloneNode(expression.object);
          } else {
            decoratorReceiverId != null ? decoratorReceiverId : decoratorReceiverId = generateLetUidIdentifier(scopeParent, "obj");
            object = _core.types.assignmentExpression("=", _core.types.cloneNode(decoratorReceiverId), expression.object);
            expression.object = _core.types.cloneNode(decoratorReceiverId);
          }
        }
        decoratorsThis.push(object);
        hasSideEffects || (hasSideEffects = !scopeParent.isStatic(expression));
        usesFnContext || (usesFnContext = usesFunctionContextOrYieldAwait(decorator));
      }
    } catch (err) {
      _didIteratorError5 = true;
      _iteratorError5 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion5 && _iterator5["return"]) {
          _iterator5["return"]();
        }
      } finally {
        if (_didIteratorError5) {
          throw _iteratorError5;
        }
      }
    }

    return {
      hasSideEffects: hasSideEffects,
      usesFnContext: usesFnContext,
      decoratorsThis: decoratorsThis
    };
  }
  var willExtractSomeElemDecs = hasComputedKeysSideEffects || elemDecsUseFnContext || version !== "2023-11";
  var needsDeclaraionForClassBinding = false;
  var classDecorationsFlag = 0;
  var classDecorations = [];
  var classDecorationsId = undefined;
  var computedKeyAssignments = [];
  if (classDecorators) {
    classInitLocal = generateLetUidIdentifier(scopeParent, "initClass");
    needsDeclaraionForClassBinding = path.isClassDeclaration();

    var _replaceClassWithVar = replaceClassWithVar(path, className);

    classIdLocal = _replaceClassWithVar.id;
    path = _replaceClassWithVar.path;

    path.node.decorators = null;
    var classDecsUsePrivateName = classDecorators.some(usesPrivateField);

    var _handleDecorators = handleDecorators(classDecorators);

    var hasSideEffects = _handleDecorators.hasSideEffects;
    var usesFnContext = _handleDecorators.usesFnContext;
    var decoratorsThis = _handleDecorators.decoratorsThis;

    var _generateDecorationList = generateDecorationList(classDecorators, decoratorsThis, version);

    var haveThis = _generateDecorationList.haveThis;
    var decs = _generateDecorationList.decs;

    classDecorationsFlag = haveThis ? 1 : 0;
    classDecorations = decs;
    if (usesFnContext || hasSideEffects && willExtractSomeElemDecs || classDecsUsePrivateName) {
      classDecorationsId = memoiseExpression(_core.types.arrayExpression(classDecorations), "classDecs", classAssignments);
    }
    if (!hasElementDecorators) {
      var _iteratorNormalCompletion6 = true;
      var _didIteratorError6 = false;
      var _iteratorError6 = undefined;

      try {
        for (var _iterator6 = path.get("body.body")[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
          var element = _step6.value;
          var node = element.node;

          var isComputed = "computed" in node && node.computed;
          if (isComputed) {
            if (element.isClassProperty({
              "static": true
            })) {
              if (!element.get("key").isConstantExpression()) {
                var key = node.key;
                var maybeAssignment = (0, _misc.memoiseComputedKey)(key, scopeParent, scopeParent.generateUid("computedKey"));
                if (maybeAssignment != null) {
                  node.key = _core.types.cloneNode(maybeAssignment.left);
                  computedKeyAssignments.push(maybeAssignment);
                }
              }
            } else if (computedKeyAssignments.length > 0) {
              prependExpressionsToComputedKey(computedKeyAssignments, element);
              computedKeyAssignments = [];
            }
          }
        }
      } catch (err) {
        _didIteratorError6 = true;
        _iteratorError6 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion6 && _iterator6["return"]) {
            _iterator6["return"]();
          }
        } finally {
          if (_didIteratorError6) {
            throw _iteratorError6;
          }
        }
      }
    }
  } else {
    assignIdForAnonymousClass(path, className);
    classIdLocal = _core.types.cloneNode(path.node.id);
  }
  var lastInstancePrivateName = undefined;
  var needsInstancePrivateBrandCheck = false;
  var fieldInitializerExpressions = [];
  var staticFieldInitializerExpressions = [];
  if (hasElementDecorators) {
    if (protoInitLocal) {
      var protoInitCall = _core.types.callExpression(_core.types.cloneNode(protoInitLocal), [_core.types.thisExpression()]);
      fieldInitializerExpressions.push(protoInitCall);
    }
    var _iteratorNormalCompletion7 = true;
    var _didIteratorError7 = false;
    var _iteratorError7 = undefined;

    try {
      for (var _iterator7 = body[Symbol.iterator](), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
        var element = _step7.value;

        if (!isClassDecoratableElementPath(element)) {
          if (staticFieldInitializerExpressions.length > 0 && element.isStaticBlock()) {
            prependExpressionsToStaticBlock(staticFieldInitializerExpressions, element);
            staticFieldInitializerExpressions = [];
          }
          continue;
        }
        var node = element.node;

        var decorators = node.decorators;
        var _hasDecorators = !!(decorators != null && decorators.length);
        var isComputed = "computed" in node && node.computed;
        var _name3 = "computedKey";
        if (node.key.type === "PrivateName") {
          _name3 = node.key.id.name;
        } else if (!isComputed && node.key.type === "Identifier") {
          _name3 = node.key.name;
        }
        var decoratorsArray = undefined;
        var decoratorsHaveThis = undefined;
        if (_hasDecorators) {
          var _handleDecorators2 = handleDecorators(decorators);

          var hasSideEffects = _handleDecorators2.hasSideEffects;
          var usesFnContext = _handleDecorators2.usesFnContext;
          var decoratorsThis = _handleDecorators2.decoratorsThis;

          var _generateDecorationList2 = generateDecorationList(decorators, decoratorsThis, version);

          var decs = _generateDecorationList2.decs;
          var haveThis = _generateDecorationList2.haveThis;

          decoratorsHaveThis = haveThis;
          decoratorsArray = decs.length === 1 ? decs[0] : _core.types.arrayExpression(decs);
          if (usesFnContext || hasSideEffects && willExtractSomeElemDecs) {
            decoratorsArray = memoiseExpression(decoratorsArray, _name3 + "Decs", computedKeyAssignments);
          }
        }
        if (isComputed) {
          if (!element.get("key").isConstantExpression()) {
            var _key = node.key;
            var maybeAssignment = (0, _misc.memoiseComputedKey)(_hasDecorators ? createToPropertyKeyCall(state, _key) : _key, scopeParent, scopeParent.generateUid("computedKey"));
            if (maybeAssignment != null) {
              if (classDecorators && element.isClassProperty({
                "static": true
              })) {
                node.key = _core.types.cloneNode(maybeAssignment.left);
                computedKeyAssignments.push(maybeAssignment);
              } else {
                node.key = maybeAssignment;
              }
            }
          }
        }
        var key = node.key;
        var isStatic = node["static"];

        var isPrivate = key.type === "PrivateName";
        var kind = getElementKind(element);
        if (isPrivate && !isStatic) {
          if (_hasDecorators) {
            needsInstancePrivateBrandCheck = true;
          }
          if (_core.types.isClassPrivateProperty(node) || !lastInstancePrivateName) {
            lastInstancePrivateName = key;
          }
        }
        if (element.isClassMethod({
          kind: "constructor"
        })) {
          constructorPath = element;
        }
        var locals = undefined;
        if (_hasDecorators) {
          var privateMethods = undefined;
          var nameExpr = undefined;
          if (isComputed) {
            nameExpr = getComputedKeyMemoiser(element.get("key"));
          } else if (key.type === "PrivateName") {
            nameExpr = _core.types.stringLiteral(key.id.name);
          } else if (key.type === "Identifier") {
            nameExpr = _core.types.stringLiteral(key.name);
          } else {
            nameExpr = _core.types.cloneNode(key);
          }
          if (kind === ACCESSOR) {
            var value = element.node.value;

            var params = version === "2023-11" && isStatic ? [] : [_core.types.thisExpression()];
            if (value) {
              params.push(_core.types.cloneNode(value));
            }
            var newId = generateClassPrivateUid();
            var newFieldInitId = generateLetUidIdentifier(scopeParent, "init_" + _name3);
            var newValue = _core.types.callExpression(_core.types.cloneNode(newFieldInitId), params);
            var newField = generateClassProperty(newId, newValue, isStatic);

            var _element$replaceWith3 = element.replaceWith(newField);

            var _element$replaceWith32 = _slicedToArray(_element$replaceWith3, 1);

            var newPath = _element$replaceWith32[0];

            if (isPrivate) {
              privateMethods = extractProxyAccessorsFor(newId, version);
              var getId = generateLetUidIdentifier(scopeParent, "get_" + _name3);
              var setId = generateLetUidIdentifier(scopeParent, "set_" + _name3);
              addCallAccessorsFor(version, newPath, key, getId, setId, isStatic);
              locals = [newFieldInitId, getId, setId];
            } else {
              assignIdForAnonymousClass(path, className);
              addProxyAccessorsFor(path.node.id, newPath, _core.types.cloneNode(key), _core.types.isAssignmentExpression(key) ? _core.types.cloneNode(key.left) : _core.types.cloneNode(key), newId, isComputed, isStatic, version);
              locals = [newFieldInitId];
            }
          } else if (kind === FIELD) {
            var initId = generateLetUidIdentifier(scopeParent, "init_" + _name3);
            var valuePath = element.get("value");
            var args = version === "2023-11" && isStatic ? [] : [_core.types.thisExpression()];
            if (valuePath.node) args.push(valuePath.node);
            valuePath.replaceWith(_core.types.callExpression(_core.types.cloneNode(initId), args));
            locals = [initId];
            if (isPrivate) {
              privateMethods = extractProxyAccessorsFor(key, version);
            }
          } else if (isPrivate) {
            var callId = generateLetUidIdentifier(scopeParent, "call_" + _name3);
            locals = [callId];
            var replaceSupers = new _helperReplaceSupers["default"]({
              constantSuper: constantSuper,
              methodPath: element,
              objectRef: classIdLocal,
              superRef: path.node.superClass,
              file: state.file,
              refToPreserve: classIdLocal
            });
            replaceSupers.replace();
            privateMethods = [createFunctionExpressionFromPrivateMethod(element.node)];
            if (kind === GETTER || kind === SETTER) {
              movePrivateAccessor(element, _core.types.cloneNode(key), _core.types.cloneNode(callId), isStatic);
            } else {
              var _node = element.node;
              path.node.body.body.unshift(_core.types.classPrivateProperty(key, _core.types.cloneNode(callId), [], _node["static"]));
              decoratedPrivateMethods.add(key.id.name);
              element.remove();
            }
          }
          elementDecoratorInfo.push({
            kind: kind,
            decoratorsArray: decoratorsArray,
            decoratorsHaveThis: decoratorsHaveThis,
            name: nameExpr,
            isStatic: isStatic,
            privateMethods: privateMethods,
            locals: locals
          });
          if (element.node) {
            element.node.decorators = null;
          }
        }
        if (isComputed && computedKeyAssignments.length > 0) {
          if (classDecorators && element.isClassProperty({
            "static": true
          })) {} else {
            prependExpressionsToComputedKey(computedKeyAssignments, kind === ACCESSOR ? element.getNextSibling() : element);
            computedKeyAssignments = [];
          }
        }
        if (fieldInitializerExpressions.length > 0 && !isStatic && (kind === FIELD || kind === ACCESSOR)) {
          prependExpressionsToFieldInitializer(fieldInitializerExpressions, element);
          fieldInitializerExpressions = [];
        }
        if (staticFieldInitializerExpressions.length > 0 && isStatic && (kind === FIELD || kind === ACCESSOR)) {
          prependExpressionsToFieldInitializer(staticFieldInitializerExpressions, element);
          staticFieldInitializerExpressions = [];
        }
        if (_hasDecorators && version === "2023-11") {
          if (kind === FIELD || kind === ACCESSOR) {
            var initExtraId = generateLetUidIdentifier(scopeParent, "init_extra_" + _name3);
            locals.push(initExtraId);
            var initExtraCall = _core.types.callExpression(_core.types.cloneNode(initExtraId), isStatic ? [] : [_core.types.thisExpression()]);
            if (!isStatic) {
              fieldInitializerExpressions.push(initExtraCall);
            } else {
              staticFieldInitializerExpressions.push(initExtraCall);
            }
          }
        }
      }
    } catch (err) {
      _didIteratorError7 = true;
      _iteratorError7 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion7 && _iterator7["return"]) {
          _iterator7["return"]();
        }
      } finally {
        if (_didIteratorError7) {
          throw _iteratorError7;
        }
      }
    }
  }
  if (computedKeyAssignments.length > 0) {
    var elements = path.get("body.body");
    var lastComputedElement = undefined;
    for (var i = elements.length - 1; i >= 0; i--) {
      var _path = elements[i];
      var node = _path.node;
      if (node.computed) {
        if (classDecorators && _core.types.isClassProperty(node, {
          "static": true
        })) {
          continue;
        }
        lastComputedElement = _path;
        break;
      }
    }
    if (lastComputedElement != null) {
      appendExpressionsToComputedKey(computedKeyAssignments, lastComputedElement);
      computedKeyAssignments = [];
    } else {}
  }
  if (fieldInitializerExpressions.length > 0) {
    var isDerivedClass = !!path.node.superClass;
    if (constructorPath) {
      if (isDerivedClass) {
        insertExpressionsAfterSuperCallAndOptimize(fieldInitializerExpressions, constructorPath, protoInitLocal);
      } else {
        prependExpressionsToConstructor(fieldInitializerExpressions, constructorPath);
      }
    } else {
      path.node.body.body.unshift(createConstructorFromExpressions(fieldInitializerExpressions, isDerivedClass));
    }
    fieldInitializerExpressions = [];
  }
  if (staticFieldInitializerExpressions.length > 0) {
    path.node.body.body.push(createStaticBlockFromExpressions(staticFieldInitializerExpressions));
    staticFieldInitializerExpressions = [];
  }
  var sortedElementDecoratorInfo = toSortedDecoratorInfo(elementDecoratorInfo);
  var elementDecorations = generateDecorationExprs(version === "2023-11" ? elementDecoratorInfo : sortedElementDecoratorInfo, version);
  var elementLocals = extractElementLocalAssignments(sortedElementDecoratorInfo);
  if (protoInitLocal) {
    elementLocals.push(protoInitLocal);
  }
  if (staticInitLocal) {
    elementLocals.push(staticInitLocal);
  }
  var classLocals = [];
  var classInitInjected = false;
  var classInitCall = classInitLocal && _core.types.callExpression(_core.types.cloneNode(classInitLocal), []);
  var originalClassPath = path;
  var originalClass = path.node;
  var staticClosures = [];
  if (classDecorators) {
    (function () {
      classLocals.push(classIdLocal, classInitLocal);
      var statics = [];
      path.get("body.body").forEach(function (element) {
        if (element.isStaticBlock()) {
          if (hasInstancePrivateAccess(element, instancePrivateNames)) {
            var staticBlockClosureId = memoiseExpression(staticBlockToFunctionClosure(element.node), "staticBlock", staticClosures);
            staticFieldInitializerExpressions.push(_core.types.callExpression(_core.types.memberExpression(staticBlockClosureId, _core.types.identifier("call")), [_core.types.thisExpression()]));
          } else {
            staticFieldInitializerExpressions.push(staticBlockToIIFE(element.node));
          }
          element.remove();
          return;
        }
        if ((element.isClassProperty() || element.isClassPrivateProperty()) && element.node["static"]) {
          var valuePath = element.get("value");
          if (hasInstancePrivateAccess(valuePath, instancePrivateNames)) {
            var fieldValueClosureId = memoiseExpression(fieldInitializerToClosure(valuePath.node), "fieldValue", staticClosures);
            valuePath.replaceWith(_core.types.callExpression(_core.types.memberExpression(fieldValueClosureId, _core.types.identifier("call")), [_core.types.thisExpression()]));
          }
          if (staticFieldInitializerExpressions.length > 0) {
            prependExpressionsToFieldInitializer(staticFieldInitializerExpressions, element);
            staticFieldInitializerExpressions = [];
          }
          element.node["static"] = false;
          statics.push(element.node);
          element.remove();
        } else if (element.isClassPrivateMethod({
          "static": true
        })) {
          if (hasInstancePrivateAccess(element, instancePrivateNames)) {
            var replaceSupers = new _helperReplaceSupers["default"]({
              constantSuper: constantSuper,
              methodPath: element,
              objectRef: classIdLocal,
              superRef: path.node.superClass,
              file: state.file,
              refToPreserve: classIdLocal
            });
            replaceSupers.replace();
            var privateMethodDelegateId = memoiseExpression(createFunctionExpressionFromPrivateMethod(element.node), element.get("key.id").node.name, staticClosures);
            if (ignoreFunctionLength) {
              element.node.params = [_core.types.restElement(_core.types.identifier("arg"))];
              element.node.body = _core.types.blockStatement([_core.types.returnStatement(_core.types.callExpression(_core.types.memberExpression(privateMethodDelegateId, _core.types.identifier("apply")), [_core.types.thisExpression(), _core.types.identifier("arg")]))]);
            } else {
              element.node.params = element.node.params.map(function (p, i) {
                if (_core.types.isRestElement(p)) {
                  return _core.types.restElement(_core.types.identifier("arg"));
                } else {
                  return _core.types.identifier("_" + i);
                }
              });
              element.node.body = _core.types.blockStatement([_core.types.returnStatement(_core.types.callExpression(_core.types.memberExpression(privateMethodDelegateId, _core.types.identifier("apply")), [_core.types.thisExpression(), _core.types.identifier("arguments")]))]);
            }
          }
          element.node["static"] = false;
          statics.push(element.node);
          element.remove();
        }
      });
      if (statics.length > 0 || staticFieldInitializerExpressions.length > 0) {
        var staticsClass = _core.template.expression.ast(_templateObject5, state.addHelper("identity"));
        staticsClass.body.body = [_core.types.classProperty(_core.types.toExpression(originalClass), undefined, undefined, undefined, true, true)].concat(statics);
        var constructorBody = [];
        var newExpr = _core.types.newExpression(staticsClass, []);
        if (staticFieldInitializerExpressions.length > 0) {
          constructorBody.push.apply(constructorBody, _toConsumableArray(staticFieldInitializerExpressions));
        }
        if (classInitCall) {
          classInitInjected = true;
          constructorBody.push(classInitCall);
        }
        if (constructorBody.length > 0) {
          constructorBody.unshift(_core.types.callExpression(_core.types["super"](), [_core.types.cloneNode(classIdLocal)]));
          staticsClass.body.body.push(createConstructorFromExpressions(constructorBody, false));
        } else {
          newExpr.arguments.push(_core.types.cloneNode(classIdLocal));
        }

        var _path$replaceWith3 = path.replaceWith(newExpr);

        var _path$replaceWith32 = _slicedToArray(_path$replaceWith3, 1);

        var newPath = _path$replaceWith32[0];

        originalClassPath = newPath.get("callee").get("body").get("body.0.key");
      }
    })();
  }
  if (!classInitInjected && classInitCall) {
    path.node.body.body.push(_core.types.staticBlock([_core.types.expressionStatement(classInitCall)]));
  }
  var superClass = originalClass.superClass;

  if (superClass && (version === "2023-11" || version === "2023-05")) {
    var id = path.scope.maybeGenerateMemoised(superClass);
    if (id) {
      originalClass.superClass = _core.types.assignmentExpression("=", id, superClass);
      superClass = id;
    }
  }
  var applyDecoratorWrapper = _core.types.staticBlock([]);
  originalClass.body.body.unshift(applyDecoratorWrapper);
  var applyDecsBody = applyDecoratorWrapper.body;
  if (computedKeyAssignments.length > 0) {
    var elements = originalClassPath.get("body.body");
    var firstPublicElement = undefined;
    var _iteratorNormalCompletion8 = true;
    var _didIteratorError8 = false;
    var _iteratorError8 = undefined;

    try {
      for (var _iterator8 = elements[Symbol.iterator](), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
        var _path2 = _step8.value;

        if ((_path2.isClassProperty() || _path2.isClassMethod()) && _path2.node.kind !== "constructor") {
          firstPublicElement = _path2;
          break;
        }
      }
    } catch (err) {
      _didIteratorError8 = true;
      _iteratorError8 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion8 && _iterator8["return"]) {
          _iterator8["return"]();
        }
      } finally {
        if (_didIteratorError8) {
          throw _iteratorError8;
        }
      }
    }

    if (firstPublicElement != null) {
      convertToComputedKey(firstPublicElement);
      prependExpressionsToComputedKey(computedKeyAssignments, firstPublicElement);
    } else {
      originalClass.body.body.unshift(_core.types.classProperty(_core.types.sequenceExpression([].concat(_toConsumableArray(computedKeyAssignments), [_core.types.stringLiteral("_")])), undefined, undefined, undefined, true, true));
      applyDecsBody.push(_core.types.expressionStatement(_core.types.unaryExpression("delete", _core.types.memberExpression(_core.types.thisExpression(), _core.types.identifier("_")))));
    }
    computedKeyAssignments = [];
  }
  applyDecsBody.push(_core.types.expressionStatement(createLocalsAssignment(elementLocals, classLocals, elementDecorations, classDecorationsId != null ? classDecorationsId : _core.types.arrayExpression(classDecorations), _core.types.numericLiteral(classDecorationsFlag), needsInstancePrivateBrandCheck ? lastInstancePrivateName : null, setClassName, _core.types.cloneNode(superClass), state, version)));
  if (staticInitLocal) {
    applyDecsBody.push(_core.types.expressionStatement(_core.types.callExpression(_core.types.cloneNode(staticInitLocal), [_core.types.thisExpression()])));
  }
  if (staticClosures.length > 0) {
    applyDecsBody.push.apply(applyDecsBody, _toConsumableArray(staticClosures.map(function (expr) {
      return _core.types.expressionStatement(expr);
    })));
  }
  path.insertBefore(classAssignments.map(function (expr) {
    return _core.types.expressionStatement(expr);
  }));
  if (needsDeclaraionForClassBinding) {
    var classBindingInfo = scopeParent.getBinding(classIdLocal.name);
    if (!classBindingInfo.constantViolations.length) {
      path.insertBefore(_core.types.variableDeclaration("let", [_core.types.variableDeclarator(_core.types.cloneNode(classIdLocal))]));
    } else {
      var classOuterBindingDelegateLocal = scopeParent.generateUidIdentifier("t" + classIdLocal.name);
      var classOuterBindingLocal = classIdLocal;
      path.replaceWithMultiple([_core.types.variableDeclaration("let", [_core.types.variableDeclarator(_core.types.cloneNode(classOuterBindingLocal)), _core.types.variableDeclarator(classOuterBindingDelegateLocal)]), _core.types.blockStatement([_core.types.variableDeclaration("let", [_core.types.variableDeclarator(_core.types.cloneNode(classIdLocal))]), path.node, _core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.cloneNode(classOuterBindingDelegateLocal), _core.types.cloneNode(classIdLocal)))]), _core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.cloneNode(classOuterBindingLocal), _core.types.cloneNode(classOuterBindingDelegateLocal)))]);
    }
  }
  if (decoratedPrivateMethods.size > 0) {
    checkPrivateMethodUpdateError(path, decoratedPrivateMethods);
  }
  path.scope.crawl();
  return path;
}
function createLocalsAssignment(elementLocals, classLocals, elementDecorations, classDecorations, classDecorationsFlag, maybePrivateBrandName, setClassName, superClass, state, version) {
  var lhs = undefined,
      rhs = undefined;
  var args = [setClassName ? createSetFunctionNameCall(state, setClassName) : _core.types.thisExpression(), classDecorations, elementDecorations];
  {
    if (version !== "2023-11") {
      args.splice(1, 2, elementDecorations, classDecorations);
    }
    if (version === "2021-12" || version === "2022-03" && !state.availableHelper("applyDecs2203R")) {
      lhs = _core.types.arrayPattern([].concat(_toConsumableArray(elementLocals), _toConsumableArray(classLocals)));
      rhs = _core.types.callExpression(state.addHelper(version === "2021-12" ? "applyDecs" : "applyDecs2203"), args);
      return _core.types.assignmentExpression("=", lhs, rhs);
    } else if (version === "2022-03") {
      rhs = _core.types.callExpression(state.addHelper("applyDecs2203R"), args);
    } else if (version === "2023-01") {
      if (maybePrivateBrandName) {
        args.push(createPrivateBrandCheckClosure(maybePrivateBrandName));
      }
      rhs = _core.types.callExpression(state.addHelper("applyDecs2301"), args);
    } else if (version === "2023-05") {
      if (maybePrivateBrandName || superClass || classDecorationsFlag.value !== 0) {
        args.push(classDecorationsFlag);
      }
      if (maybePrivateBrandName) {
        args.push(createPrivateBrandCheckClosure(maybePrivateBrandName));
      } else if (superClass) {
        args.push(_core.types.unaryExpression("void", _core.types.numericLiteral(0)));
      }
      if (superClass) args.push(superClass);
      rhs = _core.types.callExpression(state.addHelper("applyDecs2305"), args);
    }
  }
  if (version === "2023-11") {
    if (maybePrivateBrandName || superClass || classDecorationsFlag.value !== 0) {
      args.push(classDecorationsFlag);
    }
    if (maybePrivateBrandName) {
      args.push(createPrivateBrandCheckClosure(maybePrivateBrandName));
    } else if (superClass) {
      args.push(_core.types.unaryExpression("void", _core.types.numericLiteral(0)));
    }
    if (superClass) args.push(superClass);
    rhs = _core.types.callExpression(state.addHelper("applyDecs2311"), args);
  }
  if (elementLocals.length > 0) {
    if (classLocals.length > 0) {
      lhs = _core.types.objectPattern([_core.types.objectProperty(_core.types.identifier("e"), _core.types.arrayPattern(elementLocals)), _core.types.objectProperty(_core.types.identifier("c"), _core.types.arrayPattern(classLocals))]);
    } else {
      lhs = _core.types.arrayPattern(elementLocals);
      rhs = _core.types.memberExpression(rhs, _core.types.identifier("e"), false, false);
    }
  } else {
    lhs = _core.types.arrayPattern(classLocals);
    rhs = _core.types.memberExpression(rhs, _core.types.identifier("c"), false, false);
  }
  return _core.types.assignmentExpression("=", lhs, rhs);
}
function isProtoKey(node) {
  return node.type === "Identifier" ? node.name === "__proto__" : node.value === "__proto__";
}
function isDecorated(node) {
  return node.decorators && node.decorators.length > 0;
}
function shouldTransformElement(node) {
  switch (node.type) {
    case "ClassAccessorProperty":
      return true;
    case "ClassMethod":
    case "ClassProperty":
    case "ClassPrivateMethod":
    case "ClassPrivateProperty":
      return isDecorated(node);
    default:
      return false;
  }
}
function shouldTransformClass(node) {
  return isDecorated(node) || node.body.body.some(shouldTransformElement);
}
function NamedEvaluationVisitoryFactory(isAnonymous, visitor) {
  function handleComputedProperty(propertyPath, key, state) {
    switch (key.type) {
      case "StringLiteral":
        return _core.types.stringLiteral(key.value);
      case "NumericLiteral":
      case "BigIntLiteral":
        {
          var keyValue = key.value + "";
          propertyPath.get("key").replaceWith(_core.types.stringLiteral(keyValue));
          return _core.types.stringLiteral(keyValue);
        }
      default:
        {
          var ref = propertyPath.scope.maybeGenerateMemoised(key);
          propertyPath.get("key").replaceWith(_core.types.assignmentExpression("=", ref, createToPropertyKeyCall(state, key)));
          return _core.types.cloneNode(ref);
        }
    }
  }
  return {
    VariableDeclarator: function VariableDeclarator(path, state) {
      var id = path.node.id;
      if (id.type === "Identifier") {
        var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("init"));
        if (isAnonymous(initializer)) {
          var _name4 = id.name;
          visitor(initializer, state, _name4);
        }
      }
    },
    AssignmentExpression: function AssignmentExpression(path, state) {
      var id = path.node.left;
      if (id.type === "Identifier") {
        var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("right"));
        if (isAnonymous(initializer)) {
          switch (path.node.operator) {
            case "=":
            case "&&=":
            case "||=":
            case "??=":
              visitor(initializer, state, id.name);
          }
        }
      }
    },
    AssignmentPattern: function AssignmentPattern(path, state) {
      var id = path.node.left;
      if (id.type === "Identifier") {
        var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("right"));
        if (isAnonymous(initializer)) {
          var _name5 = id.name;
          visitor(initializer, state, _name5);
        }
      }
    },
    ObjectExpression: function ObjectExpression(path, state) {
      var _iteratorNormalCompletion9 = true;
      var _didIteratorError9 = false;
      var _iteratorError9 = undefined;

      try {
        for (var _iterator9 = path.get("properties")[Symbol.iterator](), _step9; !(_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done); _iteratorNormalCompletion9 = true) {
          var propertyPath = _step9.value;

          if (!propertyPath.isObjectProperty()) continue;
          var node = propertyPath.node;

          var id = node.key;
          var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(propertyPath.get("value"));
          if (isAnonymous(initializer)) {
            if (!node.computed) {
              if (!isProtoKey(id)) {
                if (id.type === "Identifier") {
                  visitor(initializer, state, id.name);
                } else {
                  var className = _core.types.stringLiteral(id.value + "");
                  visitor(initializer, state, className);
                }
              }
            } else {
              var ref = handleComputedProperty(propertyPath, id, state);
              visitor(initializer, state, ref);
            }
          }
        }
      } catch (err) {
        _didIteratorError9 = true;
        _iteratorError9 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion9 && _iterator9["return"]) {
            _iterator9["return"]();
          }
        } finally {
          if (_didIteratorError9) {
            throw _iteratorError9;
          }
        }
      }
    },
    ClassPrivateProperty: function ClassPrivateProperty(path, state) {
      var node = path.node;

      var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("value"));
      if (isAnonymous(initializer)) {
        var className = _core.types.stringLiteral("#" + node.key.id.name);
        visitor(initializer, state, className);
      }
    },
    ClassAccessorProperty: function ClassAccessorProperty(path, state) {
      var node = path.node;

      var id = node.key;
      var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("value"));
      if (isAnonymous(initializer)) {
        if (!node.computed) {
          if (id.type === "Identifier") {
            visitor(initializer, state, id.name);
          } else if (id.type === "PrivateName") {
            var className = _core.types.stringLiteral("#" + id.id.name);
            visitor(initializer, state, className);
          } else {
            var className = _core.types.stringLiteral(id.value + "");
            visitor(initializer, state, className);
          }
        } else {
          var ref = handleComputedProperty(path, id, state);
          visitor(initializer, state, ref);
        }
      }
    },
    ClassProperty: function ClassProperty(path, state) {
      var node = path.node;

      var id = node.key;
      var initializer = (0, _helperSkipTransparentExpressionWrappers.skipTransparentExprWrappers)(path.get("value"));
      if (isAnonymous(initializer)) {
        if (!node.computed) {
          if (id.type === "Identifier") {
            visitor(initializer, state, id.name);
          } else {
            var className = _core.types.stringLiteral(id.value + "");
            visitor(initializer, state, className);
          }
        } else {
          var ref = handleComputedProperty(path, id, state);
          visitor(initializer, state, ref);
        }
      }
    }
  };
}
function isDecoratedAnonymousClassExpression(path) {
  return path.isClassExpression({
    id: null
  }) && shouldTransformClass(path.node);
}
function generateLetUidIdentifier(scope, name) {
  var id = scope.generateUidIdentifier(name);
  scope.push({
    id: id,
    kind: "let"
  });
  return _core.types.cloneNode(id);
}
function _default(_ref, _ref2, version, inherits) {
  var assertVersion = _ref.assertVersion;
  var assumption = _ref.assumption;
  var loose = _ref2.loose;

  var _assumption, _assumption2;
  {
    if (version === "2023-11" || version === "2023-05" || version === "2023-01") {
      assertVersion("^7.21.0");
    } else if (version === "2021-12") {
      assertVersion("^7.16.0");
    } else {
      assertVersion("^7.19.0");
    }
  }
  var VISITED = new WeakSet();
  var constantSuper = (_assumption = assumption("constantSuper")) != null ? _assumption : loose;
  var ignoreFunctionLength = (_assumption2 = assumption("ignoreFunctionLength")) != null ? _assumption2 : loose;
  var namedEvaluationVisitor = NamedEvaluationVisitoryFactory(isDecoratedAnonymousClassExpression, visitClass);
  function visitClass(path, state, className) {
    var _node$id;
    if (VISITED.has(path)) return;
    var node = path.node;

    className != null ? className : className = (_node$id = node.id) == null ? void 0 : _node$id.name;
    var newPath = transformClass(path, state, constantSuper, ignoreFunctionLength, className, namedEvaluationVisitor, version);
    if (newPath) {
      VISITED.add(newPath);
      return;
    }
    VISITED.add(path);
  }
  return {
    name: "proposal-decorators",
    inherits: inherits,
    visitor: Object.assign({
      ExportDefaultDeclaration: function ExportDefaultDeclaration(path, state) {
        var declaration = path.node.declaration;

        if ((declaration == null ? void 0 : declaration.type) === "ClassDeclaration" && isDecorated(declaration)) {
          var isAnonymous = !declaration.id;
          {
            var _path$splitExportDecl;
            (_path$splitExportDecl = path.splitExportDeclaration) != null ? _path$splitExportDecl : path.splitExportDeclaration = require("@babel/traverse").NodePath.prototype.splitExportDeclaration;
          }
          var updatedVarDeclarationPath = path.splitExportDeclaration();
          if (isAnonymous) {
            visitClass(updatedVarDeclarationPath, state, _core.types.stringLiteral("default"));
          }
        }
      },
      ExportNamedDeclaration: function ExportNamedDeclaration(path) {
        var declaration = path.node.declaration;

        if ((declaration == null ? void 0 : declaration.type) === "ClassDeclaration" && isDecorated(declaration)) {
          {
            var _path$splitExportDecl2;
            (_path$splitExportDecl2 = path.splitExportDeclaration) != null ? _path$splitExportDecl2 : path.splitExportDeclaration = require("@babel/traverse").NodePath.prototype.splitExportDeclaration;
          }
          path.splitExportDeclaration();
        }
      },
      Class: function Class(path, state) {
        visitClass(path, state, undefined);
      }
    }, namedEvaluationVisitor)
  };
}
