// @babel/helper-create-class-features-plugin/lib/index.js
"use strict";

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FEATURES", {
  enumerable: true,
  get: function get() {
    return _features.FEATURES;
  }
});
Object.defineProperty(exports, "buildCheckInRHS", {
  enumerable: true,
  get: function get() {
    return _fields.buildCheckInRHS;
  }
});
exports.createClassFeaturePlugin = createClassFeaturePlugin;
Object.defineProperty(exports, "enableFeature", {
  enumerable: true,
  get: function get() {
    return _features.enableFeature;
  }
});
Object.defineProperty(exports, "injectInitialization", {
  enumerable: true,
  get: function get() {
    return _misc.injectInitialization;
  }
});
var _core = require("@babel/core");
var _semver = require("semver");
var _fields = require("./fields.js");
var _decorators = require("./decorators.js");
var _decorators2 = require("./decorators-2018-09.js");
var _misc = require("./misc.js");
var _features = require("./features.js");
var _typescript = require("./typescript.js");
var versionKey = "@babel/plugin-class-features/version";
function createClassFeaturePlugin(_ref) {
  var name = _ref.name;
  var feature = _ref.feature;
  var loose = _ref.loose;
  var manipulateOptions = _ref.manipulateOptions;
  var api = _ref.api;
  var inherits = _ref.inherits;
  var decoratorVersion = _ref.decoratorVersion;

  var _api$assumption;
  if (feature & _features.FEATURES.decorators) {
    {
      if (decoratorVersion === "2023-11" || decoratorVersion === "2023-05" || decoratorVersion === "2023-01" || decoratorVersion === "2022-03" || decoratorVersion === "2021-12") {
        return (0, _decorators["default"])(api, {
          loose: loose
        }, decoratorVersion, inherits);
      }
    }
  }
  {
    api != null ? api : api = {
      assumption: function assumption() {
        return void 0;
      }
    };
  }
  var setPublicClassFields = api.assumption("setPublicClassFields");
  var privateFieldsAsSymbols = api.assumption("privateFieldsAsSymbols");
  var privateFieldsAsProperties = api.assumption("privateFieldsAsProperties");
  var noUninitializedPrivateFieldAccess = (_api$assumption = api.assumption("noUninitializedPrivateFieldAccess")) != null ? _api$assumption : false;
  var constantSuper = api.assumption("constantSuper");
  var noDocumentAll = api.assumption("noDocumentAll");
  if (privateFieldsAsProperties && privateFieldsAsSymbols) {
    throw new Error("Cannot enable both the \"privateFieldsAsProperties\" and " + "\"privateFieldsAsSymbols\" assumptions as the same time.");
  }
  var privateFieldsAsSymbolsOrProperties = privateFieldsAsProperties || privateFieldsAsSymbols;
  if (loose === true) {
    var explicit = [];
    if (setPublicClassFields !== undefined) {
      explicit.push("\"setPublicClassFields\"");
    }
    if (privateFieldsAsProperties !== undefined) {
      explicit.push("\"privateFieldsAsProperties\"");
    }
    if (privateFieldsAsSymbols !== undefined) {
      explicit.push("\"privateFieldsAsSymbols\"");
    }
    if (explicit.length !== 0) {
      console.warn("[" + name + "]: You are using the \"loose: true\" option and you are" + (" explicitly setting a value for the " + explicit.join(" and ")) + (" assumption" + (explicit.length > 1 ? "s" : "") + ". The \"loose\" option") + " can cause incompatibilities with the other class features" + " plugins, so it's recommended that you replace it with the" + " following top-level option:\n" + "\t\"assumptions\": {\n" + "\t\t\"setPublicClassFields\": true,\n" + "\t\t\"privateFieldsAsSymbols\": true\n" + "\t}");
    }
  }
  return {
    name: name,
    manipulateOptions: manipulateOptions,
    inherits: inherits,
    pre: function pre(file) {
      (0, _features.enableFeature)(file, feature, loose);
      {
        if (typeof file.get(versionKey) === "number") {
          file.set(versionKey, "7.27.1");
          return;
        }
      }
      if (!file.get(versionKey) || _semver.lt(file.get(versionKey), "7.27.1")) {
        file.set(versionKey, "7.27.1");
      }
    },
    visitor: {
      Class: function Class(path, _ref2) {
        var file = _ref2.file;

        if (file.get(versionKey) !== "7.27.1") return;
        if (!(0, _features.shouldTransform)(path, file)) return;
        var pathIsClassDeclaration = path.isClassDeclaration();
        if (pathIsClassDeclaration) (0, _typescript.assertFieldTransformed)(path);
        var loose = (0, _features.isLoose)(file, feature);
        var constructor = undefined;
        var isDecorated = (0, _decorators.hasDecorators)(path.node);
        var props = [];
        var elements = [];
        var computedPaths = [];
        var privateNames = new Set();
        var body = path.get("body");
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = body.get("body")[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var _path = _step.value;

            if ((_path.isClassProperty() || _path.isClassMethod()) && _path.node.computed) {
              computedPaths.push(_path);
            }
            if (_path.isPrivate()) {
              var _name = _path.node.key.id.name;

              var getName = "get " + _name;
              var setName = "set " + _name;
              if (_path.isClassPrivateMethod()) {
                if (_path.node.kind === "get") {
                  if (privateNames.has(getName) || privateNames.has(_name) && !privateNames.has(setName)) {
                    throw _path.buildCodeFrameError("Duplicate private field");
                  }
                  privateNames.add(getName).add(_name);
                } else if (_path.node.kind === "set") {
                  if (privateNames.has(setName) || privateNames.has(_name) && !privateNames.has(getName)) {
                    throw _path.buildCodeFrameError("Duplicate private field");
                  }
                  privateNames.add(setName).add(_name);
                }
              } else {
                if (privateNames.has(_name) && !privateNames.has(getName) && !privateNames.has(setName) || privateNames.has(_name) && (privateNames.has(getName) || privateNames.has(setName))) {
                  throw _path.buildCodeFrameError("Duplicate private field");
                }
                privateNames.add(_name);
              }
            }
            if (_path.isClassMethod({
              kind: "constructor"
            })) {
              constructor = _path;
            } else {
              elements.push(_path);
              if (_path.isProperty() || _path.isPrivate() || _path.isStaticBlock != null && _path.isStaticBlock()) {
                props.push(_path);
              }
            }
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"]) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        {
          if (!props.length && !isDecorated) return;
        }
        var innerBinding = path.node.id;
        var ref = undefined;
        if (!innerBinding || !pathIsClassDeclaration) {
          {
            var _path$ensureFunctionN;
            (_path$ensureFunctionN = path.ensureFunctionName) != null ? _path$ensureFunctionN : path.ensureFunctionName = require("@babel/traverse").NodePath.prototype.ensureFunctionName;
          }
          path.ensureFunctionName(false);
          ref = path.scope.generateUidIdentifier((innerBinding == null ? void 0 : innerBinding.name) || "Class");
        }
        var classRefForDefine = ref != null ? ref : _core.types.cloneNode(innerBinding);
        var privateNamesMap = (0, _fields.buildPrivateNamesMap)(classRefForDefine.name, privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose, props, file);
        var privateNamesNodes = (0, _fields.buildPrivateNamesNodes)(privateNamesMap, privateFieldsAsProperties != null ? privateFieldsAsProperties : loose, privateFieldsAsSymbols != null ? privateFieldsAsSymbols : false, file);
        (0, _fields.transformPrivateNamesUsage)(classRefForDefine, path, privateNamesMap, {
          privateFieldsAsProperties: privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose,
          noUninitializedPrivateFieldAccess: noUninitializedPrivateFieldAccess,
          noDocumentAll: noDocumentAll,
          innerBinding: innerBinding
        }, file);
        var keysNodes = undefined,
            staticNodes = undefined,
            instanceNodes = undefined,
            lastInstanceNodeReturnsThis = undefined,
            pureStaticNodes = undefined,
            classBindingNode = undefined,
            wrapClass = undefined;
        {
          if (isDecorated) {
            staticNodes = pureStaticNodes = keysNodes = [];

            var _ref3 = (0, _decorators2.buildDecoratedClass)(classRefForDefine, path, elements, file);

            instanceNodes = _ref3.instanceNodes;
            wrapClass = _ref3.wrapClass;
          } else {
            keysNodes = (0, _misc.extractComputedKeys)(path, computedPaths, file);

            var _ref4 = (0, _fields.buildFieldsInitNodes)(ref, path.node.superClass, props, privateNamesMap, file, setPublicClassFields != null ? setPublicClassFields : loose, privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose, noUninitializedPrivateFieldAccess, constantSuper != null ? constantSuper : loose, innerBinding);

            staticNodes = _ref4.staticNodes;
            pureStaticNodes = _ref4.pureStaticNodes;
            instanceNodes = _ref4.instanceNodes;
            lastInstanceNodeReturnsThis = _ref4.lastInstanceNodeReturnsThis;
            classBindingNode = _ref4.classBindingNode;
            wrapClass = _ref4.wrapClass;
          }
        }
        if (instanceNodes.length > 0) {
          (0, _misc.injectInitialization)(path, constructor, instanceNodes, function (referenceVisitor, state) {
            {
              if (isDecorated) return;
            }
            var _iteratorNormalCompletion2 = true;
            var _didIteratorError2 = false;
            var _iteratorError2 = undefined;

            try {
              for (var _iterator2 = props[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                var prop = _step2.value;

                if (_core.types.isStaticBlock != null && _core.types.isStaticBlock(prop.node) || prop.node["static"]) continue;
                prop.traverse(referenceVisitor, state);
              }
            } catch (err) {
              _didIteratorError2 = true;
              _iteratorError2 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
                  _iterator2["return"]();
                }
              } finally {
                if (_didIteratorError2) {
                  throw _iteratorError2;
                }
              }
            }
          }, lastInstanceNodeReturnsThis);
        }
        var wrappedPath = wrapClass(path);
        wrappedPath.insertBefore([].concat(_toConsumableArray(privateNamesNodes), _toConsumableArray(keysNodes)));
        if (staticNodes.length > 0) {
          wrappedPath.insertAfter(staticNodes);
        }
        if (pureStaticNodes.length > 0) {
          wrappedPath.find(function (parent) {
            return parent.isStatement() || parent.isDeclaration();
          }).insertAfter(pureStaticNodes);
        }
        if (classBindingNode != null && pathIsClassDeclaration) {
          wrappedPath.insertAfter(classBindingNode);
        }
      },
      ExportDefaultDeclaration: function ExportDefaultDeclaration(path, _ref5) {
        var file = _ref5.file;

        {
          if (file.get(versionKey) !== "7.27.1") return;
          var decl = path.get("declaration");
          if (decl.isClassDeclaration() && (0, _decorators.hasDecorators)(decl.node)) {
            if (decl.node.id) {
              {
                var _path$splitExportDecl;
                (_path$splitExportDecl = path.splitExportDeclaration) != null ? _path$splitExportDecl : path.splitExportDeclaration = require("@babel/traverse").NodePath.prototype.splitExportDeclaration;
              }
              path.splitExportDeclaration();
            } else {
              decl.node.type = "ClassExpression";
            }
          }
        }
      }
    }
  };
}
