// @babel/core/lib/config/caching.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var marked0$0 = [genTrue, getCachedValue, getCachedValueOrWait].map(regeneratorRuntime.mark);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertSimpleType = assertSimpleType;
exports.makeStrongCache = makeStrongCache;
exports.makeStrongCacheSync = makeStrongCacheSync;
exports.makeWeakCache = makeWeakCache;
exports.makeWeakCacheSync = makeWeakCacheSync;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _async = require("../gensync-utils/async.js");
var _util = require("./util.js");
var synchronize = function synchronize(gen) {
  return _gensync()(gen).sync;
};
function genTrue() {
  return regeneratorRuntime.wrap(function genTrue$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", true);

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function makeWeakCache(handler) {
  return makeCachedFunction(WeakMap, handler);
}
function makeWeakCacheSync(handler) {
  return synchronize(makeWeakCache(handler));
}
function makeStrongCache(handler) {
  return makeCachedFunction(Map, handler);
}
function makeStrongCacheSync(handler) {
  return synchronize(makeStrongCache(handler));
}
function makeCachedFunction(CallCache, handler) {
  var callCacheSync = new CallCache();
  var callCacheAsync = new CallCache();
  var futureCache = new CallCache();
  return regeneratorRuntime.mark(function cachedFunction(arg, data) {
    var asyncContext, callCache, cached, cache, handlerResult, finishLock, value;
    return regeneratorRuntime.wrap(function cachedFunction$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          return context$2$0.delegateYield((0, _async.isAsync)(), "t0", 1);

        case 1:
          asyncContext = context$2$0.t0;
          callCache = asyncContext ? callCacheAsync : callCacheSync;
          return context$2$0.delegateYield(getCachedValueOrWait(asyncContext, callCache, futureCache, arg, data), "t1", 4);

        case 4:
          cached = context$2$0.t1;

          if (!cached.valid) {
            context$2$0.next = 7;
            break;
          }

          return context$2$0.abrupt("return", cached.value);

        case 7:
          cache = new CacheConfigurator(data);
          handlerResult = handler(arg, cache);
          finishLock = undefined;
          value = undefined;

          if (!(0, _util.isIterableIterator)(handlerResult)) {
            context$2$0.next = 16;
            break;
          }

          return context$2$0.delegateYield((0, _async.onFirstPause)(handlerResult, function () {
            finishLock = setupAsyncLocks(cache, futureCache, arg);
          }), "t2", 13);

        case 13:
          value = context$2$0.t2;
          context$2$0.next = 17;
          break;

        case 16:
          value = handlerResult;

        case 17:
          updateFunctionCache(callCache, cache, arg, value);
          if (finishLock) {
            futureCache["delete"](arg);
            finishLock.release(value);
          }
          return context$2$0.abrupt("return", value);

        case 20:
        case "end":
          return context$2$0.stop();
      }
    }, cachedFunction, this);
  });
}
function getCachedValue(cache, arg, data) {
  var cachedValue, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _step$value, value, valid;

  return regeneratorRuntime.wrap(function getCachedValue$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        cachedValue = cache.get(arg);

        if (!cachedValue) {
          context$1$0.next = 31;
          break;
        }

        _iteratorNormalCompletion = true;
        _didIteratorError = false;
        _iteratorError = undefined;
        context$1$0.prev = 5;
        _iterator = cachedValue[Symbol.iterator]();

      case 7:
        if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
          context$1$0.next = 17;
          break;
        }

        _step$value = _step.value;
        value = _step$value.value;
        valid = _step$value.valid;
        return context$1$0.delegateYield(valid(data), "t0", 12);

      case 12:
        if (!context$1$0.t0) {
          context$1$0.next = 14;
          break;
        }

        return context$1$0.abrupt("return", {
          valid: true,
          value: value
        });

      case 14:
        _iteratorNormalCompletion = true;
        context$1$0.next = 7;
        break;

      case 17:
        context$1$0.next = 23;
        break;

      case 19:
        context$1$0.prev = 19;
        context$1$0.t1 = context$1$0["catch"](5);
        _didIteratorError = true;
        _iteratorError = context$1$0.t1;

      case 23:
        context$1$0.prev = 23;
        context$1$0.prev = 24;

        if (!_iteratorNormalCompletion && _iterator["return"]) {
          _iterator["return"]();
        }

      case 26:
        context$1$0.prev = 26;

        if (!_didIteratorError) {
          context$1$0.next = 29;
          break;
        }

        throw _iteratorError;

      case 29:
        return context$1$0.finish(26);

      case 30:
        return context$1$0.finish(23);

      case 31:
        return context$1$0.abrupt("return", {
          valid: false,
          value: null
        });

      case 32:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this, [[5, 19, 23, 31], [24,, 26, 30]]);
}
function getCachedValueOrWait(asyncContext, callCache, futureCache, arg, data) {
  var cached, _cached, value;

  return regeneratorRuntime.wrap(function getCachedValueOrWait$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(getCachedValue(callCache, arg, data), "t0", 1);

      case 1:
        cached = context$1$0.t0;

        if (!cached.valid) {
          context$1$0.next = 4;
          break;
        }

        return context$1$0.abrupt("return", cached);

      case 4:
        if (!asyncContext) {
          context$1$0.next = 11;
          break;
        }

        return context$1$0.delegateYield(getCachedValue(futureCache, arg, data), "t1", 6);

      case 6:
        _cached = context$1$0.t1;

        if (!_cached.valid) {
          context$1$0.next = 11;
          break;
        }

        return context$1$0.delegateYield((0, _async.waitFor)(_cached.value.promise), "t2", 9);

      case 9:
        value = context$1$0.t2;
        return context$1$0.abrupt("return", {
          valid: true,
          value: value
        });

      case 11:
        return context$1$0.abrupt("return", {
          valid: false,
          value: null
        });

      case 12:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
function setupAsyncLocks(config, futureCache, arg) {
  var finishLock = new Lock();
  updateFunctionCache(futureCache, config, arg, finishLock);
  return finishLock;
}
function updateFunctionCache(cache, config, arg, value) {
  if (!config.configured()) config.forever();
  var cachedValue = cache.get(arg);
  config.deactivate();
  switch (config.mode()) {
    case "forever":
      cachedValue = [{
        value: value,
        valid: genTrue
      }];
      cache.set(arg, cachedValue);
      break;
    case "invalidate":
      cachedValue = [{
        value: value,
        valid: config.validator()
      }];
      cache.set(arg, cachedValue);
      break;
    case "valid":
      if (cachedValue) {
        cachedValue.push({
          value: value,
          valid: config.validator()
        });
      } else {
        cachedValue = [{
          value: value,
          valid: config.validator()
        }];
        cache.set(arg, cachedValue);
      }
  }
}

var CacheConfigurator = (function () {
  function CacheConfigurator(data) {
    _classCallCheck(this, CacheConfigurator);

    this._active = true;
    this._never = false;
    this._forever = false;
    this._invalidate = false;
    this._configured = false;
    this._pairs = [];
    this._data = void 0;
    this._data = data;
  }

  _createClass(CacheConfigurator, [{
    key: "simple",
    value: function simple() {
      return makeSimpleConfigurator(this);
    }
  }, {
    key: "mode",
    value: function mode() {
      if (this._never) return "never";
      if (this._forever) return "forever";
      if (this._invalidate) return "invalidate";
      return "valid";
    }
  }, {
    key: "forever",
    value: function forever() {
      if (!this._active) {
        throw new Error("Cannot change caching after evaluation has completed.");
      }
      if (this._never) {
        throw new Error("Caching has already been configured with .never()");
      }
      this._forever = true;
      this._configured = true;
    }
  }, {
    key: "never",
    value: function never() {
      if (!this._active) {
        throw new Error("Cannot change caching after evaluation has completed.");
      }
      if (this._forever) {
        throw new Error("Caching has already been configured with .forever()");
      }
      this._never = true;
      this._configured = true;
    }
  }, {
    key: "using",
    value: function using(handler) {
      var _this = this;

      if (!this._active) {
        throw new Error("Cannot change caching after evaluation has completed.");
      }
      if (this._never || this._forever) {
        throw new Error("Caching has already been configured with .never or .forever()");
      }
      this._configured = true;
      var key = handler(this._data);
      var fn = (0, _async.maybeAsync)(handler, "You appear to be using an async cache handler, but Babel has been called synchronously");
      if ((0, _async.isThenable)(key)) {
        return key.then(function (key) {
          _this._pairs.push([key, fn]);
          return key;
        });
      }
      this._pairs.push([key, fn]);
      return key;
    }
  }, {
    key: "invalidate",
    value: function invalidate(handler) {
      this._invalidate = true;
      return this.using(handler);
    }
  }, {
    key: "validator",
    value: function validator() {
      var pairs = this._pairs;
      return regeneratorRuntime.mark(function callee$2$0(data) {
        var _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _step2$value, key, fn;

        return regeneratorRuntime.wrap(function callee$2$0$(context$3$0) {
          while (1) switch (context$3$0.prev = context$3$0.next) {
            case 0:
              _iteratorNormalCompletion2 = true;
              _didIteratorError2 = false;
              _iteratorError2 = undefined;
              context$3$0.prev = 3;
              _iterator2 = pairs[Symbol.iterator]();

            case 5:
              if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
                context$3$0.next = 17;
                break;
              }

              _step2$value = _slicedToArray(_step2.value, 2);
              key = _step2$value[0];
              fn = _step2$value[1];
              context$3$0.t0 = key;
              return context$3$0.delegateYield(fn(data), "t1", 11);

            case 11:
              context$3$0.t2 = context$3$0.t1;

              if (!(context$3$0.t0 !== context$3$0.t2)) {
                context$3$0.next = 14;
                break;
              }

              return context$3$0.abrupt("return", false);

            case 14:
              _iteratorNormalCompletion2 = true;
              context$3$0.next = 5;
              break;

            case 17:
              context$3$0.next = 23;
              break;

            case 19:
              context$3$0.prev = 19;
              context$3$0.t3 = context$3$0["catch"](3);
              _didIteratorError2 = true;
              _iteratorError2 = context$3$0.t3;

            case 23:
              context$3$0.prev = 23;
              context$3$0.prev = 24;

              if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
                _iterator2["return"]();
              }

            case 26:
              context$3$0.prev = 26;

              if (!_didIteratorError2) {
                context$3$0.next = 29;
                break;
              }

              throw _iteratorError2;

            case 29:
              return context$3$0.finish(26);

            case 30:
              return context$3$0.finish(23);

            case 31:
              return context$3$0.abrupt("return", true);

            case 32:
            case "end":
              return context$3$0.stop();
          }
        }, callee$2$0, this, [[3, 19, 23, 31], [24,, 26, 30]]);
      });
    }
  }, {
    key: "deactivate",
    value: function deactivate() {
      this._active = false;
    }
  }, {
    key: "configured",
    value: function configured() {
      return this._configured;
    }
  }]);

  return CacheConfigurator;
})();

function makeSimpleConfigurator(cache) {
  function cacheFn(val) {
    if (typeof val === "boolean") {
      if (val) cache.forever();else cache.never();
      return;
    }
    return cache.using(function () {
      return assertSimpleType(val());
    });
  }
  cacheFn.forever = function () {
    return cache.forever();
  };
  cacheFn.never = function () {
    return cache.never();
  };
  cacheFn.using = function (cb) {
    return cache.using(function () {
      return assertSimpleType(cb());
    });
  };
  cacheFn.invalidate = function (cb) {
    return cache.invalidate(function () {
      return assertSimpleType(cb());
    });
  };
  return cacheFn;
}
function assertSimpleType(value) {
  if ((0, _async.isThenable)(value)) {
    throw new Error("You appear to be using an async cache handler, " + "which your current version of Babel does not support. " + "We may add support for this in the future, " + "but if you're on the most recent version of @babel/core and still " + "seeing this error, then you'll need to synchronously handle your caching logic.");
  }
  if (value != null && typeof value !== "string" && typeof value !== "boolean" && typeof value !== "number") {
    throw new Error("Cache keys must be either string, boolean, number, null, or undefined.");
  }
  return value;
}

var Lock = (function () {
  function Lock() {
    var _this2 = this;

    _classCallCheck(this, Lock);

    this.released = false;
    this.promise = void 0;
    this._resolve = void 0;
    this.promise = new Promise(function (resolve) {
      _this2._resolve = resolve;
    });
  }

  _createClass(Lock, [{
    key: "release",
    value: function release(value) {
      this.released = true;
      this._resolve(value);
    }
  }]);

  return Lock;
})();

0 && 0;
