// @babel/core/lib/config/config-descriptors.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var marked0$0 = [handlerOf, createPresetDescriptors, createPluginDescriptors, createDescriptors, createDescriptor].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCachedDescriptors = createCachedDescriptors;
exports.createDescriptor = createDescriptor;
exports.createUncachedDescriptors = createUncachedDescriptors;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _functional = require("../gensync-utils/functional.js");
var _index = require("./files/index.js");
var _item = require("./item.js");
var _caching = require("./caching.js");
var _resolveTargets = require("./resolve-targets.js");
function isEqualDescriptor(a, b) {
  var _a$file, _b$file, _a$file2, _b$file2;
  return a.name === b.name && a.value === b.value && a.options === b.options && a.dirname === b.dirname && a.alias === b.alias && a.ownPass === b.ownPass && ((_a$file = a.file) == null ? void 0 : _a$file.request) === ((_b$file = b.file) == null ? void 0 : _b$file.request) && ((_a$file2 = a.file) == null ? void 0 : _a$file2.resolved) === ((_b$file2 = b.file) == null ? void 0 : _b$file2.resolved);
}
function handlerOf(value) {
  return regeneratorRuntime.wrap(function handlerOf$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", value);

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function optionsWithResolvedBrowserslistConfigFile(options, dirname) {
  if (typeof options.browserslistConfigFile === "string") {
    options.browserslistConfigFile = (0, _resolveTargets.resolveBrowserslistConfigFile)(options.browserslistConfigFile, dirname);
  }
  return options;
}
function createCachedDescriptors(dirname, options, alias) {
  var plugins = options.plugins;
  var presets = options.presets;
  var passPerPreset = options.passPerPreset;

  return {
    options: optionsWithResolvedBrowserslistConfigFile(options, dirname),
    plugins: plugins ? function () {
      return createCachedPluginDescriptors(plugins, dirname)(alias);
    } : function () {
      return handlerOf([]);
    },
    presets: presets ? function () {
      return createCachedPresetDescriptors(presets, dirname)(alias)(!!passPerPreset);
    } : function () {
      return handlerOf([]);
    }
  };
}
function createUncachedDescriptors(dirname, options, alias) {
  return {
    options: optionsWithResolvedBrowserslistConfigFile(options, dirname),
    plugins: (0, _functional.once)(function () {
      return createPluginDescriptors(options.plugins || [], dirname, alias);
    }),
    presets: (0, _functional.once)(function () {
      return createPresetDescriptors(options.presets || [], dirname, alias, !!options.passPerPreset);
    })
  };
}
var PRESET_DESCRIPTOR_CACHE = new WeakMap();
var createCachedPresetDescriptors = (0, _caching.makeWeakCacheSync)(function (items, cache) {
  var dirname = cache.using(function (dir) {
    return dir;
  });
  return (0, _caching.makeStrongCacheSync)(function (alias) {
    return (0, _caching.makeStrongCache)(regeneratorRuntime.mark(function callee$2$0(passPerPreset) {
      var descriptors;
      return regeneratorRuntime.wrap(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            return context$3$0.delegateYield(createPresetDescriptors(items, dirname, alias, passPerPreset), "t0", 1);

          case 1:
            descriptors = context$3$0.t0;
            return context$3$0.abrupt("return", descriptors.map(function (desc) {
              return loadCachedDescriptor(PRESET_DESCRIPTOR_CACHE, desc);
            }));

          case 3:
          case "end":
            return context$3$0.stop();
        }
      }, callee$2$0, this);
    }));
  });
});
var PLUGIN_DESCRIPTOR_CACHE = new WeakMap();
var createCachedPluginDescriptors = (0, _caching.makeWeakCacheSync)(function (items, cache) {
  var dirname = cache.using(function (dir) {
    return dir;
  });
  return (0, _caching.makeStrongCache)(regeneratorRuntime.mark(function callee$1$0(alias) {
    var descriptors;
    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          return context$2$0.delegateYield(createPluginDescriptors(items, dirname, alias), "t0", 1);

        case 1:
          descriptors = context$2$0.t0;
          return context$2$0.abrupt("return", descriptors.map(function (desc) {
            return loadCachedDescriptor(PLUGIN_DESCRIPTOR_CACHE, desc);
          }));

        case 3:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this);
  }));
});
var DEFAULT_OPTIONS = {};
function loadCachedDescriptor(cache, desc) {
  var value = desc.value;
  var _desc$options = desc.options;
  var options = _desc$options === undefined ? DEFAULT_OPTIONS : _desc$options;

  if (options === false) return desc;
  var cacheByOptions = cache.get(value);
  if (!cacheByOptions) {
    cacheByOptions = new WeakMap();
    cache.set(value, cacheByOptions);
  }
  var possibilities = cacheByOptions.get(options);
  if (!possibilities) {
    possibilities = [];
    cacheByOptions.set(options, possibilities);
  }
  if (!possibilities.includes(desc)) {
    var matches = possibilities.filter(function (possibility) {
      return isEqualDescriptor(possibility, desc);
    });
    if (matches.length > 0) {
      return matches[0];
    }
    possibilities.push(desc);
  }
  return desc;
}
function createPresetDescriptors(items, dirname, alias, passPerPreset) {
  return regeneratorRuntime.wrap(function createPresetDescriptors$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(createDescriptors("preset", items, dirname, alias, passPerPreset), "t0", 1);

      case 1:
        return context$1$0.abrupt("return", context$1$0.t0);

      case 2:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}
function createPluginDescriptors(items, dirname, alias) {
  return regeneratorRuntime.wrap(function createPluginDescriptors$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(createDescriptors("plugin", items, dirname, alias), "t0", 1);

      case 1:
        return context$1$0.abrupt("return", context$1$0.t0);

      case 2:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
function createDescriptors(type, items, dirname, alias, ownPass) {
  var descriptors;
  return regeneratorRuntime.wrap(function createDescriptors$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(_gensync().all(items.map(function (item, index) {
          return createDescriptor(item, dirname, {
            type: type,
            alias: alias + "$" + index,
            ownPass: !!ownPass
          });
        })), "t0", 1);

      case 1:
        descriptors = context$1$0.t0;

        assertNoDuplicates(descriptors);
        return context$1$0.abrupt("return", descriptors);

      case 4:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[3], this);
}
function createDescriptor(pair, dirname, _ref) {
  var type = _ref.type;
  var alias = _ref.alias;
  var ownPass = _ref.ownPass;

  var desc, name, options, value, _value, _value2, _value3, _value32, file, filepath, resolver, request, _ref2;

  return regeneratorRuntime.wrap(function createDescriptor$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        desc = (0, _item.getItemDescriptor)(pair);

        if (!desc) {
          context$1$0.next = 3;
          break;
        }

        return context$1$0.abrupt("return", desc);

      case 3:
        name = undefined;
        options = undefined;
        value = pair;

        if (Array.isArray(value)) {
          if (value.length === 3) {
            _value = value;
            _value2 = _slicedToArray(_value, 3);
            value = _value2[0];
            options = _value2[1];
            name = _value2[2];
          } else {
            _value3 = value;
            _value32 = _slicedToArray(_value3, 2);
            value = _value32[0];
            options = _value32[1];
          }
        }
        file = undefined;
        filepath = null;

        if (!(typeof value === "string")) {
          context$1$0.next = 19;
          break;
        }

        if (!(typeof type !== "string")) {
          context$1$0.next = 12;
          break;
        }

        throw new Error("To resolve a string-based item, the type of item must be given");

      case 12:
        resolver = type === "plugin" ? _index.loadPlugin : _index.loadPreset;
        request = value;
        return context$1$0.delegateYield(resolver(value, dirname), "t0", 15);

      case 15:
        _ref2 = context$1$0.t0;
        filepath = _ref2.filepath;
        value = _ref2.value;

        file = {
          request: request,
          resolved: filepath
        };

      case 19:
        if (value) {
          context$1$0.next = 21;
          break;
        }

        throw new Error("Unexpected falsy value: " + String(value));

      case 21:
        if (!(typeof value === "object" && value.__esModule)) {
          context$1$0.next = 27;
          break;
        }

        if (!value["default"]) {
          context$1$0.next = 26;
          break;
        }

        value = value["default"];
        context$1$0.next = 27;
        break;

      case 26:
        throw new Error("Must export a default export when using ES6 modules.");

      case 27:
        if (!(typeof value !== "object" && typeof value !== "function")) {
          context$1$0.next = 29;
          break;
        }

        throw new Error("Unsupported format: " + typeof value + ". Expected an object or a function.");

      case 29:
        if (!(filepath !== null && typeof value === "object" && value)) {
          context$1$0.next = 31;
          break;
        }

        throw new Error("Plugin/Preset files are not allowed to export objects, only functions. In " + filepath);

      case 31:
        return context$1$0.abrupt("return", {
          name: name,
          alias: filepath || alias,
          value: value,
          options: options,
          dirname: dirname,
          ownPass: ownPass,
          file: file
        });

      case 32:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[4], this);
}
function assertNoDuplicates(items) {
  var map = new Map();
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    var _loop = function () {
      var item = _step.value;

      if (typeof item.value !== "function") return "continue";
      var nameMap = map.get(item.value);
      if (!nameMap) {
        nameMap = new Set();
        map.set(item.value, nameMap);
      }
      if (nameMap.has(item.name)) {
        var conflicts = items.filter(function (i) {
          return i.value === item.value;
        });
        throw new Error(["Duplicate plugin/preset detected.", "If you'd like to use two separate instances of a plugin,", "they need separate names, e.g.", "", "  plugins: [", "    ['some-plugin', {}],", "    ['some-plugin', {}, 'some unique name'],", "  ]", "", "Duplicates detected are:", "" + JSON.stringify(conflicts, null, 2)].join("\n"));
      }
      nameMap.add(item.name);
    };

    for (var _iterator = items[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var _ret = _loop();

      if (_ret === "continue") continue;
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
}
0 && 0;
