// @babel/core/lib/config/config-chain.js
"use strict";

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

var marked0$0 = [buildPresetChain, buildRootChain, loadFileChain, mergeExtendsChain, mergeChainOpts].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildPresetChain = buildPresetChain;
exports.buildPresetChainWalker = void 0;
exports.buildRootChain = buildRootChain;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _debug() {
  var data = require("debug");
  _debug = function () {
    return data;
  };
  return data;
}
var _options = require("./validation/options.js");
var _patternToRegex = require("./pattern-to-regex.js");
var _printer = require("./printer.js");
var _rewriteStackTrace = require("../errors/rewrite-stack-trace.js");
var _configError = require("../errors/config-error.js");
var _index = require("./files/index.js");
var _caching = require("./caching.js");
var _configDescriptors = require("./config-descriptors.js");
var debug = _debug()("babel:config:config-chain");
function buildPresetChain(arg, context) {
  var chain;
  return regeneratorRuntime.wrap(function buildPresetChain$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(buildPresetChainWalker(arg, context), "t0", 1);

      case 1:
        chain = context$1$0.t0;

        if (chain) {
          context$1$0.next = 4;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 4:
        return context$1$0.abrupt("return", {
          plugins: dedupDescriptors(chain.plugins),
          presets: dedupDescriptors(chain.presets),
          options: chain.options.map(function (o) {
            return normalizeOptions(o);
          }),
          files: new Set()
        });

      case 5:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
var buildPresetChainWalker = exports.buildPresetChainWalker = makeChainWalker({
  root: function root(preset) {
    return loadPresetDescriptors(preset);
  },
  env: function env(preset, envName) {
    return loadPresetEnvDescriptors(preset)(envName);
  },
  overrides: function overrides(preset, index) {
    return loadPresetOverridesDescriptors(preset)(index);
  },
  overridesEnv: function overridesEnv(preset, index, envName) {
    return loadPresetOverridesEnvDescriptors(preset)(index)(envName);
  },
  createLogger: function createLogger() {
    return function () {};
  }
});
var loadPresetDescriptors = (0, _caching.makeWeakCacheSync)(function (preset) {
  return buildRootDescriptors(preset, preset.alias, _configDescriptors.createUncachedDescriptors);
});
var loadPresetEnvDescriptors = (0, _caching.makeWeakCacheSync)(function (preset) {
  return (0, _caching.makeStrongCacheSync)(function (envName) {
    return buildEnvDescriptors(preset, preset.alias, _configDescriptors.createUncachedDescriptors, envName);
  });
});
var loadPresetOverridesDescriptors = (0, _caching.makeWeakCacheSync)(function (preset) {
  return (0, _caching.makeStrongCacheSync)(function (index) {
    return buildOverrideDescriptors(preset, preset.alias, _configDescriptors.createUncachedDescriptors, index);
  });
});
var loadPresetOverridesEnvDescriptors = (0, _caching.makeWeakCacheSync)(function (preset) {
  return (0, _caching.makeStrongCacheSync)(function (index) {
    return (0, _caching.makeStrongCacheSync)(function (envName) {
      return buildOverrideEnvDescriptors(preset, preset.alias, _configDescriptors.createUncachedDescriptors, index, envName);
    });
  });
});
function buildRootChain(opts, context) {
  var configReport, babelRcReport, programmaticLogger, programmaticChain, programmaticReport, configFile, babelrc, babelrcRoots, babelrcRootsDirectory, configFileChain, configFileLogger, validatedFile, result, ignoreFile, babelrcFile, isIgnored, fileChain, pkgData, _ref, babelrcLogger, chain;

  return regeneratorRuntime.wrap(function buildRootChain$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        configReport = undefined, babelRcReport = undefined;
        programmaticLogger = new _printer.ConfigPrinter();
        return context$1$0.delegateYield(loadProgrammaticChain({
          options: opts,
          dirname: context.cwd
        }, context, undefined, programmaticLogger), "t0", 3);

      case 3:
        programmaticChain = context$1$0.t0;

        if (programmaticChain) {
          context$1$0.next = 6;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 6:
        return context$1$0.delegateYield(programmaticLogger.output(), "t1", 7);

      case 7:
        programmaticReport = context$1$0.t1;
        configFile = undefined;

        if (!(typeof opts.configFile === "string")) {
          context$1$0.next = 14;
          break;
        }

        return context$1$0.delegateYield((0, _index.loadConfig)(opts.configFile, context.cwd, context.envName, context.caller), "t2", 11);

      case 11:
        configFile = context$1$0.t2;
        context$1$0.next = 17;
        break;

      case 14:
        if (!(opts.configFile !== false)) {
          context$1$0.next = 17;
          break;
        }

        return context$1$0.delegateYield((0, _index.findRootConfig)(context.root, context.envName, context.caller), "t3", 16);

      case 16:
        configFile = context$1$0.t3;

      case 17:
        babelrc = opts.babelrc;
        babelrcRoots = opts.babelrcRoots;
        babelrcRootsDirectory = context.cwd;
        configFileChain = emptyChain();
        configFileLogger = new _printer.ConfigPrinter();

        if (!configFile) {
          context$1$0.next = 33;
          break;
        }

        validatedFile = validateConfigFile(configFile);
        return context$1$0.delegateYield(loadFileChain(validatedFile, context, undefined, configFileLogger), "t4", 25);

      case 25:
        result = context$1$0.t4;

        if (result) {
          context$1$0.next = 28;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 28:
        return context$1$0.delegateYield(configFileLogger.output(), "t5", 29);

      case 29:
        configReport = context$1$0.t5;

        if (babelrc === undefined) {
          babelrc = validatedFile.options.babelrc;
        }
        if (babelrcRoots === undefined) {
          babelrcRootsDirectory = validatedFile.dirname;
          babelrcRoots = validatedFile.options.babelrcRoots;
        }
        mergeChain(configFileChain, result);

      case 33:
        ignoreFile = undefined, babelrcFile = undefined;
        isIgnored = false;
        fileChain = emptyChain();

        if (!((babelrc === true || babelrc === undefined) && typeof context.filename === "string")) {
          context$1$0.next = 59;
          break;
        }

        return context$1$0.delegateYield((0, _index.findPackageData)(context.filename), "t6", 38);

      case 38:
        pkgData = context$1$0.t6;

        if (!(pkgData && babelrcLoadEnabled(context, pkgData, babelrcRoots, babelrcRootsDirectory))) {
          context$1$0.next = 59;
          break;
        }

        return context$1$0.delegateYield((0, _index.findRelativeConfig)(pkgData, context.envName, context.caller), "t7", 41);

      case 41:
        _ref = context$1$0.t7;
        ignoreFile = _ref.ignore;
        babelrcFile = _ref.config;

        if (ignoreFile) {
          fileChain.files.add(ignoreFile.filepath);
        }
        if (ignoreFile && shouldIgnore(context, ignoreFile.ignore, null, ignoreFile.dirname)) {
          isIgnored = true;
        }

        if (!(babelrcFile && !isIgnored)) {
          context$1$0.next = 58;
          break;
        }

        validatedFile = validateBabelrcFile(babelrcFile);
        babelrcLogger = new _printer.ConfigPrinter();
        return context$1$0.delegateYield(loadFileChain(validatedFile, context, undefined, babelrcLogger), "t8", 50);

      case 50:
        result = context$1$0.t8;

        if (result) {
          context$1$0.next = 55;
          break;
        }

        isIgnored = true;
        context$1$0.next = 58;
        break;

      case 55:
        return context$1$0.delegateYield(babelrcLogger.output(), "t9", 56);

      case 56:
        babelRcReport = context$1$0.t9;

        mergeChain(fileChain, result);

      case 58:
        if (babelrcFile && isIgnored) {
          fileChain.files.add(babelrcFile.filepath);
        }

      case 59:
        if (context.showConfig) {
          console.log("Babel configs on \"" + context.filename + "\" (ascending priority):\n" + [configReport, babelRcReport, programmaticReport].filter(function (x) {
            return !!x;
          }).join("\n\n") + "\n-----End Babel configs-----");
        }
        chain = mergeChain(mergeChain(mergeChain(emptyChain(), configFileChain), fileChain), programmaticChain);
        return context$1$0.abrupt("return", {
          plugins: isIgnored ? [] : dedupDescriptors(chain.plugins),
          presets: isIgnored ? [] : dedupDescriptors(chain.presets),
          options: isIgnored ? [] : chain.options.map(function (o) {
            return normalizeOptions(o);
          }),
          fileHandling: isIgnored ? "ignored" : "transpile",
          ignore: ignoreFile || undefined,
          babelrc: babelrcFile || undefined,
          config: configFile || undefined,
          files: chain.files
        });

      case 62:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}
function babelrcLoadEnabled(context, pkgData, babelrcRoots, babelrcRootsDirectory) {
  if (typeof babelrcRoots === "boolean") return babelrcRoots;
  var absoluteRoot = context.root;
  if (babelrcRoots === undefined) {
    return pkgData.directories.includes(absoluteRoot);
  }
  var babelrcPatterns = babelrcRoots;
  if (!Array.isArray(babelrcPatterns)) {
    babelrcPatterns = [babelrcPatterns];
  }
  babelrcPatterns = babelrcPatterns.map(function (pat) {
    return typeof pat === "string" ? _path().resolve(babelrcRootsDirectory, pat) : pat;
  });
  if (babelrcPatterns.length === 1 && babelrcPatterns[0] === absoluteRoot) {
    return pkgData.directories.includes(absoluteRoot);
  }
  return babelrcPatterns.some(function (pat) {
    if (typeof pat === "string") {
      pat = (0, _patternToRegex["default"])(pat, babelrcRootsDirectory);
    }
    return pkgData.directories.some(function (directory) {
      return matchPattern(pat, babelrcRootsDirectory, directory, context);
    });
  });
}
var validateConfigFile = (0, _caching.makeWeakCacheSync)(function (file) {
  return {
    filepath: file.filepath,
    dirname: file.dirname,
    options: (0, _options.validate)("configfile", file.options, file.filepath)
  };
});
var validateBabelrcFile = (0, _caching.makeWeakCacheSync)(function (file) {
  return {
    filepath: file.filepath,
    dirname: file.dirname,
    options: (0, _options.validate)("babelrcfile", file.options, file.filepath)
  };
});
var validateExtendFile = (0, _caching.makeWeakCacheSync)(function (file) {
  return {
    filepath: file.filepath,
    dirname: file.dirname,
    options: (0, _options.validate)("extendsfile", file.options, file.filepath)
  };
});
var loadProgrammaticChain = makeChainWalker({
  root: function root(input) {
    return buildRootDescriptors(input, "base", _configDescriptors.createCachedDescriptors);
  },
  env: function env(input, envName) {
    return buildEnvDescriptors(input, "base", _configDescriptors.createCachedDescriptors, envName);
  },
  overrides: function overrides(input, index) {
    return buildOverrideDescriptors(input, "base", _configDescriptors.createCachedDescriptors, index);
  },
  overridesEnv: function overridesEnv(input, index, envName) {
    return buildOverrideEnvDescriptors(input, "base", _configDescriptors.createCachedDescriptors, index, envName);
  },
  createLogger: function createLogger(input, context, baseLogger) {
    return buildProgrammaticLogger(input, context, baseLogger);
  }
});
var loadFileChainWalker = makeChainWalker({
  root: function root(file) {
    return loadFileDescriptors(file);
  },
  env: function env(file, envName) {
    return loadFileEnvDescriptors(file)(envName);
  },
  overrides: function overrides(file, index) {
    return loadFileOverridesDescriptors(file)(index);
  },
  overridesEnv: function overridesEnv(file, index, envName) {
    return loadFileOverridesEnvDescriptors(file)(index)(envName);
  },
  createLogger: function createLogger(file, context, baseLogger) {
    return buildFileLogger(file.filepath, context, baseLogger);
  }
});
function loadFileChain(input, context, files, baseLogger) {
  var chain;
  return regeneratorRuntime.wrap(function loadFileChain$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(loadFileChainWalker(input, context, files, baseLogger), "t0", 1);

      case 1:
        chain = context$1$0.t0;

        chain == null || chain.files.add(input.filepath);
        return context$1$0.abrupt("return", chain);

      case 4:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
var loadFileDescriptors = (0, _caching.makeWeakCacheSync)(function (file) {
  return buildRootDescriptors(file, file.filepath, _configDescriptors.createUncachedDescriptors);
});
var loadFileEnvDescriptors = (0, _caching.makeWeakCacheSync)(function (file) {
  return (0, _caching.makeStrongCacheSync)(function (envName) {
    return buildEnvDescriptors(file, file.filepath, _configDescriptors.createUncachedDescriptors, envName);
  });
});
var loadFileOverridesDescriptors = (0, _caching.makeWeakCacheSync)(function (file) {
  return (0, _caching.makeStrongCacheSync)(function (index) {
    return buildOverrideDescriptors(file, file.filepath, _configDescriptors.createUncachedDescriptors, index);
  });
});
var loadFileOverridesEnvDescriptors = (0, _caching.makeWeakCacheSync)(function (file) {
  return (0, _caching.makeStrongCacheSync)(function (index) {
    return (0, _caching.makeStrongCacheSync)(function (envName) {
      return buildOverrideEnvDescriptors(file, file.filepath, _configDescriptors.createUncachedDescriptors, index, envName);
    });
  });
});
function buildFileLogger(filepath, context, baseLogger) {
  if (!baseLogger) {
    return function () {};
  }
  return baseLogger.configure(context.showConfig, _printer.ChainFormatter.Config, {
    filepath: filepath
  });
}
function buildRootDescriptors(_ref2, alias, descriptors) {
  var dirname = _ref2.dirname;
  var options = _ref2.options;

  return descriptors(dirname, options, alias);
}
function buildProgrammaticLogger(_, context, baseLogger) {
  var _context$caller;
  if (!baseLogger) {
    return function () {};
  }
  return baseLogger.configure(context.showConfig, _printer.ChainFormatter.Programmatic, {
    callerName: (_context$caller = context.caller) == null ? void 0 : _context$caller.name
  });
}
function buildEnvDescriptors(_ref3, alias, descriptors, envName) {
  var dirname = _ref3.dirname;
  var options = _ref3.options;

  var _options$env;
  var opts = (_options$env = options.env) == null ? void 0 : _options$env[envName];
  return opts ? descriptors(dirname, opts, alias + ".env[\"" + envName + "\"]") : null;
}
function buildOverrideDescriptors(_ref4, alias, descriptors, index) {
  var dirname = _ref4.dirname;
  var options = _ref4.options;

  var _options$overrides;
  var opts = (_options$overrides = options.overrides) == null ? void 0 : _options$overrides[index];
  if (!opts) throw new Error("Assertion failure - missing override");
  return descriptors(dirname, opts, alias + ".overrides[" + index + "]");
}
function buildOverrideEnvDescriptors(_ref5, alias, descriptors, index, envName) {
  var dirname = _ref5.dirname;
  var options = _ref5.options;

  var _options$overrides2, _override$env;
  var override = (_options$overrides2 = options.overrides) == null ? void 0 : _options$overrides2[index];
  if (!override) throw new Error("Assertion failure - missing override");
  var opts = (_override$env = override.env) == null ? void 0 : _override$env[envName];
  return opts ? descriptors(dirname, opts, alias + ".overrides[" + index + "].env[\"" + envName + "\"]") : null;
}
function makeChainWalker(_ref6) {
  var root = _ref6.root;
  var env = _ref6.env;
  var overrides = _ref6.overrides;
  var overridesEnv = _ref6.overridesEnv;
  var createLogger = _ref6.createLogger;

  return regeneratorRuntime.mark(function chainWalker(input, context, files, baseLogger) {
    if (files === undefined) files = new Set();

    var dirname, flattenedConfigs, rootOpts, envOpts, chain, logger, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _step$value, config, index, envName;

    return regeneratorRuntime.wrap(function chainWalker$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          dirname = input.dirname;
          flattenedConfigs = [];
          rootOpts = root(input);

          if (configIsApplicable(rootOpts, dirname, context, input.filepath)) {
            flattenedConfigs.push({
              config: rootOpts,
              envName: undefined,
              index: undefined
            });
            envOpts = env(input, context.envName);

            if (envOpts && configIsApplicable(envOpts, dirname, context, input.filepath)) {
              flattenedConfigs.push({
                config: envOpts,
                envName: context.envName,
                index: undefined
              });
            }
            (rootOpts.options.overrides || []).forEach(function (_, index) {
              var overrideOps = overrides(input, index);
              if (configIsApplicable(overrideOps, dirname, context, input.filepath)) {
                flattenedConfigs.push({
                  config: overrideOps,
                  index: index,
                  envName: undefined
                });
                var overrideEnvOpts = overridesEnv(input, index, context.envName);
                if (overrideEnvOpts && configIsApplicable(overrideEnvOpts, dirname, context, input.filepath)) {
                  flattenedConfigs.push({
                    config: overrideEnvOpts,
                    index: index,
                    envName: context.envName
                  });
                }
              }
            });
          }

          if (!flattenedConfigs.some(function (_ref7) {
            var _ref7$config$options = _ref7.config.options;
            var ignore = _ref7$config$options.ignore;
            var only = _ref7$config$options.only;
            return shouldIgnore(context, ignore, only, dirname);
          })) {
            context$2$0.next = 6;
            break;
          }

          return context$2$0.abrupt("return", null);

        case 6:
          chain = emptyChain();
          logger = createLogger(input, context, baseLogger);
          _iteratorNormalCompletion = true;
          _didIteratorError = false;
          _iteratorError = undefined;
          context$2$0.prev = 11;
          _iterator = flattenedConfigs[Symbol.iterator]();

        case 13:
          if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
            context$2$0.next = 26;
            break;
          }

          _step$value = _step.value;
          config = _step$value.config;
          index = _step$value.index;
          envName = _step$value.envName;
          return context$2$0.delegateYield(mergeExtendsChain(chain, config.options, dirname, context, files, baseLogger), "t0", 19);

        case 19:
          if (context$2$0.t0) {
            context$2$0.next = 21;
            break;
          }

          return context$2$0.abrupt("return", null);

        case 21:
          logger(config, index, envName);
          return context$2$0.delegateYield(mergeChainOpts(chain, config), "t1", 23);

        case 23:
          _iteratorNormalCompletion = true;
          context$2$0.next = 13;
          break;

        case 26:
          context$2$0.next = 32;
          break;

        case 28:
          context$2$0.prev = 28;
          context$2$0.t2 = context$2$0["catch"](11);
          _didIteratorError = true;
          _iteratorError = context$2$0.t2;

        case 32:
          context$2$0.prev = 32;
          context$2$0.prev = 33;

          if (!_iteratorNormalCompletion && _iterator["return"]) {
            _iterator["return"]();
          }

        case 35:
          context$2$0.prev = 35;

          if (!_didIteratorError) {
            context$2$0.next = 38;
            break;
          }

          throw _iteratorError;

        case 38:
          return context$2$0.finish(35);

        case 39:
          return context$2$0.finish(32);

        case 40:
          return context$2$0.abrupt("return", chain);

        case 41:
        case "end":
          return context$2$0.stop();
      }
    }, chainWalker, this, [[11, 28, 32, 40], [33,, 35, 39]]);
  });
}
function mergeExtendsChain(chain, opts, dirname, context, files, baseLogger) {
  var file, fileChain;
  return regeneratorRuntime.wrap(function mergeExtendsChain$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(opts["extends"] === undefined)) {
          context$1$0.next = 2;
          break;
        }

        return context$1$0.abrupt("return", true);

      case 2:
        return context$1$0.delegateYield((0, _index.loadConfig)(opts["extends"], dirname, context.envName, context.caller), "t0", 3);

      case 3:
        file = context$1$0.t0;

        if (!files.has(file)) {
          context$1$0.next = 6;
          break;
        }

        throw new Error("Configuration cycle detected loading " + file.filepath + ".\n" + "File already loaded following the config chain:\n" + Array.from(files, function (file) {
          return " - " + file.filepath;
        }).join("\n"));

      case 6:
        files.add(file);
        return context$1$0.delegateYield(loadFileChain(validateExtendFile(file), context, files, baseLogger), "t1", 8);

      case 8:
        fileChain = context$1$0.t1;

        files["delete"](file);

        if (fileChain) {
          context$1$0.next = 12;
          break;
        }

        return context$1$0.abrupt("return", false);

      case 12:
        mergeChain(chain, fileChain);
        return context$1$0.abrupt("return", true);

      case 14:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[3], this);
}
function mergeChain(target, source) {
  var _target$options, _target$plugins, _target$presets;

  (_target$options = target.options).push.apply(_target$options, _toConsumableArray(source.options));
  (_target$plugins = target.plugins).push.apply(_target$plugins, _toConsumableArray(source.plugins));
  (_target$presets = target.presets).push.apply(_target$presets, _toConsumableArray(source.presets));
  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = source.files[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var file = _step2.value;

      target.files.add(file);
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
        _iterator2["return"]();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  return target;
}
function mergeChainOpts(target, _ref8) {
  var _target$plugins2, _target$presets2;

  var options = _ref8.options;
  var plugins = _ref8.plugins;
  var presets = _ref8.presets;
  return regeneratorRuntime.wrap(function mergeChainOpts$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        target.options.push(options);
        context$1$0.t0 = (_target$plugins2 = target.plugins).push;
        context$1$0.t1 = _target$plugins2;
        return context$1$0.delegateYield(plugins(), "t2", 4);

      case 4:
        context$1$0.t3 = context$1$0.t2;
        context$1$0.t4 = _toConsumableArray(context$1$0.t3);
        context$1$0.t0.apply.call(context$1$0.t0, context$1$0.t1, context$1$0.t4);
        context$1$0.t5 = (_target$presets2 = target.presets).push;
        context$1$0.t6 = _target$presets2;
        return context$1$0.delegateYield(presets(), "t7", 10);

      case 10:
        context$1$0.t8 = context$1$0.t7;
        context$1$0.t9 = _toConsumableArray(context$1$0.t8);
        context$1$0.t5.apply.call(context$1$0.t5, context$1$0.t6, context$1$0.t9);
        return context$1$0.abrupt("return", target);

      case 14:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[4], this);
}
function emptyChain() {
  return {
    options: [],
    presets: [],
    plugins: [],
    files: new Set()
  };
}
function normalizeOptions(opts) {
  var options = Object.assign({}, opts);
  delete options["extends"];
  delete options.env;
  delete options.overrides;
  delete options.plugins;
  delete options.presets;
  delete options.passPerPreset;
  delete options.ignore;
  delete options.only;
  delete options.test;
  delete options.include;
  delete options.exclude;
  if (hasOwnProperty.call(options, "sourceMap")) {
    options.sourceMaps = options.sourceMap;
    delete options.sourceMap;
  }
  return options;
}
function dedupDescriptors(items) {
  var map = new Map();
  var descriptors = [];
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = items[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var item = _step3.value;

      if (typeof item.value === "function") {
        var fnKey = item.value;
        var nameMap = map.get(fnKey);
        if (!nameMap) {
          nameMap = new Map();
          map.set(fnKey, nameMap);
        }
        var desc = nameMap.get(item.name);
        if (!desc) {
          desc = {
            value: item
          };
          descriptors.push(desc);
          if (!item.ownPass) nameMap.set(item.name, desc);
        } else {
          desc.value = item;
        }
      } else {
        descriptors.push({
          value: item
        });
      }
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  return descriptors.reduce(function (acc, desc) {
    acc.push(desc.value);
    return acc;
  }, []);
}
function configIsApplicable(_ref9, dirname, context, configName) {
  var options = _ref9.options;

  return (options.test === undefined || configFieldIsApplicable(context, options.test, dirname, configName)) && (options.include === undefined || configFieldIsApplicable(context, options.include, dirname, configName)) && (options.exclude === undefined || !configFieldIsApplicable(context, options.exclude, dirname, configName));
}
function configFieldIsApplicable(context, test, dirname, configName) {
  var patterns = Array.isArray(test) ? test : [test];
  return matchesPatterns(context, patterns, dirname, configName);
}
function ignoreListReplacer(_key, value) {
  if (value instanceof RegExp) {
    return String(value);
  }
  return value;
}
function shouldIgnore(context, ignore, only, dirname) {
  if (ignore && matchesPatterns(context, ignore, dirname)) {
    var _context$filename;
    var message = "No config is applied to \"" + ((_context$filename = context.filename) != null ? _context$filename : "(unknown)") + "\" because it matches one of `ignore: " + JSON.stringify(ignore, ignoreListReplacer) + "` from \"" + dirname + "\"";
    debug(message);
    if (context.showConfig) {
      console.log(message);
    }
    return true;
  }
  if (only && !matchesPatterns(context, only, dirname)) {
    var _context$filename2;
    var message = "No config is applied to \"" + ((_context$filename2 = context.filename) != null ? _context$filename2 : "(unknown)") + "\" because it fails to match one of `only: " + JSON.stringify(only, ignoreListReplacer) + "` from \"" + dirname + "\"";
    debug(message);
    if (context.showConfig) {
      console.log(message);
    }
    return true;
  }
  return false;
}
function matchesPatterns(context, patterns, dirname, configName) {
  return patterns.some(function (pattern) {
    return matchPattern(pattern, dirname, context.filename, context, configName);
  });
}
function matchPattern(pattern, dirname, pathToTest, context, configName) {
  if (typeof pattern === "function") {
    return !!(0, _rewriteStackTrace.endHiddenCallStack)(pattern)(pathToTest, {
      dirname: dirname,
      envName: context.envName,
      caller: context.caller
    });
  }
  if (typeof pathToTest !== "string") {
    throw new _configError["default"]("Configuration contains string/RegExp pattern, but no filename was passed to Babel", configName);
  }
  if (typeof pattern === "string") {
    pattern = (0, _patternToRegex["default"])(pattern, dirname);
  }
  return pattern.test(pathToTest);
}
0 && 0;
