// @babel/core/lib/config/full.js
"use strict";

var marked0$0 = [loadPluginDescriptor, loadPresetDescriptor].map(regeneratorRuntime.mark);

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _async = require("../gensync-utils/async.js");
var _util = require("./util.js");
var context = require("../index.js");
var _plugin = require("./plugin.js");
var _item = require("./item.js");
var _configChain = require("./config-chain.js");
var _deepArray = require("./helpers/deep-array.js");
function _traverse() {
  var data = require("@babel/traverse");
  _traverse = function () {
    return data;
  };
  return data;
}
var _caching = require("./caching.js");
var _options = require("./validation/options.js");
var _plugins = require("./validation/plugins.js");
var _configApi = require("./helpers/config-api.js");
var _partial = require("./partial.js");
var _configError = require("../errors/config-error.js");
var _default = exports["default"] = _gensync()(regeneratorRuntime.mark(function loadFullConfig(inputOpts) {
  var _opts$assumptions, result, options, context, fileHandling, optionDefaults, plugins, presets, presetContext, toDescriptor, presetsDescriptors, initialPluginsDescriptors, pluginDescriptorsByPass, passes, externalDependencies, ignored, opts, pluginContext;

  return regeneratorRuntime.wrap(function loadFullConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield((0, _partial["default"])(inputOpts), "t0", 1);

      case 1:
        result = context$1$0.t0;

        if (result) {
          context$1$0.next = 4;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 4:
        options = result.options;
        context = result.context;
        fileHandling = result.fileHandling;

        if (!(fileHandling === "ignored")) {
          context$1$0.next = 9;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 9:
        optionDefaults = {};
        plugins = options.plugins;
        presets = options.presets;

        if (!(!plugins || !presets)) {
          context$1$0.next = 14;
          break;
        }

        throw new Error("Assertion failure - plugins and presets exist");

      case 14:
        presetContext = Object.assign({}, context, {
          targets: options.targets
        });

        toDescriptor = function toDescriptor(item) {
          var desc = (0, _item.getItemDescriptor)(item);
          if (!desc) {
            throw new Error("Assertion failure - must be config item");
          }
          return desc;
        };

        presetsDescriptors = presets.map(toDescriptor);
        initialPluginsDescriptors = plugins.map(toDescriptor);
        pluginDescriptorsByPass = [[]];
        passes = [];
        externalDependencies = [];
        return context$1$0.delegateYield(enhanceError(context, regeneratorRuntime.mark(function recursePresetDescriptors(rawPresets, pluginDescriptorsPass) {
          var presets, i, descriptor, preset, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _step$value, _preset, pass, _ignored;

          return regeneratorRuntime.wrap(function recursePresetDescriptors$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                presets = [];
                i = 0;

              case 2:
                if (!(i < rawPresets.length)) {
                  context$2$0.next = 19;
                  break;
                }

                descriptor = rawPresets[i];

                if (!(descriptor.options !== false)) {
                  context$2$0.next = 16;
                  break;
                }

                context$2$0.prev = 5;
                return context$2$0.delegateYield(loadPresetDescriptor(descriptor, presetContext), "t0", 7);

              case 7:
                preset = context$2$0.t0;
                context$2$0.next = 14;
                break;

              case 10:
                context$2$0.prev = 10;
                context$2$0.t1 = context$2$0["catch"](5);

                if (context$2$0.t1.code === "BABEL_UNKNOWN_OPTION") {
                  (0, _options.checkNoUnwrappedItemOptionPairs)(rawPresets, i, "preset", context$2$0.t1);
                }
                throw context$2$0.t1;

              case 14:
                externalDependencies.push(preset.externalDependencies);
                if (descriptor.ownPass) {
                  presets.push({
                    preset: preset.chain,
                    pass: []
                  });
                } else {
                  presets.unshift({
                    preset: preset.chain,
                    pass: pluginDescriptorsPass
                  });
                }

              case 16:
                i++;
                context$2$0.next = 2;
                break;

              case 19:
                if (!(presets.length > 0)) {
                  context$2$0.next = 55;
                  break;
                }

                pluginDescriptorsByPass.splice.apply(pluginDescriptorsByPass, [1, 0].concat(_toConsumableArray(presets.map(function (o) {
                  return o.pass;
                }).filter(function (p) {
                  return p !== pluginDescriptorsPass;
                }))));
                _iteratorNormalCompletion = true;
                _didIteratorError = false;
                _iteratorError = undefined;
                context$2$0.prev = 24;
                _iterator = presets[Symbol.iterator]();

              case 26:
                if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                  context$2$0.next = 41;
                  break;
                }

                _step$value = _step.value;
                _preset = _step$value.preset;
                pass = _step$value.pass;

                if (_preset) {
                  context$2$0.next = 32;
                  break;
                }

                return context$2$0.abrupt("return", true);

              case 32:
                pass.push.apply(pass, _toConsumableArray(_preset.plugins));
                return context$2$0.delegateYield(recursePresetDescriptors(_preset.presets, pass), "t2", 34);

              case 34:
                _ignored = context$2$0.t2;

                if (!_ignored) {
                  context$2$0.next = 37;
                  break;
                }

                return context$2$0.abrupt("return", true);

              case 37:
                _preset.options.forEach(function (opts) {
                  (0, _util.mergeOptions)(optionDefaults, opts);
                });

              case 38:
                _iteratorNormalCompletion = true;
                context$2$0.next = 26;
                break;

              case 41:
                context$2$0.next = 47;
                break;

              case 43:
                context$2$0.prev = 43;
                context$2$0.t3 = context$2$0["catch"](24);
                _didIteratorError = true;
                _iteratorError = context$2$0.t3;

              case 47:
                context$2$0.prev = 47;
                context$2$0.prev = 48;

                if (!_iteratorNormalCompletion && _iterator["return"]) {
                  _iterator["return"]();
                }

              case 50:
                context$2$0.prev = 50;

                if (!_didIteratorError) {
                  context$2$0.next = 53;
                  break;
                }

                throw _iteratorError;

              case 53:
                return context$2$0.finish(50);

              case 54:
                return context$2$0.finish(47);

              case 55:
              case "end":
                return context$2$0.stop();
            }
          }, recursePresetDescriptors, this, [[5, 10], [24, 43, 47, 55], [48,, 50, 54]]);
        }))(presetsDescriptors, pluginDescriptorsByPass[0]), "t1", 22);

      case 22:
        ignored = context$1$0.t1;

        if (!ignored) {
          context$1$0.next = 25;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 25:
        opts = optionDefaults;

        (0, _util.mergeOptions)(opts, options);
        pluginContext = Object.assign({}, presetContext, {
          assumptions: (_opts$assumptions = opts.assumptions) != null ? _opts$assumptions : {}
        });
        return context$1$0.delegateYield(enhanceError(context, regeneratorRuntime.mark(function loadPluginDescriptors() {
          var _pluginDescriptorsByPass$0;

          var _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, descs, pass, i, descriptor, plugin;

          return regeneratorRuntime.wrap(function loadPluginDescriptors$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                (_pluginDescriptorsByPass$0 = pluginDescriptorsByPass[0]).unshift.apply(_pluginDescriptorsByPass$0, _toConsumableArray(initialPluginsDescriptors));
                _iteratorNormalCompletion2 = true;
                _didIteratorError2 = false;
                _iteratorError2 = undefined;
                context$2$0.prev = 4;
                _iterator2 = pluginDescriptorsByPass[Symbol.iterator]();

              case 6:
                if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
                  context$2$0.next = 31;
                  break;
                }

                descs = _step2.value;
                pass = [];

                passes.push(pass);
                i = 0;

              case 11:
                if (!(i < descs.length)) {
                  context$2$0.next = 28;
                  break;
                }

                descriptor = descs[i];

                if (!(descriptor.options !== false)) {
                  context$2$0.next = 25;
                  break;
                }

                context$2$0.prev = 14;
                return context$2$0.delegateYield(loadPluginDescriptor(descriptor, pluginContext), "t0", 16);

              case 16:
                plugin = context$2$0.t0;
                context$2$0.next = 23;
                break;

              case 19:
                context$2$0.prev = 19;
                context$2$0.t1 = context$2$0["catch"](14);

                if (context$2$0.t1.code === "BABEL_UNKNOWN_PLUGIN_PROPERTY") {
                  (0, _options.checkNoUnwrappedItemOptionPairs)(descs, i, "plugin", context$2$0.t1);
                }
                throw context$2$0.t1;

              case 23:
                pass.push(plugin);
                externalDependencies.push(plugin.externalDependencies);

              case 25:
                i++;
                context$2$0.next = 11;
                break;

              case 28:
                _iteratorNormalCompletion2 = true;
                context$2$0.next = 6;
                break;

              case 31:
                context$2$0.next = 37;
                break;

              case 33:
                context$2$0.prev = 33;
                context$2$0.t2 = context$2$0["catch"](4);
                _didIteratorError2 = true;
                _iteratorError2 = context$2$0.t2;

              case 37:
                context$2$0.prev = 37;
                context$2$0.prev = 38;

                if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
                  _iterator2["return"]();
                }

              case 40:
                context$2$0.prev = 40;

                if (!_didIteratorError2) {
                  context$2$0.next = 43;
                  break;
                }

                throw _iteratorError2;

              case 43:
                return context$2$0.finish(40);

              case 44:
                return context$2$0.finish(37);

              case 45:
              case "end":
                return context$2$0.stop();
            }
          }, loadPluginDescriptors, this, [[4, 33, 37, 45], [14, 19], [38,, 40, 44]]);
        }))(), "t2", 29);

      case 29:
        opts.plugins = passes[0];
        opts.presets = passes.slice(1).filter(function (plugins) {
          return plugins.length > 0;
        }).map(function (plugins) {
          return {
            plugins: plugins
          };
        });
        opts.passPerPreset = opts.presets.length > 0;
        return context$1$0.abrupt("return", {
          options: opts,
          passes: passes,
          externalDependencies: (0, _deepArray.finalize)(externalDependencies)
        });

      case 33:
      case "end":
        return context$1$0.stop();
    }
  }, loadFullConfig, this);
}));
function enhanceError(context, fn) {
  return regeneratorRuntime.mark(function callee$1$0(arg1, arg2) {
    var _context$filename;

    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          context$2$0.prev = 0;
          return context$2$0.delegateYield(fn(arg1, arg2), "t0", 2);

        case 2:
          return context$2$0.abrupt("return", context$2$0.t0);

        case 5:
          context$2$0.prev = 5;
          context$2$0.t1 = context$2$0["catch"](0);

          if (!/^\[BABEL\]/.test(context$2$0.t1.message)) {
            context$2$0.t1.message = "[BABEL] " + ((_context$filename = context.filename) != null ? _context$filename : "unknown file") + ": " + context$2$0.t1.message;
          }
          throw context$2$0.t1;

        case 9:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this, [[0, 5]]);
  });
}
var makeDescriptorLoader = function makeDescriptorLoader(apiFactory) {
  return (0, _caching.makeWeakCache)(regeneratorRuntime.mark(function callee$1$0(_ref, cache) {
    var value = _ref.value;
    var options = _ref.options;
    var dirname = _ref.dirname;
    var alias = _ref.alias;
    var externalDependencies, item, factory, api, error;
    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          if (!(options === false)) {
            context$2$0.next = 2;
            break;
          }

          throw new Error("Assertion failure");

        case 2:
          options = options || {};
          externalDependencies = [];
          item = value;

          if (!(typeof value === "function")) {
            context$2$0.next = 17;
            break;
          }

          factory = (0, _async.maybeAsync)(value, "You appear to be using an async plugin/preset, but Babel has been called synchronously");
          api = Object.assign({}, context, apiFactory(cache, externalDependencies));
          context$2$0.prev = 8;
          return context$2$0.delegateYield(factory(api, options, dirname), "t0", 10);

        case 10:
          item = context$2$0.t0;
          context$2$0.next = 17;
          break;

        case 13:
          context$2$0.prev = 13;
          context$2$0.t1 = context$2$0["catch"](8);

          if (alias) {
            context$2$0.t1.message += " (While processing: " + JSON.stringify(alias) + ")";
          }
          throw context$2$0.t1;

        case 17:
          if (!(!item || typeof item !== "object")) {
            context$2$0.next = 19;
            break;
          }

          throw new Error("Plugin/Preset did not return an object.");

        case 19:
          if (!(0, _async.isThenable)(item)) {
            context$2$0.next = 22;
            break;
          }

          return context$2$0.delegateYield([], "t2", 21);

        case 21:
          throw new Error("You appear to be using a promise as a plugin, " + "which your current version of Babel does not support. " + "If you're using a published plugin, " + "you may need to upgrade your @babel/core version. " + "As an alternative, you can prefix the promise with \"await\". " + ("(While processing: " + JSON.stringify(alias) + ")"));

        case 22:
          if (!(externalDependencies.length > 0 && (!cache.configured() || cache.mode() === "forever"))) {
            context$2$0.next = 27;
            break;
          }

          error = "A plugin/preset has external untracked dependencies " + ("(" + externalDependencies[0] + "), but the cache ");

          if (!cache.configured()) {
            error += "has not been configured to be invalidated when the external dependencies change. ";
          } else {
            error += " has been configured to never be invalidated. ";
          }
          error += "Plugins/presets should configure their cache to be invalidated when the external " + "dependencies change, for example using `api.cache.invalidate(() => " + "statSync(filepath).mtimeMs)` or `api.cache.never()`\n" + ("(While processing: " + JSON.stringify(alias) + ")");
          throw new Error(error);

        case 27:
          return context$2$0.abrupt("return", {
            value: item,
            options: options,
            dirname: dirname,
            alias: alias,
            externalDependencies: (0, _deepArray.finalize)(externalDependencies)
          });

        case 28:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this, [[8, 13]]);
  }));
};
var pluginDescriptorLoader = makeDescriptorLoader(_configApi.makePluginAPI);
var presetDescriptorLoader = makeDescriptorLoader(_configApi.makePresetAPI);
var instantiatePlugin = (0, _caching.makeWeakCache)(regeneratorRuntime.mark(function callee$0$0(_ref2, cache) {
  var value = _ref2.value;
  var options = _ref2.options;
  var dirname = _ref2.dirname;
  var alias = _ref2.alias;
  var externalDependencies = _ref2.externalDependencies;
  var pluginObj, plugin;
  return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        pluginObj = (0, _plugins.validatePluginObject)(value);
        plugin = Object.assign({}, pluginObj);

        if (plugin.visitor) {
          plugin.visitor = _traverse()["default"].explode(Object.assign({}, plugin.visitor));
        }

        if (!plugin.inherits) {
          context$1$0.next = 5;
          break;
        }

        return context$1$0.delegateYield(regeneratorRuntime.mark(function callee$1$0() {
          var inheritsDescriptor, inherits;
          return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                inheritsDescriptor = {
                  name: undefined,
                  alias: alias + "$inherits",
                  value: plugin.inherits,
                  options: options,
                  dirname: dirname
                };
                return context$2$0.delegateYield((0, _async.forwardAsync)(loadPluginDescriptor, function (run) {
                  return cache.invalidate(function (data) {
                    return run(inheritsDescriptor, data);
                  });
                }), "t0", 2);

              case 2:
                inherits = context$2$0.t0;

                plugin.pre = chainMaybeAsync(inherits.pre, plugin.pre);
                plugin.post = chainMaybeAsync(inherits.post, plugin.post);
                plugin.manipulateOptions = chainMaybeAsync(inherits.manipulateOptions, plugin.manipulateOptions);
                plugin.visitor = _traverse()["default"].visitors.merge([inherits.visitor || {}, plugin.visitor || {}]);
                if (inherits.externalDependencies.length > 0) {
                  if (externalDependencies.length === 0) {
                    externalDependencies = inherits.externalDependencies;
                  } else {
                    externalDependencies = (0, _deepArray.finalize)([externalDependencies, inherits.externalDependencies]);
                  }
                }

              case 8:
              case "end":
                return context$2$0.stop();
            }
          }, callee$1$0, _this);
        })(), "t0", 5);

      case 5:
        return context$1$0.abrupt("return", new _plugin["default"](plugin, options, alias, externalDependencies));

      case 6:
      case "end":
        return context$1$0.stop();
    }
  }, callee$0$0, this);
}));
function loadPluginDescriptor(descriptor, context) {
  return regeneratorRuntime.wrap(function loadPluginDescriptor$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(descriptor.value instanceof _plugin["default"])) {
          context$1$0.next = 4;
          break;
        }

        if (!descriptor.options) {
          context$1$0.next = 3;
          break;
        }

        throw new Error("Passed options to an existing Plugin instance will not work.");

      case 3:
        return context$1$0.abrupt("return", descriptor.value);

      case 4:
        return context$1$0.delegateYield(pluginDescriptorLoader(descriptor, context), "t0", 5);

      case 5:
        context$1$0.t1 = context$1$0.t0;
        context$1$0.t2 = context;
        return context$1$0.delegateYield(instantiatePlugin(context$1$0.t1, context$1$0.t2), "t3", 8);

      case 8:
        return context$1$0.abrupt("return", context$1$0.t3);

      case 9:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
var needsFilename = function needsFilename(val) {
  return val && typeof val !== "function";
};
var validateIfOptionNeedsFilename = function validateIfOptionNeedsFilename(options, descriptor) {
  if (needsFilename(options.test) || needsFilename(options.include) || needsFilename(options.exclude)) {
    var formattedPresetName = descriptor.name ? "\"" + descriptor.name + "\"" : "/* your preset */";
    throw new _configError["default"](["Preset " + formattedPresetName + " requires a filename to be set when babel is called directly,", "```", "babel.transformSync(code, { filename: 'file.ts', presets: [" + formattedPresetName + "] });", "```", "See https://babeljs.io/docs/en/options#filename for more information."].join("\n"));
  }
};
var validatePreset = function validatePreset(preset, context, descriptor) {
  if (!context.filename) {
    var _options$overrides;
    var options = preset.options;

    validateIfOptionNeedsFilename(options, descriptor);
    (_options$overrides = options.overrides) == null || _options$overrides.forEach(function (overrideOptions) {
      return validateIfOptionNeedsFilename(overrideOptions, descriptor);
    });
  }
};
var instantiatePreset = (0, _caching.makeWeakCacheSync)(function (_ref3) {
  var value = _ref3.value;
  var dirname = _ref3.dirname;
  var alias = _ref3.alias;
  var externalDependencies = _ref3.externalDependencies;

  return {
    options: (0, _options.validate)("preset", value),
    alias: alias,
    dirname: dirname,
    externalDependencies: externalDependencies
  };
});
function loadPresetDescriptor(descriptor, context) {
  var preset;
  return regeneratorRuntime.wrap(function loadPresetDescriptor$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(presetDescriptorLoader(descriptor, context), "t0", 1);

      case 1:
        context$1$0.t1 = context$1$0.t0;
        preset = instantiatePreset(context$1$0.t1);

        validatePreset(preset, context, descriptor);
        return context$1$0.delegateYield((0, _configChain.buildPresetChain)(preset, context), "t2", 5);

      case 5:
        context$1$0.t3 = context$1$0.t2;
        context$1$0.t4 = preset.externalDependencies;
        return context$1$0.abrupt("return", {
          chain: context$1$0.t3,
          externalDependencies: context$1$0.t4
        });

      case 8:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}
function chainMaybeAsync(a, b) {
  if (!a) return b;
  if (!b) return a;
  return function () {
    var _this2 = this;

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var res = a.apply(this, args);
    if (res && typeof res.then === "function") {
      return res.then(function () {
        return b.apply(_this2, args);
      });
    }
    return b.apply(this, args);
  };
}
0 && 0;
