// @babel/core/lib/config/index.js
"use strict";

var marked0$0 = [loadOptionsImpl].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createConfigItem = createConfigItem;
exports.createConfigItemAsync = createConfigItemAsync;
exports.createConfigItemSync = createConfigItemSync;
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _full["default"];
  }
});
exports.loadOptions = loadOptions;
exports.loadOptionsAsync = loadOptionsAsync;
exports.loadOptionsSync = loadOptionsSync;
exports.loadPartialConfig = loadPartialConfig;
exports.loadPartialConfigAsync = loadPartialConfigAsync;
exports.loadPartialConfigSync = loadPartialConfigSync;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _full = require("./full.js");
var _partial = require("./partial.js");
var _item = require("./item.js");
var _rewriteStackTrace = require("../errors/rewrite-stack-trace.js");
var loadPartialConfigRunner = _gensync()(_partial.loadPartialConfig);
function loadPartialConfigAsync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(loadPartialConfigRunner.async).apply(undefined, arguments);
}
function loadPartialConfigSync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(loadPartialConfigRunner.sync).apply(undefined, arguments);
}
function loadPartialConfig(opts, callback) {
  if (callback !== undefined) {
    (0, _rewriteStackTrace.beginHiddenCallStack)(loadPartialConfigRunner.errback)(opts, callback);
  } else if (typeof opts === "function") {
    (0, _rewriteStackTrace.beginHiddenCallStack)(loadPartialConfigRunner.errback)(undefined, opts);
  } else {
    {
      return loadPartialConfigSync(opts);
    }
  }
}
function loadOptionsImpl(opts) {
  var _config$options, config;

  return regeneratorRuntime.wrap(function loadOptionsImpl$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield((0, _full["default"])(opts), "t0", 1);

      case 1:
        config = context$1$0.t0;
        return context$1$0.abrupt("return", (_config$options = config == null ? void 0 : config.options) != null ? _config$options : null);

      case 3:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
var loadOptionsRunner = _gensync()(loadOptionsImpl);
function loadOptionsAsync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(loadOptionsRunner.async).apply(undefined, arguments);
}
function loadOptionsSync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(loadOptionsRunner.sync).apply(undefined, arguments);
}
function loadOptions(opts, callback) {
  if (callback !== undefined) {
    (0, _rewriteStackTrace.beginHiddenCallStack)(loadOptionsRunner.errback)(opts, callback);
  } else if (typeof opts === "function") {
    (0, _rewriteStackTrace.beginHiddenCallStack)(loadOptionsRunner.errback)(undefined, opts);
  } else {
    {
      return loadOptionsSync(opts);
    }
  }
}
var createConfigItemRunner = _gensync()(_item.createConfigItem);
function createConfigItemAsync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(createConfigItemRunner.async).apply(undefined, arguments);
}
function createConfigItemSync() {
  return (0, _rewriteStackTrace.beginHiddenCallStack)(createConfigItemRunner.sync).apply(undefined, arguments);
}
function createConfigItem(target, options, callback) {
  if (callback !== undefined) {
    (0, _rewriteStackTrace.beginHiddenCallStack)(createConfigItemRunner.errback)(target, options, callback);
  } else if (typeof options === "function") {
    (0, _rewriteStackTrace.beginHiddenCallStack)(createConfigItemRunner.errback)(target, undefined, callback);
  } else {
    {
      return createConfigItemSync(target, options);
    }
  }
}
0 && 0;
