// @babel/core/lib/config/item.js
"use strict";

var marked0$0 = [createConfigItem].map(regeneratorRuntime.mark);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createConfigItem = createConfigItem;
exports.createItemFromDescriptor = createItemFromDescriptor;
exports.getItemDescriptor = getItemDescriptor;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
var _configDescriptors = require("./config-descriptors.js");
function createItemFromDescriptor(desc) {
  return new ConfigItem(desc);
}
function createConfigItem(value) {
  var _ref,
      _ref$dirname,
      dirname,
      type,
      descriptor,
      args$1$0 = arguments;

  return regeneratorRuntime.wrap(function createConfigItem$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _ref = args$1$0.length <= 1 || args$1$0[1] === undefined ? {} : args$1$0[1];
        _ref$dirname = _ref.dirname;
        dirname = _ref$dirname === undefined ? "." : _ref$dirname;
        type = _ref.type;
        return context$1$0.delegateYield((0, _configDescriptors.createDescriptor)(value, _path().resolve(dirname), {
          type: type,
          alias: "programmatic item"
        }), "t0", 5);

      case 5:
        descriptor = context$1$0.t0;
        return context$1$0.abrupt("return", createItemFromDescriptor(descriptor));

      case 7:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
var CONFIG_ITEM_BRAND = Symbol["for"]("@babel/core@7 - ConfigItem");
function getItemDescriptor(item) {
  if (item != null && item[CONFIG_ITEM_BRAND]) {
    return item._descriptor;
  }
  return undefined;
}

var ConfigItem = function ConfigItem(descriptor) {
  _classCallCheck(this, ConfigItem);

  this._descriptor = void 0;
  this[CONFIG_ITEM_BRAND] = true;
  this.value = void 0;
  this.options = void 0;
  this.dirname = void 0;
  this.name = void 0;
  this.file = void 0;
  this._descriptor = descriptor;
  Object.defineProperty(this, "_descriptor", {
    enumerable: false
  });
  Object.defineProperty(this, CONFIG_ITEM_BRAND, {
    enumerable: false
  });
  this.value = this._descriptor.value;
  this.options = this._descriptor.options;
  this.dirname = this._descriptor.dirname;
  this.name = this._descriptor.name;
  this.file = this._descriptor.file ? {
    request: this._descriptor.file.request,
    resolved: this._descriptor.file.resolved
  } : undefined;
  Object.freeze(this);
};

Object.freeze(ConfigItem.prototype);
0 && 0;
