// @babel/core/lib/config/partial.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var marked0$0 = [loadPrivatePartialConfig, loadPartialConfig].map(regeneratorRuntime.mark);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = loadPrivatePartialConfig;
exports.loadPartialConfig = loadPartialConfig;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
var _plugin = require("./plugin.js");
var _util = require("./util.js");
var _item = require("./item.js");
var _configChain = require("./config-chain.js");
var _environment = require("./helpers/environment.js");
var _options = require("./validation/options.js");
var _index = require("./files/index.js");
var _resolveTargets = require("./resolve-targets.js");
var _excluded = ["showIgnoredFiles"];
function _objectWithoutPropertiesLoose(r, e) {
  if (null == r) return {};var t = {};for (var n in r) if (({}).hasOwnProperty.call(r, n)) {
    if (-1 !== e.indexOf(n)) continue;t[n] = r[n];
  }return t;
}
function resolveRootMode(rootDir, rootMode) {
  switch (rootMode) {
    case "root":
      return rootDir;
    case "upward-optional":
      {
        var upwardRootDir = (0, _index.findConfigUpwards)(rootDir);
        return upwardRootDir === null ? rootDir : upwardRootDir;
      }
    case "upward":
      {
        var upwardRootDir = (0, _index.findConfigUpwards)(rootDir);
        if (upwardRootDir !== null) return upwardRootDir;
        throw Object.assign(new Error("Babel was run with rootMode:\"upward\" but a root could not " + ("be found when searching upward from \"" + rootDir + "\".\n") + "One of the following config files must be in the directory tree: " + ("\"" + _index.ROOT_CONFIG_FILENAMES.join(", ") + "\".")), {
          code: "BABEL_ROOT_NOT_FOUND",
          dirname: rootDir
        });
      }
    default:
      throw new Error("Assertion failure - unknown rootMode value.");
  }
}
function loadPrivatePartialConfig(inputOpts) {
  var args, _args$envName, envName, _args$cwd, cwd, _args$root, rootDir, _args$rootMode, rootMode, caller, _args$cloneInputAst, cloneInputAst, absoluteCwd, absoluteRootDir, filename, showConfigPath, context, configChain, merged, options;

  return regeneratorRuntime.wrap(function loadPrivatePartialConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(inputOpts != null && (typeof inputOpts !== "object" || Array.isArray(inputOpts)))) {
          context$1$0.next = 2;
          break;
        }

        throw new Error("Babel options must be an object, null, or undefined");

      case 2:
        args = inputOpts ? (0, _options.validate)("arguments", inputOpts) : {};
        _args$envName = args.envName;
        envName = _args$envName === undefined ? (0, _environment.getEnv)() : _args$envName;
        _args$cwd = args.cwd;
        cwd = _args$cwd === undefined ? "." : _args$cwd;
        _args$root = args.root;
        rootDir = _args$root === undefined ? "." : _args$root;
        _args$rootMode = args.rootMode;
        rootMode = _args$rootMode === undefined ? "root" : _args$rootMode;
        caller = args.caller;
        _args$cloneInputAst = args.cloneInputAst;
        cloneInputAst = _args$cloneInputAst === undefined ? true : _args$cloneInputAst;
        absoluteCwd = _path().resolve(cwd);
        absoluteRootDir = resolveRootMode(_path().resolve(absoluteCwd, rootDir), rootMode);
        filename = typeof args.filename === "string" ? _path().resolve(cwd, args.filename) : undefined;
        return context$1$0.delegateYield((0, _index.resolveShowConfigPath)(absoluteCwd), "t0", 18);

      case 18:
        showConfigPath = context$1$0.t0;
        context = {
          filename: filename,
          cwd: absoluteCwd,
          root: absoluteRootDir,
          envName: envName,
          caller: caller,
          showConfig: showConfigPath === filename
        };
        return context$1$0.delegateYield((0, _configChain.buildRootChain)(args, context), "t1", 21);

      case 21:
        configChain = context$1$0.t1;

        if (configChain) {
          context$1$0.next = 24;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 24:
        merged = {
          assumptions: {}
        };

        configChain.options.forEach(function (opts) {
          (0, _util.mergeOptions)(merged, opts);
        });
        options = Object.assign({}, merged, {
          targets: (0, _resolveTargets.resolveTargets)(merged, absoluteRootDir),
          cloneInputAst: cloneInputAst,
          babelrc: false,
          configFile: false,
          browserslistConfigFile: false,
          passPerPreset: false,
          envName: context.envName,
          cwd: context.cwd,
          root: context.root,
          rootMode: "root",
          filename: typeof context.filename === "string" ? context.filename : undefined,
          plugins: configChain.plugins.map(function (descriptor) {
            return (0, _item.createItemFromDescriptor)(descriptor);
          }),
          presets: configChain.presets.map(function (descriptor) {
            return (0, _item.createItemFromDescriptor)(descriptor);
          })
        });
        return context$1$0.abrupt("return", {
          options: options,
          context: context,
          fileHandling: configChain.fileHandling,
          ignore: configChain.ignore,
          babelrc: configChain.babelrc,
          config: configChain.config,
          files: configChain.files
        });

      case 28:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function loadPartialConfig(opts) {
  var showIgnoredFiles, _opts, result, options, babelrc, ignore, config, fileHandling, files;

  return regeneratorRuntime.wrap(function loadPartialConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        showIgnoredFiles = false;

        if (typeof opts === "object" && opts !== null && !Array.isArray(opts)) {
          _opts = opts;
          showIgnoredFiles = _opts.showIgnoredFiles;

          opts = _objectWithoutPropertiesLoose(_opts, _excluded);
          _opts;
        }
        return context$1$0.delegateYield(loadPrivatePartialConfig(opts), "t0", 3);

      case 3:
        result = context$1$0.t0;

        if (result) {
          context$1$0.next = 6;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 6:
        options = result.options;
        babelrc = result.babelrc;
        ignore = result.ignore;
        config = result.config;
        fileHandling = result.fileHandling;
        files = result.files;

        if (!(fileHandling === "ignored" && !showIgnoredFiles)) {
          context$1$0.next = 14;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 14:
        (options.plugins || []).forEach(function (item) {
          if (item.value instanceof _plugin["default"]) {
            throw new Error("Passing cached plugin instances is not supported in " + "babel.loadPartialConfig()");
          }
        });
        return context$1$0.abrupt("return", new PartialConfig(options, babelrc ? babelrc.filepath : undefined, ignore ? ignore.filepath : undefined, config ? config.filepath : undefined, fileHandling, files));

      case 16:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}

var PartialConfig = (function () {
  function PartialConfig(options, babelrc, ignore, config, fileHandling, files) {
    _classCallCheck(this, PartialConfig);

    this.options = void 0;
    this.babelrc = void 0;
    this.babelignore = void 0;
    this.config = void 0;
    this.fileHandling = void 0;
    this.files = void 0;
    this.options = options;
    this.babelignore = ignore;
    this.babelrc = babelrc;
    this.config = config;
    this.fileHandling = fileHandling;
    this.files = files;
    Object.freeze(this);
  }

  _createClass(PartialConfig, [{
    key: "hasFilesystemConfig",
    value: function hasFilesystemConfig() {
      return this.babelrc !== undefined || this.config !== undefined;
    }
  }]);

  return PartialConfig;
})();

Object.freeze(PartialConfig.prototype);
0 && 0;
