// @babel/core/lib/config/printer.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfigPrinter = exports.ChainFormatter = void 0;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var ChainFormatter = exports.ChainFormatter = {
  Programmatic: 0,
  Config: 1
};
var Formatter = {
  title: function title(type, callerName, filepath) {
    var title = "";
    if (type === ChainFormatter.Programmatic) {
      title = "programmatic options";
      if (callerName) {
        title += " from " + callerName;
      }
    } else {
      title = "config " + filepath;
    }
    return title;
  },
  loc: function loc(index, envName) {
    var loc = "";
    if (index != null) {
      loc += ".overrides[" + index + "]";
    }
    if (envName != null) {
      loc += ".env[\"" + envName + "\"]";
    }
    return loc;
  },
  optionsAndDescriptors: regeneratorRuntime.mark(function optionsAndDescriptors(opt) {
    var content, pluginDescriptors, presetDescriptors;
    return regeneratorRuntime.wrap(function optionsAndDescriptors$(context$1$0) {
      while (1) switch (context$1$0.prev = context$1$0.next) {
        case 0:
          content = Object.assign({}, opt.options);

          delete content.overrides;
          delete content.env;
          context$1$0.t0 = [];
          return context$1$0.delegateYield(opt.plugins(), "t1", 5);

        case 5:
          context$1$0.t2 = context$1$0.t1;
          context$1$0.t3 = _toConsumableArray(context$1$0.t2);
          pluginDescriptors = context$1$0.t0.concat.call(context$1$0.t0, context$1$0.t3);

          if (pluginDescriptors.length) {
            content.plugins = pluginDescriptors.map(function (d) {
              return descriptorToConfig(d);
            });
          }
          context$1$0.t4 = [];
          return context$1$0.delegateYield(opt.presets(), "t5", 11);

        case 11:
          context$1$0.t6 = context$1$0.t5;
          context$1$0.t7 = _toConsumableArray(context$1$0.t6);
          presetDescriptors = context$1$0.t4.concat.call(context$1$0.t4, context$1$0.t7);

          if (presetDescriptors.length) {
            content.presets = [].concat(_toConsumableArray(presetDescriptors)).map(function (d) {
              return descriptorToConfig(d);
            });
          }
          return context$1$0.abrupt("return", JSON.stringify(content, undefined, 2));

        case 16:
        case "end":
          return context$1$0.stop();
      }
    }, optionsAndDescriptors, this);
  })
};
function descriptorToConfig(d) {
  var _d$file;
  var name = (_d$file = d.file) == null ? void 0 : _d$file.request;
  if (name == null) {
    if (typeof d.value === "object") {
      name = d.value;
    } else if (typeof d.value === "function") {
      name = "[Function: " + d.value.toString().slice(0, 50) + " ... ]";
    }
  }
  if (name == null) {
    name = "[Unknown]";
  }
  if (d.options === undefined) {
    return name;
  } else if (d.name == null) {
    return [name, d.options];
  } else {
    return [name, d.options, d.name];
  }
}

var ConfigPrinter = (function () {
  function ConfigPrinter() {
    _classCallCheck(this, ConfigPrinter);

    this._stack = [];
  }

  _createClass(ConfigPrinter, [{
    key: "configure",
    value: function configure(enabled, type, _ref) {
      var _this = this;

      var callerName = _ref.callerName;
      var filepath = _ref.filepath;

      if (!enabled) return function () {};
      return function (content, index, envName) {
        _this._stack.push({
          type: type,
          callerName: callerName,
          filepath: filepath,
          content: content,
          index: index,
          envName: envName
        });
      };
    }
  }, {
    key: "output",
    value: regeneratorRuntime.mark(function output() {
      var configs;
      return regeneratorRuntime.wrap(function output$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(this._stack.length === 0)) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt("return", "");

          case 2:
            return context$2$0.delegateYield(_gensync().all(this._stack.map(function (s) {
              return ConfigPrinter.format(s);
            })), "t0", 3);

          case 3:
            configs = context$2$0.t0;
            return context$2$0.abrupt("return", configs.join("\n\n"));

          case 5:
          case "end":
            return context$2$0.stop();
        }
      }, output, this);
    })
  }], [{
    key: "format",
    value: regeneratorRuntime.mark(function format(config) {
      var title, loc, content;
      return regeneratorRuntime.wrap(function format$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            title = Formatter.title(config.type, config.callerName, config.filepath);
            loc = Formatter.loc(config.index, config.envName);

            if (loc) title += " " + loc;
            return context$2$0.delegateYield(Formatter.optionsAndDescriptors(config.content), "t0", 4);

          case 4:
            content = context$2$0.t0;
            return context$2$0.abrupt("return", title + "\n" + content);

          case 6:
          case "end":
            return context$2$0.stop();
        }
      }, format, this);
    })
  }]);

  return ConfigPrinter;
})();

exports.ConfigPrinter = ConfigPrinter;
0 && 0;
