// @babel/helper-member-expression-to-functions/lib/index.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

Object.defineProperty(exports, '__esModule', { value: true });

var _t = require('@babel/types');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function get() {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var _t__namespace = /*#__PURE__*/_interopNamespace(_t);

function willPathCastToBoolean(_x) {
  var _again = true;

  _function: while (_again) {
    var path = _x;
    _again = false;

    var maybeWrapped = path;
    var node = maybeWrapped.node;
    var parentPath = maybeWrapped.parentPath;

    if (parentPath.isLogicalExpression()) {
      var _parentPath$node = parentPath.node;
      var operator = _parentPath$node.operator;
      var right = _parentPath$node.right;

      if (operator === "&&" || operator === "||" || operator === "??" && node === right) {
        _x = parentPath;
        _again = true;
        maybeWrapped = node = parentPath = _parentPath$node = operator = right = undefined;
        continue _function;
      }
    }
    if (parentPath.isSequenceExpression()) {
      var expressions = parentPath.node.expressions;

      if (expressions[expressions.length - 1] === node) {
        _x = parentPath;
        _again = true;
        maybeWrapped = node = parentPath = _parentPath$node = operator = right = expressions = undefined;
        continue _function;
      } else {
        return true;
      }
    }
    return parentPath.isConditional({
      test: node
    }) || parentPath.isUnaryExpression({
      operator: "!"
    }) || parentPath.isLoop({
      test: node
    });
  }
}

var LOGICAL_OPERATORS = _t__namespace.LOGICAL_OPERATORS;
var arrowFunctionExpression = _t__namespace.arrowFunctionExpression;
var assignmentExpression = _t__namespace.assignmentExpression;
var binaryExpression = _t__namespace.binaryExpression;
var booleanLiteral = _t__namespace.booleanLiteral;
var callExpression = _t__namespace.callExpression;
var cloneNode = _t__namespace.cloneNode;
var conditionalExpression = _t__namespace.conditionalExpression;
var identifier = _t__namespace.identifier;
var isMemberExpression = _t__namespace.isMemberExpression;
var isOptionalCallExpression = _t__namespace.isOptionalCallExpression;
var isOptionalMemberExpression = _t__namespace.isOptionalMemberExpression;
var isUpdateExpression = _t__namespace.isUpdateExpression;
var logicalExpression = _t__namespace.logicalExpression;
var memberExpression = _t__namespace.memberExpression;
var nullLiteral = _t__namespace.nullLiteral;
var optionalCallExpression = _t__namespace.optionalCallExpression;
var optionalMemberExpression = _t__namespace.optionalMemberExpression;
var sequenceExpression = _t__namespace.sequenceExpression;
var updateExpression = _t__namespace.updateExpression;

var AssignmentMemoiser = (function () {
  function AssignmentMemoiser() {
    _classCallCheck(this, AssignmentMemoiser);

    this._map = void 0;
    this._map = new WeakMap();
  }

  _createClass(AssignmentMemoiser, [{
    key: 'has',
    value: function has(key) {
      return this._map.has(key);
    }
  }, {
    key: 'get',
    value: function get(key) {
      if (!this.has(key)) return;
      var record = this._map.get(key);
      var value = record.value;

      record.count--;
      if (record.count === 0) {
        return assignmentExpression("=", value, key);
      }
      return value;
    }
  }, {
    key: 'set',
    value: function set(key, value, count) {
      return this._map.set(key, {
        count: count,
        value: value
      });
    }
  }]);

  return AssignmentMemoiser;
})();

function toNonOptional(path, base) {
  var node = path.node;

  if (isOptionalMemberExpression(node)) {
    return memberExpression(base, node.property, node.computed);
  }
  if (path.isOptionalCallExpression()) {
    var callee = path.get("callee");
    if (path.node.optional && callee.isOptionalMemberExpression()) {
      var object = callee.node.object;
      var context = path.scope.maybeGenerateMemoised(object);
      callee.get("object").replaceWith(assignmentExpression("=", context, object));
      return callExpression(memberExpression(base, identifier("call")), [context].concat(_toConsumableArray(path.node.arguments)));
    }
    return callExpression(base, path.node.arguments);
  }
  return path.node;
}
function isInDetachedTree(path) {
  while (path) {
    if (path.isProgram()) break;
    var _path = path;
    var parentPath = _path.parentPath;
    var container = _path.container;
    var listKey = _path.listKey;

    var parentNode = parentPath.node;
    if (listKey) {
      if (container !== parentNode[listKey]) {
        return true;
      }
    } else {
      if (container !== parentNode) return true;
    }
    path = parentPath;
  }
  return false;
}
var handle = {
  memoise: function memoise() {},
  handle: function handle(member, noDocumentAll) {
    var _this = this;

    var node = member.node;
    var parent = member.parent;
    var parentPath = member.parentPath;
    var scope = member.scope;

    if (member.isOptionalMemberExpression()) {
      var _ret = (function () {
        if (isInDetachedTree(member)) return {
            v: undefined
          };
        var endPath = member.find(function (_ref) {
          var node = _ref.node;
          var parent = _ref.parent;

          if (isOptionalMemberExpression(parent)) {
            return parent.optional || parent.object !== node;
          }
          if (isOptionalCallExpression(parent)) {
            return node !== member.node && parent.optional || parent.callee !== node;
          }
          return true;
        });
        if (scope.path.isPattern()) {
          endPath.replaceWith(callExpression(arrowFunctionExpression([], endPath.node), []));
          return {
            v: undefined
          };
        }
        var willEndPathCastToBoolean = willPathCastToBoolean(endPath);
        var rootParentPath = endPath.parentPath;
        if (rootParentPath.isUpdateExpression({
          argument: node
        })) {
          throw member.buildCodeFrameError('can\'t handle update expression');
        }
        var isAssignment = rootParentPath.isAssignmentExpression({
          left: endPath.node
        });
        var isDeleteOperation = rootParentPath.isUnaryExpression({
          operator: "delete"
        });
        if (isDeleteOperation && endPath.isOptionalMemberExpression() && endPath.get("property").isPrivateName()) {
          throw member.buildCodeFrameError('can\'t delete a private class element');
        }
        var startingOptional = member;
        for (;;) {
          if (startingOptional.isOptionalMemberExpression()) {
            if (startingOptional.node.optional) break;
            startingOptional = startingOptional.get("object");
            continue;
          } else if (startingOptional.isOptionalCallExpression()) {
            if (startingOptional.node.optional) break;
            startingOptional = startingOptional.get("callee");
            continue;
          }
          throw new Error('Internal error: unexpected ' + startingOptional.node.type);
        }
        var startingNode = startingOptional.isOptionalMemberExpression() ? startingOptional.node.object : startingOptional.node.callee;
        var baseNeedsMemoised = scope.maybeGenerateMemoised(startingNode);
        var baseRef = baseNeedsMemoised != null ? baseNeedsMemoised : startingNode;
        var parentIsOptionalCall = parentPath.isOptionalCallExpression({
          callee: node
        });
        var isOptionalCall = function isOptionalCall(parent) {
          return parentIsOptionalCall;
        };
        var parentIsCall = parentPath.isCallExpression({
          callee: node
        });
        startingOptional.replaceWith(toNonOptional(startingOptional, baseRef));
        if (isOptionalCall()) {
          if (parent.optional) {
            parentPath.replaceWith(_this.optionalCall(member, parent.arguments));
          } else {
            parentPath.replaceWith(_this.call(member, parent.arguments));
          }
        } else if (parentIsCall) {
          member.replaceWith(_this.boundGet(member));
        } else if (_this['delete'] && parentPath.isUnaryExpression({
          operator: "delete"
        })) {
          parentPath.replaceWith(_this['delete'](member));
        } else if (parentPath.isAssignmentExpression()) {
          handleAssignment(_this, member, parentPath);
        } else {
          member.replaceWith(_this.get(member));
        }
        var regular = member.node;
        for (var current = member; current !== endPath;) {
          var _parentPath = current.parentPath;
          if (_parentPath === endPath && isOptionalCall() && parent.optional) {
            regular = _parentPath.node;
            break;
          }
          regular = toNonOptional(_parentPath, regular);
          current = _parentPath;
        }
        var context = undefined;
        var endParentPath = endPath.parentPath;
        if (isMemberExpression(regular) && endParentPath.isOptionalCallExpression({
          callee: endPath.node,
          optional: true
        })) {
          var _regular = regular;
          var object = _regular.object;

          context = member.scope.maybeGenerateMemoised(object);
          if (context) {
            regular.object = assignmentExpression("=", context, object);
          }
        }
        var replacementPath = endPath;
        if (isDeleteOperation || isAssignment) {
          replacementPath = endParentPath;
          regular = endParentPath.node;
        }
        var baseMemoised = baseNeedsMemoised ? assignmentExpression("=", cloneNode(baseRef), cloneNode(startingNode)) : cloneNode(baseRef);
        if (willEndPathCastToBoolean) {
          var nonNullishCheck = undefined;
          if (noDocumentAll) {
            nonNullishCheck = binaryExpression("!=", baseMemoised, nullLiteral());
          } else {
            nonNullishCheck = logicalExpression("&&", binaryExpression("!==", baseMemoised, nullLiteral()), binaryExpression("!==", cloneNode(baseRef), scope.buildUndefinedNode()));
          }
          replacementPath.replaceWith(logicalExpression("&&", nonNullishCheck, regular));
        } else {
          var nullishCheck = undefined;
          if (noDocumentAll) {
            nullishCheck = binaryExpression("==", baseMemoised, nullLiteral());
          } else {
            nullishCheck = logicalExpression("||", binaryExpression("===", baseMemoised, nullLiteral()), binaryExpression("===", cloneNode(baseRef), scope.buildUndefinedNode()));
          }
          replacementPath.replaceWith(conditionalExpression(nullishCheck, isDeleteOperation ? booleanLiteral(true) : scope.buildUndefinedNode(), regular));
        }
        if (context) {
          var endParent = endParentPath.node;
          endParentPath.replaceWith(optionalCallExpression(optionalMemberExpression(endParent.callee, identifier("call"), false, true), [cloneNode(context)].concat(_toConsumableArray(endParent.arguments)), false));
        }
        return {
          v: undefined
        };
      })();

      if (typeof _ret === 'object') return _ret.v;
    }
    if (isUpdateExpression(parent, {
      argument: node
    })) {
      if (this.simpleSet) {
        member.replaceWith(this.simpleSet(member));
        return;
      }
      var operator = parent.operator;
      var prefix = parent.prefix;

      this.memoise(member, 2);
      var ref = scope.generateUidIdentifierBasedOnNode(node);
      scope.push({
        id: ref
      });
      var seq = [assignmentExpression("=", cloneNode(ref), this.get(member))];
      if (prefix) {
        seq.push(updateExpression(operator, cloneNode(ref), prefix));
        var value = sequenceExpression(seq);
        parentPath.replaceWith(this.set(member, value));
        return;
      } else {
        var ref2 = scope.generateUidIdentifierBasedOnNode(node);
        scope.push({
          id: ref2
        });
        seq.push(assignmentExpression("=", cloneNode(ref2), updateExpression(operator, cloneNode(ref), prefix)), cloneNode(ref));
        var value = sequenceExpression(seq);
        parentPath.replaceWith(sequenceExpression([this.set(member, value), cloneNode(ref2)]));
        return;
      }
    }
    if (parentPath.isAssignmentExpression({
      left: node
    })) {
      handleAssignment(this, member, parentPath);
      return;
    }
    if (parentPath.isCallExpression({
      callee: node
    })) {
      parentPath.replaceWith(this.call(member, parentPath.node.arguments));
      return;
    }
    if (parentPath.isOptionalCallExpression({
      callee: node
    })) {
      if (scope.path.isPattern()) {
        parentPath.replaceWith(callExpression(arrowFunctionExpression([], parentPath.node), []));
        return;
      }
      parentPath.replaceWith(this.optionalCall(member, parentPath.node.arguments));
      return;
    }
    if (this['delete'] && parentPath.isUnaryExpression({
      operator: "delete"
    })) {
      parentPath.replaceWith(this['delete'](member));
      return;
    }
    if (parentPath.isForXStatement({
      left: node
    }) || parentPath.isObjectProperty({
      value: node
    }) && parentPath.parentPath.isObjectPattern() || parentPath.isAssignmentPattern({
      left: node
    }) && parentPath.parentPath.isObjectProperty({
      value: parent
    }) && parentPath.parentPath.parentPath.isObjectPattern() || parentPath.isArrayPattern() || parentPath.isAssignmentPattern({
      left: node
    }) && parentPath.parentPath.isArrayPattern() || parentPath.isRestElement()) {
      member.replaceWith(this.destructureSet(member));
      return;
    }
    if (parentPath.isTaggedTemplateExpression()) {
      member.replaceWith(this.boundGet(member));
    } else {
      member.replaceWith(this.get(member));
    }
  }
};
function handleAssignment(state, member, parentPath) {
  if (state.simpleSet) {
    member.replaceWith(state.simpleSet(member));
    return;
  }
  var _parentPath$node2 = parentPath.node;
  var operator = _parentPath$node2.operator;
  var value = _parentPath$node2.right;

  if (operator === "=") {
    parentPath.replaceWith(state.set(member, value));
  } else {
    var operatorTrunc = operator.slice(0, -1);
    if (LOGICAL_OPERATORS.includes(operatorTrunc)) {
      state.memoise(member, 1);
      parentPath.replaceWith(logicalExpression(operatorTrunc, state.get(member), state.set(member, value)));
    } else {
      state.memoise(member, 2);
      parentPath.replaceWith(state.set(member, binaryExpression(operatorTrunc, state.get(member), value)));
    }
  }
}
function memberExpressionToFunctions(path, visitor, state) {
  path.traverse(visitor, Object.assign({}, handle, state, {
    memoiser: new AssignmentMemoiser()
  }));
}

exports['default'] = memberExpressionToFunctions;
