// @babel/helper-replace-supers/lib/index.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _templateObject = _taggedTemplateLiteral(["\n          function () { throw new ReferenceError(\"'delete super[expr]' is invalid\"); }()\n        "], ["\n          function () { throw new ReferenceError(\"'delete super[expr]' is invalid\"); }()\n        "]),
    _templateObject2 = _taggedTemplateLiteral(["\n        function () { throw new ReferenceError(\"'delete super.prop' is invalid\"); }()\n      "], ["\n        function () { throw new ReferenceError(\"'delete super.prop' is invalid\"); }()\n      "]);

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _helperMemberExpressionToFunctions = require("@babel/helper-member-expression-to-functions");
var _helperOptimiseCallExpression = require("@babel/helper-optimise-call-expression");
var _core = require("@babel/core");
var _traverse = require("@babel/traverse");
var _core$types = _core.types;
var assignmentExpression = _core$types.assignmentExpression;
var callExpression = _core$types.callExpression;
var cloneNode = _core$types.cloneNode;
var identifier = _core$types.identifier;
var memberExpression = _core$types.memberExpression;
var sequenceExpression = _core$types.sequenceExpression;
var stringLiteral = _core$types.stringLiteral;
var thisExpression = _core$types.thisExpression;

{
  exports.environmentVisitor = _traverse.visitors.environmentVisitor({});
  exports.skipAllButComputedKey = function skipAllButComputedKey(path) {
    path.skip();
    if (path.node.computed) {
      path.context.maybeQueue(path.get("key"));
    }
  };
}
var visitor = _traverse.visitors.environmentVisitor({
  Super: function Super(path, state) {
    var node = path.node;
    var parentPath = path.parentPath;

    if (!parentPath.isMemberExpression({
      object: node
    })) return;
    state.handle(parentPath);
  }
});
var unshadowSuperBindingVisitor = _traverse.visitors.environmentVisitor({
  Scopable: function Scopable(path, _ref) {
    var refName = _ref.refName;

    var binding = path.scope.getOwnBinding(refName);
    if (binding && binding.identifier.name === refName) {
      path.scope.rename(refName);
    }
  }
});
var specHandlers = {
  memoise: function memoise(superMember, count) {
    var scope = superMember.scope;
    var node = superMember.node;
    var computed = node.computed;
    var property = node.property;

    if (!computed) {
      return;
    }
    var memo = scope.maybeGenerateMemoised(property);
    if (!memo) {
      return;
    }
    this.memoiser.set(property, memo, count);
  },
  prop: function prop(superMember) {
    var _superMember$node = superMember.node;
    var computed = _superMember$node.computed;
    var property = _superMember$node.property;

    if (this.memoiser.has(property)) {
      return cloneNode(this.memoiser.get(property));
    }
    if (computed) {
      return cloneNode(property);
    }
    return stringLiteral(property.name);
  },
  _getPrototypeOfExpression: function _getPrototypeOfExpression() {
    var objectRef = cloneNode(this.getObjectRef());
    var targetRef = this.isStatic || this.isPrivateMethod ? objectRef : memberExpression(objectRef, identifier("prototype"));
    return callExpression(this.file.addHelper("getPrototypeOf"), [targetRef]);
  },
  get: function get(superMember) {
    var objectRef = cloneNode(this.getObjectRef());
    return callExpression(this.file.addHelper("superPropGet"), [this.isDerivedConstructor ? sequenceExpression([thisExpression(), objectRef]) : objectRef, this.prop(superMember), thisExpression()].concat(_toConsumableArray(this.isStatic || this.isPrivateMethod ? [] : [_core.types.numericLiteral(1)])));
  },
  _call: function _call(superMember, args, optional) {
    var objectRef = cloneNode(this.getObjectRef());
    var argsNode = undefined;
    if (args.length === 1 && _core.types.isSpreadElement(args[0]) && (_core.types.isIdentifier(args[0].argument) || _core.types.isArrayExpression(args[0].argument))) {
      argsNode = args[0].argument;
    } else {
      argsNode = _core.types.arrayExpression(args);
    }
    var call = _core.types.callExpression(this.file.addHelper("superPropGet"), [this.isDerivedConstructor ? sequenceExpression([thisExpression(), objectRef]) : objectRef, this.prop(superMember), thisExpression(), _core.types.numericLiteral(2 | (this.isStatic || this.isPrivateMethod ? 0 : 1))]);
    if (optional) {
      return _core.types.optionalCallExpression(call, [argsNode], true);
    }
    return callExpression(call, [argsNode]);
  },
  set: function set(superMember, value) {
    var objectRef = cloneNode(this.getObjectRef());
    return callExpression(this.file.addHelper("superPropSet"), [this.isDerivedConstructor ? sequenceExpression([thisExpression(), objectRef]) : objectRef, this.prop(superMember), value, thisExpression(), _core.types.numericLiteral(superMember.isInStrictMode() ? 1 : 0)].concat(_toConsumableArray(this.isStatic || this.isPrivateMethod ? [] : [_core.types.numericLiteral(1)])));
  },
  destructureSet: function destructureSet(superMember) {
    throw superMember.buildCodeFrameError("Destructuring to a super field is not supported yet.");
  },
  call: function call(superMember, args) {
    return this._call(superMember, args, false);
  },
  optionalCall: function optionalCall(superMember, args) {
    return this._call(superMember, args, true);
  },
  "delete": function _delete(superMember) {
    if (superMember.node.computed) {
      return sequenceExpression([callExpression(this.file.addHelper("toPropertyKey"), [cloneNode(superMember.node.property)]), _core.template.expression.ast(_templateObject)]);
    } else {
      return _core.template.expression.ast(_templateObject2);
    }
  }
};
var specHandlers_old = {
  memoise: function memoise(superMember, count) {
    var scope = superMember.scope;
    var node = superMember.node;
    var computed = node.computed;
    var property = node.property;

    if (!computed) {
      return;
    }
    var memo = scope.maybeGenerateMemoised(property);
    if (!memo) {
      return;
    }
    this.memoiser.set(property, memo, count);
  },
  prop: function prop(superMember) {
    var _superMember$node2 = superMember.node;
    var computed = _superMember$node2.computed;
    var property = _superMember$node2.property;

    if (this.memoiser.has(property)) {
      return cloneNode(this.memoiser.get(property));
    }
    if (computed) {
      return cloneNode(property);
    }
    return stringLiteral(property.name);
  },
  _getPrototypeOfExpression: function _getPrototypeOfExpression() {
    var objectRef = cloneNode(this.getObjectRef());
    var targetRef = this.isStatic || this.isPrivateMethod ? objectRef : memberExpression(objectRef, identifier("prototype"));
    return callExpression(this.file.addHelper("getPrototypeOf"), [targetRef]);
  },
  get: function get(superMember) {
    return this._get(superMember);
  },
  _get: function _get(superMember) {
    var proto = this._getPrototypeOfExpression();
    return callExpression(this.file.addHelper("get"), [this.isDerivedConstructor ? sequenceExpression([thisExpression(), proto]) : proto, this.prop(superMember), thisExpression()]);
  },
  set: function set(superMember, value) {
    var proto = this._getPrototypeOfExpression();
    return callExpression(this.file.addHelper("set"), [this.isDerivedConstructor ? sequenceExpression([thisExpression(), proto]) : proto, this.prop(superMember), value, thisExpression(), _core.types.booleanLiteral(superMember.isInStrictMode())]);
  },
  destructureSet: function destructureSet(superMember) {
    throw superMember.buildCodeFrameError("Destructuring to a super field is not supported yet.");
  },
  call: function call(superMember, args) {
    return (0, _helperOptimiseCallExpression["default"])(this._get(superMember), thisExpression(), args, false);
  },
  optionalCall: function optionalCall(superMember, args) {
    return (0, _helperOptimiseCallExpression["default"])(this._get(superMember), cloneNode(thisExpression()), args, true);
  },
  "delete": function _delete(superMember) {
    if (superMember.node.computed) {
      return sequenceExpression([callExpression(this.file.addHelper("toPropertyKey"), [cloneNode(superMember.node.property)]), _core.template.expression.ast(_templateObject)]);
    } else {
      return _core.template.expression.ast(_templateObject2);
    }
  }
};
var looseHandlers = Object.assign({}, specHandlers, {
  prop: function prop(superMember) {
    var property = superMember.node.property;

    if (this.memoiser.has(property)) {
      return cloneNode(this.memoiser.get(property));
    }
    return cloneNode(property);
  },
  get: function get(superMember) {
    var isStatic = this.isStatic;
    var getSuperRef = this.getSuperRef;
    var computed = superMember.node.computed;

    var prop = this.prop(superMember);
    var object = undefined;
    if (isStatic) {
      var _getSuperRef;
      object = (_getSuperRef = getSuperRef()) != null ? _getSuperRef : memberExpression(identifier("Function"), identifier("prototype"));
    } else {
      var _getSuperRef2;
      object = memberExpression((_getSuperRef2 = getSuperRef()) != null ? _getSuperRef2 : identifier("Object"), identifier("prototype"));
    }
    return memberExpression(object, prop, computed);
  },
  set: function set(superMember, value) {
    var computed = superMember.node.computed;

    var prop = this.prop(superMember);
    return assignmentExpression("=", memberExpression(thisExpression(), prop, computed), value);
  },
  destructureSet: function destructureSet(superMember) {
    var computed = superMember.node.computed;

    var prop = this.prop(superMember);
    return memberExpression(thisExpression(), prop, computed);
  },
  call: function call(superMember, args) {
    return (0, _helperOptimiseCallExpression["default"])(this.get(superMember), thisExpression(), args, false);
  },
  optionalCall: function optionalCall(superMember, args) {
    return (0, _helperOptimiseCallExpression["default"])(this.get(superMember), thisExpression(), args, true);
  }
});

var ReplaceSupers = (function () {
  function ReplaceSupers(opts) {
    _classCallCheck(this, ReplaceSupers);

    var _opts$constantSuper;
    var path = opts.methodPath;
    this.methodPath = path;
    this.isDerivedConstructor = path.isClassMethod({
      kind: "constructor"
    }) && !!opts.superRef;
    this.isStatic = path.isObjectMethod() || path.node["static"] || (path.isStaticBlock == null ? void 0 : path.isStaticBlock());
    this.isPrivateMethod = path.isPrivate() && path.isMethod();
    this.file = opts.file;
    this.constantSuper = (_opts$constantSuper = opts.constantSuper) != null ? _opts$constantSuper : opts.isLoose;
    this.opts = opts;
  }

  _createClass(ReplaceSupers, [{
    key: "getObjectRef",
    value: function getObjectRef() {
      return cloneNode(this.opts.objectRef || this.opts.getObjectRef());
    }
  }, {
    key: "getSuperRef",
    value: function getSuperRef() {
      if (this.opts.superRef) return cloneNode(this.opts.superRef);
      if (this.opts.getSuperRef) {
        return cloneNode(this.opts.getSuperRef());
      }
    }
  }, {
    key: "replace",
    value: function replace() {
      var methodPath = this.methodPath;

      if (this.opts.refToPreserve) {
        methodPath.traverse(unshadowSuperBindingVisitor, {
          refName: this.opts.refToPreserve.name
        });
      }
      var handler = this.constantSuper ? looseHandlers : this.file.availableHelper("superPropSet") ? specHandlers : specHandlers_old;
      visitor.shouldSkip = function (path) {
        if (path.parentPath === methodPath) {
          if (path.parentKey === "decorators" || path.parentKey === "key") {
            return true;
          }
        }
      };
      (0, _helperMemberExpressionToFunctions["default"])(methodPath, visitor, Object.assign({
        file: this.file,
        scope: this.methodPath.scope,
        isDerivedConstructor: this.isDerivedConstructor,
        isStatic: this.isStatic,
        isPrivateMethod: this.isPrivateMethod,
        getObjectRef: this.getObjectRef.bind(this),
        getSuperRef: this.getSuperRef.bind(this),
        boundGet: handler.get
      }, handler));
    }
  }]);

  return ReplaceSupers;
})();

exports["default"] = ReplaceSupers;
