// @babel/core/lib/transform-file.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transformFile = transformFile;
exports.transformFileAsync = transformFileAsync;
exports.transformFileSync = transformFileSync;
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _index = require("./config/index.js");
var _index2 = require("./transformation/index.js");
var fs = require("./gensync-utils/fs.js");
({});
var transformFileRunner = _gensync()(regeneratorRuntime.mark(function callee$0$0(filename, opts) {
  var options, config, code;
  return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        options = Object.assign({}, opts, {
          filename: filename
        });
        return context$1$0.delegateYield((0, _index["default"])(options), "t0", 2);

      case 2:
        config = context$1$0.t0;

        if (!(config === null)) {
          context$1$0.next = 5;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 5:
        return context$1$0.delegateYield(fs.readFile(filename, "utf8"), "t1", 6);

      case 6:
        code = context$1$0.t1;
        return context$1$0.delegateYield((0, _index2.run)(config, code), "t2", 8);

      case 8:
        return context$1$0.abrupt("return", context$1$0.t2);

      case 9:
      case "end":
        return context$1$0.stop();
    }
  }, callee$0$0, this);
}));
function transformFile() {
  transformFileRunner.errback.apply(transformFileRunner, arguments);
}
function transformFileSync() {
  return transformFileRunner.sync.apply(transformFileRunner, arguments);
}
function transformFileAsync() {
  return transformFileRunner.async.apply(transformFileRunner, arguments);
}
0 && 0;
