// @babel/core/lib/config/files/configuration.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var marked0$0 = [readConfigCode, findRelativeConfig, loadOneConfig, loadConfig, resolveShowConfigPath].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ROOT_CONFIG_FILENAMES = void 0;
exports.findConfigUpwards = findConfigUpwards;
exports.findRelativeConfig = findRelativeConfig;
exports.findRootConfig = findRootConfig;
exports.loadConfig = loadConfig;
exports.resolveShowConfigPath = resolveShowConfigPath;
function _debug() {
  var data = require("debug");
  _debug = function () {
    return data;
  };
  return data;
}
function _fs() {
  var data = require("fs");
  _fs = function () {
    return data;
  };
  return data;
}
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _json() {
  var data = require("json5");
  _json = function () {
    return data;
  };
  return data;
}
function _gensync() {
  var data = require("gensync");
  _gensync = function () {
    return data;
  };
  return data;
}
var _caching = require("../caching.js");
var _configApi = require("../helpers/config-api.js");
var _utils = require("./utils.js");
var _moduleTypes = require("./module-types.js");
var _patternToRegex = require("../pattern-to-regex.js");
var _configError = require("../../errors/config-error.js");
var fs = require("../../gensync-utils/fs.js");
require("module");
var _rewriteStackTrace = require("../../errors/rewrite-stack-trace.js");
var _async = require("../../gensync-utils/async.js");
var debug = _debug()("babel:config:loading:files:configuration");
var ROOT_CONFIG_FILENAMES = exports.ROOT_CONFIG_FILENAMES = ["babel.config.js", "babel.config.cjs", "babel.config.mjs", "babel.config.json", "babel.config.cts"];
var RELATIVE_CONFIG_FILENAMES = [".babelrc", ".babelrc.js", ".babelrc.cjs", ".babelrc.mjs", ".babelrc.json", ".babelrc.cts"];
var BABELIGNORE_FILENAME = ".babelignore";
var runConfig = (0, _caching.makeWeakCache)(regeneratorRuntime.mark(function runConfig(options, cache) {
  return regeneratorRuntime.wrap(function runConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield([], "t0", 1);

      case 1:
        return context$1$0.abrupt("return", {
          options: (0, _rewriteStackTrace.endHiddenCallStack)(options)((0, _configApi.makeConfigAPI)(cache)),
          cacheNeedsConfiguration: !cache.configured()
        });

      case 2:
      case "end":
        return context$1$0.stop();
    }
  }, runConfig, this);
}));
function readConfigCode(filepath, data) {
  var options, cacheNeedsConfiguration, _ref;

  return regeneratorRuntime.wrap(function readConfigCode$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (_fs().existsSync(filepath)) {
          context$1$0.next = 2;
          break;
        }

        return context$1$0.abrupt("return", null);

      case 2:
        context$1$0.t0 = filepath;
        return context$1$0.delegateYield((0, _async.isAsync)(), "t1", 4);

      case 4:
        if (!context$1$0.t1) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.t2 = "auto";
        context$1$0.next = 9;
        break;

      case 8:
        context$1$0.t2 = "require";

      case 9:
        context$1$0.t3 = context$1$0.t2;
        context$1$0.t4 = "You appear to be using a native ECMAScript module configuration " + "file, which is only supported when running Babel asynchronously " + "or when using the Node.js `--experimental-require-module` flag.";
        context$1$0.t5 = "You appear to be using a configuration file that contains top-level " + "await, which is only supported when running Babel asynchronously.";
        return context$1$0.delegateYield((0, _moduleTypes["default"])(context$1$0.t0, context$1$0.t3, context$1$0.t4, context$1$0.t5), "t6", 13);

      case 13:
        options = context$1$0.t6;
        cacheNeedsConfiguration = false;

        if (!(typeof options === "function")) {
          context$1$0.next = 20;
          break;
        }

        return context$1$0.delegateYield(runConfig(options, data), "t7", 17);

      case 17:
        _ref = context$1$0.t7;
        options = _ref.options;
        cacheNeedsConfiguration = _ref.cacheNeedsConfiguration;

      case 20:
        if (!(!options || typeof options !== "object" || Array.isArray(options))) {
          context$1$0.next = 22;
          break;
        }

        throw new _configError["default"]("Configuration should be an exported JavaScript object.", filepath);

      case 22:
        if (!(typeof options.then === "function")) {
          context$1$0.next = 25;
          break;
        }

        options["catch"] == null || options["catch"](function () {});
        throw new _configError["default"]("You appear to be using an async configuration, " + "which your current version of Babel does not support. " + "We may add support for this in the future, " + "but if you're on the most recent version of @babel/core and still " + "seeing this error, then you'll need to synchronously return your config.", filepath);

      case 25:
        if (cacheNeedsConfiguration) throwConfigError(filepath);
        return context$1$0.abrupt("return", buildConfigFileObject(options, filepath));

      case 27:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
var cfboaf = new WeakMap();
function buildConfigFileObject(options, filepath) {
  var configFilesByFilepath = cfboaf.get(options);
  if (!configFilesByFilepath) {
    cfboaf.set(options, configFilesByFilepath = new Map());
  }
  var configFile = configFilesByFilepath.get(filepath);
  if (!configFile) {
    configFile = {
      filepath: filepath,
      dirname: _path().dirname(filepath),
      options: options
    };
    configFilesByFilepath.set(filepath, configFile);
  }
  return configFile;
}
var packageToBabelConfig = (0, _caching.makeWeakCacheSync)(function (file) {
  var babel = file.options.babel;
  if (babel === undefined) return null;
  if (typeof babel !== "object" || Array.isArray(babel) || babel === null) {
    throw new _configError["default"](".babel property must be an object", file.filepath);
  }
  return {
    filepath: file.filepath,
    dirname: file.dirname,
    options: babel
  };
});
var readConfigJSON5 = (0, _utils.makeStaticFileCache)(function (filepath, content) {
  var options = undefined;
  try {
    options = _json().parse(content);
  } catch (err) {
    throw new _configError["default"]("Error while parsing config - " + err.message, filepath);
  }
  if (!options) throw new _configError["default"]("No config detected", filepath);
  if (typeof options !== "object") {
    throw new _configError["default"]("Config returned typeof " + typeof options, filepath);
  }
  if (Array.isArray(options)) {
    throw new _configError["default"]("Expected config object but found array", filepath);
  }
  delete options.$schema;
  return {
    filepath: filepath,
    dirname: _path().dirname(filepath),
    options: options
  };
});
var readIgnoreConfig = (0, _utils.makeStaticFileCache)(function (filepath, content) {
  var ignoreDir = _path().dirname(filepath);
  var ignorePatterns = content.split("\n").map(function (line) {
    return line.replace(/#.*$/, "").trim();
  }).filter(Boolean);
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = ignorePatterns[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var pattern = _step.value;

      if (pattern[0] === "!") {
        throw new _configError["default"]("Negation of file paths is not supported.", filepath);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"]) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return {
    filepath: filepath,
    dirname: _path().dirname(filepath),
    ignore: ignorePatterns.map(function (pattern) {
      return (0, _patternToRegex["default"])(pattern, ignoreDir);
    })
  };
});
function findConfigUpwards(rootDir) {
  var dirname = rootDir;
  for (;;) {
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = ROOT_CONFIG_FILENAMES[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var filename = _step2.value;

        if (_fs().existsSync(_path().join(dirname, filename))) {
          return dirname;
        }
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2["return"]) {
          _iterator2["return"]();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }

    var nextDir = _path().dirname(dirname);
    if (dirname === nextDir) break;
    dirname = nextDir;
  }
  return null;
}
function findRelativeConfig(packageData, envName, caller) {
  var config, ignore, dirname, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, loc, _packageData$pkg, ignoreLoc;

  return regeneratorRuntime.wrap(function findRelativeConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        config = null;
        ignore = null;
        dirname = _path().dirname(packageData.filepath);
        _iteratorNormalCompletion3 = true;
        _didIteratorError3 = false;
        _iteratorError3 = undefined;
        context$1$0.prev = 6;
        _iterator3 = packageData.directories[Symbol.iterator]();

      case 8:
        if (_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done) {
          context$1$0.next = 21;
          break;
        }

        loc = _step3.value;

        if (config) {
          context$1$0.next = 13;
          break;
        }

        return context$1$0.delegateYield(loadOneConfig(RELATIVE_CONFIG_FILENAMES, loc, envName, caller, ((_packageData$pkg = packageData.pkg) == null ? void 0 : _packageData$pkg.dirname) === loc ? packageToBabelConfig(packageData.pkg) : null), "t0", 12);

      case 12:
        config = context$1$0.t0;

      case 13:
        if (ignore) {
          context$1$0.next = 18;
          break;
        }

        ignoreLoc = _path().join(loc, BABELIGNORE_FILENAME);
        return context$1$0.delegateYield(readIgnoreConfig(ignoreLoc), "t1", 16);

      case 16:
        ignore = context$1$0.t1;

        if (ignore) {
          debug("Found ignore %o from %o.", ignore.filepath, dirname);
        }

      case 18:
        _iteratorNormalCompletion3 = true;
        context$1$0.next = 8;
        break;

      case 21:
        context$1$0.next = 27;
        break;

      case 23:
        context$1$0.prev = 23;
        context$1$0.t2 = context$1$0["catch"](6);
        _didIteratorError3 = true;
        _iteratorError3 = context$1$0.t2;

      case 27:
        context$1$0.prev = 27;
        context$1$0.prev = 28;

        if (!_iteratorNormalCompletion3 && _iterator3["return"]) {
          _iterator3["return"]();
        }

      case 30:
        context$1$0.prev = 30;

        if (!_didIteratorError3) {
          context$1$0.next = 33;
          break;
        }

        throw _iteratorError3;

      case 33:
        return context$1$0.finish(30);

      case 34:
        return context$1$0.finish(27);

      case 35:
        return context$1$0.abrupt("return", {
          config: config,
          ignore: ignore
        });

      case 36:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this, [[6, 23, 27, 35], [28,, 30, 34]]);
}
function findRootConfig(dirname, envName, caller) {
  return loadOneConfig(ROOT_CONFIG_FILENAMES, dirname, envName, caller);
}
function loadOneConfig(names, dirname, envName, caller) {
  var previousConfig = arguments.length <= 4 || arguments[4] === undefined ? null : arguments[4];
  var configs, config;
  return regeneratorRuntime.wrap(function loadOneConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield(_gensync().all(names.map(function (filename) {
          return readConfig(_path().join(dirname, filename), envName, caller);
        })), "t0", 1);

      case 1:
        configs = context$1$0.t0;
        config = configs.reduce(function (previousConfig, config) {
          if (config && previousConfig) {
            throw new _configError["default"]("Multiple configuration files found. Please remove one:\n" + (" - " + _path().basename(previousConfig.filepath) + "\n") + (" - " + config.filepath + "\n") + ("from " + dirname));
          }
          return config || previousConfig;
        }, previousConfig);

        if (config) {
          debug("Found configuration %o from %o.", config.filepath, dirname);
        }
        return context$1$0.abrupt("return", config);

      case 5:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
function loadConfig(name, dirname, envName, caller) {
  var filepath, conf;
  return regeneratorRuntime.wrap(function loadConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        filepath = ((function (v, w) {
          return v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1];
        })(process.versions.node, "8.9") ? require.resolve : function (r, _ref2) {
          var _ref2$paths = _slicedToArray(_ref2.paths, 1);

          var b = _ref2$paths[0];
          var M = arguments.length <= 2 || arguments[2] === undefined ? require("module") : arguments[2];

          var f = M._findPath(r, M._nodeModulePaths(b).concat(b));
          if (f) return f;
          f = new Error("Cannot resolve module '" + r + "'");
          f.code = "MODULE_NOT_FOUND";
          throw f;
        })(name, {
          paths: [dirname]
        });
        return context$1$0.delegateYield(readConfig(filepath, envName, caller), "t0", 2);

      case 2:
        conf = context$1$0.t0;

        if (conf) {
          context$1$0.next = 5;
          break;
        }

        throw new _configError["default"]("Config file contains no configuration data", filepath);

      case 5:
        debug("Loaded config %o from %o.", name, dirname);
        return context$1$0.abrupt("return", conf);

      case 7:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[3], this);
}
function readConfig(filepath, envName, caller) {
  var ext = _path().extname(filepath);
  switch (ext) {
    case ".js":
    case ".cjs":
    case ".mjs":
    case ".ts":
    case ".cts":
    case ".mts":
      return readConfigCode(filepath, {
        envName: envName,
        caller: caller
      });
    default:
      return readConfigJSON5(filepath);
  }
}
function resolveShowConfigPath(dirname) {
  var targetPath, absolutePath, stats;
  return regeneratorRuntime.wrap(function resolveShowConfigPath$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        targetPath = process.env.BABEL_SHOW_CONFIG_FOR;

        if (!(targetPath != null)) {
          context$1$0.next = 8;
          break;
        }

        absolutePath = _path().resolve(dirname, targetPath);
        return context$1$0.delegateYield(fs.stat(absolutePath), "t0", 4);

      case 4:
        stats = context$1$0.t0;

        if (stats.isFile()) {
          context$1$0.next = 7;
          break;
        }

        throw new Error(absolutePath + ": BABEL_SHOW_CONFIG_FOR must refer to a regular file, directories are not supported.");

      case 7:
        return context$1$0.abrupt("return", absolutePath);

      case 8:
        return context$1$0.abrupt("return", null);

      case 9:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[4], this);
}
function throwConfigError(filepath) {
  throw new _configError["default"]("Caching was left unconfigured. Babel's plugins, presets, and .babelrc.js files can be configured\nfor various types of caching, using the first param of their handler functions:\n\nmodule.exports = function(api) {\n  // The API exposes the following:\n\n  // Cache the returned value forever and don't call this function again.\n  api.cache(true);\n\n  // Don't cache at all. Not recommended because it will be very slow.\n  api.cache(false);\n\n  // Cached based on the value of some function. If this function returns a value different from\n  // a previously-encountered value, the plugins will re-evaluate.\n  var env = api.cache(() => process.env.NODE_ENV);\n\n  // If testing for a specific env, we recommend specifics to avoid instantiating a plugin for\n  // any possible NODE_ENV value that might come up during plugin execution.\n  var isProd = api.cache(() => process.env.NODE_ENV === \"production\");\n\n  // .cache(fn) will perform a linear search though instances to find the matching plugin based\n  // based on previous instantiated plugins. If you want to recreate the plugin and discard the\n  // previous instance whenever something changes, you may use:\n  var isProd = api.cache.invalidate(() => process.env.NODE_ENV === \"production\");\n\n  // Note, we also expose the following more-verbose versions of the above examples:\n  api.cache.forever(); // api.cache(true)\n  api.cache.never();   // api.cache(false)\n  api.cache.using(fn); // api.cache(fn)\n\n  // Return the value that will be cached.\n  return { };\n};", filepath);
}
0 && 0;
