// @babel/core/lib/config/files/index-browser.js
"use strict";

var marked0$0 = [findPackageData, findRelativeConfig, findRootConfig, loadConfig, resolveShowConfigPath].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ROOT_CONFIG_FILENAMES = void 0;
exports.findConfigUpwards = findConfigUpwards;
exports.findPackageData = findPackageData;
exports.findRelativeConfig = findRelativeConfig;
exports.findRootConfig = findRootConfig;
exports.loadConfig = loadConfig;
exports.loadPlugin = loadPlugin;
exports.loadPreset = loadPreset;
exports.resolvePlugin = resolvePlugin;
exports.resolvePreset = resolvePreset;
exports.resolveShowConfigPath = resolveShowConfigPath;
function findConfigUpwards(rootDir) {
  return null;
}
function findPackageData(filepath) {
  return regeneratorRuntime.wrap(function findPackageData$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", {
          filepath: filepath,
          directories: [],
          pkg: null,
          isPackage: false
        });

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function findRelativeConfig(pkgData, envName, caller) {
  return regeneratorRuntime.wrap(function findRelativeConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", {
          config: null,
          ignore: null
        });

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}
function findRootConfig(dirname, envName, caller) {
  return regeneratorRuntime.wrap(function findRootConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", null);

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
function loadConfig(name, dirname, envName, caller) {
  return regeneratorRuntime.wrap(function loadConfig$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        throw new Error("Cannot load " + name + " relative to " + dirname + " in a browser");

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[3], this);
}
function resolveShowConfigPath(dirname) {
  return regeneratorRuntime.wrap(function resolveShowConfigPath$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt("return", null);

      case 1:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[4], this);
}
var ROOT_CONFIG_FILENAMES = exports.ROOT_CONFIG_FILENAMES = [];
function resolvePlugin(name, dirname) {
  return null;
}
function resolvePreset(name, dirname) {
  return null;
}
function loadPlugin(name, dirname) {
  throw new Error("Cannot load plugin " + name + " relative to " + dirname + " in a browser");
}
function loadPreset(name, dirname) {
  throw new Error("Cannot load preset " + name + " relative to " + dirname + " in a browser");
}
0 && 0;
