// @babel/core/lib/config/files/module-types.js
"use strict";

var marked0$0 = [loadCodeDefault].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = loadCodeDefault;
exports.supportsESM = void 0;
var _async = require("../../gensync-utils/async.js");
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _url() {
  var data = require("url");
  _url = function () {
    return data;
  };
  return data;
}
require("module");
function _semver() {
  var data = require("semver");
  _semver = function () {
    return data;
  };
  return data;
}
function _debug() {
  var data = require("debug");
  _debug = function () {
    return data;
  };
  return data;
}
var _rewriteStackTrace = require("../../errors/rewrite-stack-trace.js");
var _configError = require("../../errors/config-error.js");
var _transformFile = require("../../transform-file.js");
function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
        u = i.value;
  } catch (n) {
    return void e(n);
  }i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
        e = arguments;return new Promise(function (r, o) {
      var a = n.apply(t, e);function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }_next(void 0);
    });
  };
}
var debug = _debug()("babel:config:loading:files:module-types");
{
  try {
    var import_ = require("./import.cjs");
  } catch (_unused) {}
}
var supportsESM = exports.supportsESM = _semver().satisfies(process.versions.node, "^12.17 || >=13.2");
var LOADING_CJS_FILES = new Set();
function loadCjsDefault(filepath) {
  if (LOADING_CJS_FILES.has(filepath)) {
    debug("Auto-ignoring usage of config %o.", filepath);
    return {};
  }
  var module = undefined;
  try {
    LOADING_CJS_FILES.add(filepath);
    module = (0, _rewriteStackTrace.endHiddenCallStack)(require)(filepath);
  } finally {
    LOADING_CJS_FILES["delete"](filepath);
  }
  {
    return module != null && (module.__esModule || module[Symbol.toStringTag] === "Module") ? module["default"] || (arguments[1] ? module : undefined) : module;
  }
}
var loadMjsFromPath = (0, _rewriteStackTrace.endHiddenCallStack)((function () {
  var _loadMjsFromPath = _asyncToGenerator(regeneratorRuntime.mark(function callee$1$0(filepath) {
    var url;
    return regeneratorRuntime.wrap(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          url = (0, _url().pathToFileURL)(filepath).toString() + "?import";

          if (import_) {
            context$2$0.next = 3;
            break;
          }

          throw new _configError["default"]("Internal error: Native ECMAScript modules aren't supported by this platform.\n", filepath);

        case 3:
          context$2$0.next = 5;
          return import_(url);

        case 5:
          return context$2$0.abrupt("return", context$2$0.sent);

        case 6:
        case "end":
          return context$2$0.stop();
      }
    }, callee$1$0, this);
  }));
  function loadMjsFromPath(_x) {
    return _loadMjsFromPath.apply(this, arguments);
  }
  return loadMjsFromPath;
})());
var SUPPORTED_EXTENSIONS = {
  ".js": "unknown",
  ".mjs": "esm",
  ".cjs": "cjs",
  ".ts": "unknown",
  ".mts": "esm",
  ".cts": "cjs"
};
var asyncModules = new Set();
function loadCodeDefault(filepath, loader, esmError, tlaError) {
  var async,
      ext,
      isTS,
      type,
      pattern,
      promise,
      args$1$0 = arguments;
  return regeneratorRuntime.wrap(function loadCodeDefault$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        async = undefined;
        ext = _path().extname(filepath);
        isTS = ext === ".ts" || ext === ".cts" || ext === ".mts";
        type = SUPPORTED_EXTENSIONS[hasOwnProperty.call(SUPPORTED_EXTENSIONS, ext) ? ext : ".js"];
        pattern = loader + " " + type;
        context$1$0.t0 = pattern;
        context$1$0.next = context$1$0.t0 === "require cjs" ? 8 : context$1$0.t0 === "auto cjs" ? 8 : context$1$0.t0 === "auto unknown" ? 13 : context$1$0.t0 === "require unknown" ? 13 : context$1$0.t0 === "require esm" ? 13 : context$1$0.t0 === "auto esm" ? 39 : 50;
        break;

      case 8:
        if (!isTS) {
          context$1$0.next = 12;
          break;
        }

        return context$1$0.abrupt("return", ensureTsSupport(filepath, ext, function () {
          return loadCjsDefault(filepath);
        }));

      case 12:
        return context$1$0.abrupt("return", loadCjsDefault(filepath, args$1$0[2]));

      case 13:
        context$1$0.prev = 13;

        if (!isTS) {
          context$1$0.next = 18;
          break;
        }

        return context$1$0.abrupt("return", ensureTsSupport(filepath, ext, function () {
          return loadCjsDefault(filepath);
        }));

      case 18:
        return context$1$0.abrupt("return", loadCjsDefault(filepath, args$1$0[2]));

      case 19:
        context$1$0.next = 39;
        break;

      case 21:
        context$1$0.prev = 21;
        context$1$0.t1 = context$1$0["catch"](13);

        if (!(context$1$0.t1.code === "ERR_REQUIRE_ASYNC_MODULE" || context$1$0.t1.code === "ERR_REQUIRE_CYCLE_MODULE" && asyncModules.has(filepath))) {
          context$1$0.next = 35;
          break;
        }

        asyncModules.add(filepath);

        if (!(async != null)) {
          context$1$0.next = 29;
          break;
        }

        context$1$0.t2 = async;
        context$1$0.next = 31;
        break;

      case 29:
        return context$1$0.delegateYield((0, _async.isAsync)(), "t3", 30);

      case 30:
        context$1$0.t2 = async = context$1$0.t3;

      case 31:
        if (context$1$0.t2) {
          context$1$0.next = 33;
          break;
        }

        throw new _configError["default"](tlaError, filepath);

      case 33:
        context$1$0.next = 39;
        break;

      case 35:
        if (!(context$1$0.t1.code === "ERR_REQUIRE_ESM" || type === "esm")) {
          context$1$0.next = 38;
          break;
        }

        context$1$0.next = 39;
        break;

      case 38:
        throw context$1$0.t1;

      case 39:
        if (!(async != null)) {
          context$1$0.next = 43;
          break;
        }

        context$1$0.t4 = async;
        context$1$0.next = 45;
        break;

      case 43:
        return context$1$0.delegateYield((0, _async.isAsync)(), "t5", 44);

      case 44:
        context$1$0.t4 = async = context$1$0.t5;

      case 45:
        if (!context$1$0.t4) {
          context$1$0.next = 49;
          break;
        }

        promise = isTS ? ensureTsSupport(filepath, ext, function () {
          return loadMjsFromPath(filepath);
        }) : loadMjsFromPath(filepath);
        return context$1$0.delegateYield((0, _async.waitFor)(promise), "t6", 48);

      case 48:
        return context$1$0.abrupt("return", context$1$0.t6["default"]);

      case 49:
        throw new _configError["default"](esmError, filepath);

      case 50:
        throw new Error("Internal Babel error: unreachable code.");

      case 51:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this, [[13, 21]]);
}
function ensureTsSupport(filepath, ext, callback) {
  if (require.extensions[".ts"] || require.extensions[".cts"] || require.extensions[".mts"]) {
    return callback();
  }
  if (ext !== ".cts") {
    throw new _configError["default"]("You are using a " + ext + " config file, but Babel only supports transpiling .cts configs. Either:\n- Use a .cts config file\n- Update to Node.js 23.6.0, which has native TypeScript support\n- Install ts-node to transpile " + ext + " files on the fly", filepath);
  }
  var opts = {
    babelrc: false,
    configFile: false,
    sourceType: "unambiguous",
    sourceMaps: "inline",
    sourceFileName: _path().basename(filepath),
    presets: [[getTSPreset(filepath), Object.assign({
      onlyRemoveTypeImports: true,
      optimizeConstEnums: true
    }, {
      allowDeclareFields: true
    })]]
  };
  var _handler = function handler(m, filename) {
    if (_handler && filename.endsWith(".cts")) {
      try {
        return m._compile((0, _transformFile.transformFileSync)(filename, Object.assign({}, opts, {
          filename: filename
        })).code, filename);
      } catch (error) {
        var packageJson = require("@babel/preset-typescript/package.json");
        if (_semver().lt(packageJson.version, "7.21.4")) {
          console.error("`.cts` configuration file failed to load, please try to update `@babel/preset-typescript`.");
        }
        throw error;
      }
    }
    return require.extensions[".js"](m, filename);
  };
  require.extensions[ext] = _handler;
  try {
    return callback();
  } finally {
    if (require.extensions[ext] === _handler) delete require.extensions[ext];
    _handler = undefined;
  }
}
function getTSPreset(filepath) {
  try {
    return require("@babel/preset-typescript");
  } catch (error) {
    if (error.code !== "MODULE_NOT_FOUND") throw error;
    var message = "You appear to be using a .cts file as Babel configuration, but the `@babel/preset-typescript` package was not found: please install it!";
    {
      if (process.versions.pnp) {
        message += "\nIf you are using Yarn Plug'n'Play, you may also need to add the following configuration to your .yarnrc.yml file:\n\npackageExtensions:\n\t\"@babel/core@*\":\n\t\tpeerDependencies:\n\t\t\t\"@babel/preset-typescript\": \"*\"\n";
      }
    }
    throw new _configError["default"](message, filepath);
  }
}
0 && 0;
