// @babel/core/lib/config/files/package.js
"use strict";

var marked0$0 = [findPackageData].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findPackageData = findPackageData;
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
var _utils = require("./utils.js");
var _configError = require("../../errors/config-error.js");
var PACKAGE_FILENAME = "package.json";
var readConfigPackage = (0, _utils.makeStaticFileCache)(function (filepath, content) {
  var options = undefined;
  try {
    options = JSON.parse(content);
  } catch (err) {
    throw new _configError["default"]("Error while parsing JSON - " + err.message, filepath);
  }
  if (!options) throw new Error(filepath + ": No config detected");
  if (typeof options !== "object") {
    throw new _configError["default"]("Config returned typeof " + typeof options, filepath);
  }
  if (Array.isArray(options)) {
    throw new _configError["default"]("Expected config object but found array", filepath);
  }
  return {
    filepath: filepath,
    dirname: _path().dirname(filepath),
    options: options
  };
});
function findPackageData(filepath) {
  var pkg, directories, isPackage, dirname, nextLoc;
  return regeneratorRuntime.wrap(function findPackageData$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        pkg = null;
        directories = [];
        isPackage = true;
        dirname = _path().dirname(filepath);

      case 4:
        if (!(!pkg && _path().basename(dirname) !== "node_modules")) {
          context$1$0.next = 15;
          break;
        }

        directories.push(dirname);
        return context$1$0.delegateYield(readConfigPackage(_path().join(dirname, PACKAGE_FILENAME)), "t0", 7);

      case 7:
        pkg = context$1$0.t0;
        nextLoc = _path().dirname(dirname);

        if (!(dirname === nextLoc)) {
          context$1$0.next = 12;
          break;
        }

        isPackage = false;
        return context$1$0.abrupt("break", 15);

      case 12:
        dirname = nextLoc;
        context$1$0.next = 4;
        break;

      case 15:
        return context$1$0.abrupt("return", {
          filepath: filepath,
          directories: directories,
          pkg: pkg,
          isPackage: isPackage
        });

      case 16:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
0 && 0;
