// @babel/core/lib/config/files/plugins.js
"use strict";

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

var marked0$0 = [loadPlugin, loadPreset, resolveAlternativesHelper, requireModule].map(regeneratorRuntime.mark);
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadPlugin = loadPlugin;
exports.loadPreset = loadPreset;
exports.resolvePreset = exports.resolvePlugin = void 0;
function _debug() {
  var data = require("debug");
  _debug = function () {
    return data;
  };
  return data;
}
function _path() {
  var data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
var _async = require("../../gensync-utils/async.js");
var _moduleTypes = require("./module-types.js");
function _url() {
  var data = require("url");
  _url = function () {
    return data;
  };
  return data;
}
var _importMetaResolve = require("../../vendor/import-meta-resolve.js");
require("module");
function _fs() {
  var data = require("fs");
  _fs = function () {
    return data;
  };
  return data;
}
var debug = _debug()("babel:config:loading:files:plugins");
var EXACT_RE = /^module:/;
var BABEL_PLUGIN_PREFIX_RE = /^(?!@|module:|[^/]+\/|babel-plugin-)/;
var BABEL_PRESET_PREFIX_RE = /^(?!@|module:|[^/]+\/|babel-preset-)/;
var BABEL_PLUGIN_ORG_RE = /^(@babel\/)(?!plugin-|[^/]+\/)/;
var BABEL_PRESET_ORG_RE = /^(@babel\/)(?!preset-|[^/]+\/)/;
var OTHER_PLUGIN_ORG_RE = /^(@(?!babel\/)[^/]+\/)(?![^/]*babel-plugin(?:-|\/|$)|[^/]+\/)/;
var OTHER_PRESET_ORG_RE = /^(@(?!babel\/)[^/]+\/)(?![^/]*babel-preset(?:-|\/|$)|[^/]+\/)/;
var OTHER_ORG_DEFAULT_RE = /^(@(?!babel$)[^/]+)$/;
var resolvePlugin = exports.resolvePlugin = resolveStandardizedName.bind(null, "plugin");
var resolvePreset = exports.resolvePreset = resolveStandardizedName.bind(null, "preset");
function loadPlugin(name, dirname) {
  var _resolvePlugin, filepath, loader, value;

  return regeneratorRuntime.wrap(function loadPlugin$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.t0 = name;
        context$1$0.t1 = dirname;
        return context$1$0.delegateYield((0, _async.isAsync)(), "t2", 3);

      case 3:
        context$1$0.t3 = context$1$0.t2;
        _resolvePlugin = resolvePlugin(context$1$0.t0, context$1$0.t1, context$1$0.t3);
        filepath = _resolvePlugin.filepath;
        loader = _resolvePlugin.loader;
        return context$1$0.delegateYield(requireModule("plugin", loader, filepath), "t4", 8);

      case 8:
        value = context$1$0.t4;

        debug("Loaded plugin %o from %o.", name, dirname);
        return context$1$0.abrupt("return", {
          filepath: filepath,
          value: value
        });

      case 11:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[0], this);
}
function loadPreset(name, dirname) {
  var _resolvePreset, filepath, loader, value;

  return regeneratorRuntime.wrap(function loadPreset$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.t0 = name;
        context$1$0.t1 = dirname;
        return context$1$0.delegateYield((0, _async.isAsync)(), "t2", 3);

      case 3:
        context$1$0.t3 = context$1$0.t2;
        _resolvePreset = resolvePreset(context$1$0.t0, context$1$0.t1, context$1$0.t3);
        filepath = _resolvePreset.filepath;
        loader = _resolvePreset.loader;
        return context$1$0.delegateYield(requireModule("preset", loader, filepath), "t4", 8);

      case 8:
        value = context$1$0.t4;

        debug("Loaded preset %o from %o.", name, dirname);
        return context$1$0.abrupt("return", {
          filepath: filepath,
          value: value
        });

      case 11:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[1], this);
}
function standardizeName(type, name) {
  if (_path().isAbsolute(name)) return name;
  var isPreset = type === "preset";
  return name.replace(isPreset ? BABEL_PRESET_PREFIX_RE : BABEL_PLUGIN_PREFIX_RE, "babel-" + type + "-").replace(isPreset ? BABEL_PRESET_ORG_RE : BABEL_PLUGIN_ORG_RE, "$1" + type + "-").replace(isPreset ? OTHER_PRESET_ORG_RE : OTHER_PLUGIN_ORG_RE, "$1babel-" + type + "-").replace(OTHER_ORG_DEFAULT_RE, "$1/babel-" + type).replace(EXACT_RE, "");
}
function resolveAlternativesHelper(type, name) {
  var standardizedName, _ref, error, value, oppositeType, transformName;

  return regeneratorRuntime.wrap(function resolveAlternativesHelper$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        standardizedName = standardizeName(type, name);
        context$1$0.next = 3;
        return standardizedName;

      case 3:
        _ref = context$1$0.sent;
        error = _ref.error;
        value = _ref.value;

        if (error) {
          context$1$0.next = 8;
          break;
        }

        return context$1$0.abrupt("return", value);

      case 8:
        if (!(error.code !== "MODULE_NOT_FOUND")) {
          context$1$0.next = 10;
          break;
        }

        throw error;

      case 10:
        context$1$0.t0 = standardizedName !== name;

        if (!context$1$0.t0) {
          context$1$0.next = 15;
          break;
        }

        context$1$0.next = 14;
        return name;

      case 14:
        context$1$0.t0 = !context$1$0.sent.error;

      case 15:
        if (!context$1$0.t0) {
          context$1$0.next = 17;
          break;
        }

        error.message += "\n- If you want to resolve \"" + name + "\", use \"module:" + name + "\"";

      case 17:
        context$1$0.next = 19;
        return standardizeName(type, "@babel/" + name);

      case 19:
        if (context$1$0.sent.error) {
          context$1$0.next = 21;
          break;
        }

        error.message += "\n- Did you mean \"@babel/" + name + "\"?";

      case 21:
        oppositeType = type === "preset" ? "plugin" : "preset";
        context$1$0.next = 24;
        return standardizeName(oppositeType, name);

      case 24:
        if (context$1$0.sent.error) {
          context$1$0.next = 26;
          break;
        }

        error.message += "\n- Did you accidentally pass a " + oppositeType + " as a " + type + "?";

      case 26:
        if (!(type === "plugin")) {
          context$1$0.next = 35;
          break;
        }

        transformName = standardizedName.replace("-proposal-", "-transform-");
        context$1$0.t1 = transformName !== standardizedName;

        if (!context$1$0.t1) {
          context$1$0.next = 33;
          break;
        }

        context$1$0.next = 32;
        return transformName;

      case 32:
        context$1$0.t1 = !context$1$0.sent.error;

      case 33:
        if (!context$1$0.t1) {
          context$1$0.next = 35;
          break;
        }

        error.message += "\n- Did you mean \"" + transformName + "\"?";

      case 35:
        error.message += "\n\nMake sure that all the Babel plugins and presets you are using\nare defined as dependencies or devDependencies in your package.json\nfile. It's possible that the missing plugin is loaded by a preset\nyou are using that forgot to add the plugin to its dependencies: you\ncan workaround this problem by explicitly adding the missing package\nto your top-level package.json.\n";
        throw error;

      case 37:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[2], this);
}
function tryRequireResolve(id, dirname) {
  try {
    if (dirname) {
      return {
        error: null,
        value: ((function (v, w) {
          return v = v.split("."), w = w.split("."), +v[0] > +w[0] || v[0] == w[0] && +v[1] >= +w[1];
        })(process.versions.node, "8.9") ? require.resolve : function (r, _ref2) {
          var _ref2$paths = _slicedToArray(_ref2.paths, 1);

          var b = _ref2$paths[0];
          var M = arguments.length <= 2 || arguments[2] === undefined ? require("module") : arguments[2];

          var f = M._findPath(r, M._nodeModulePaths(b).concat(b));
          if (f) return f;
          f = new Error("Cannot resolve module '" + r + "'");
          f.code = "MODULE_NOT_FOUND";
          throw f;
        })(id, {
          paths: [dirname]
        })
      };
    } else {
      return {
        error: null,
        value: require.resolve(id)
      };
    }
  } catch (error) {
    return {
      error: error,
      value: null
    };
  }
}
function tryImportMetaResolve(id, options) {
  try {
    return {
      error: null,
      value: (0, _importMetaResolve.resolve)(id, options)
    };
  } catch (error) {
    return {
      error: error,
      value: null
    };
  }
}
function resolveStandardizedNameForRequire(type, name, dirname) {
  var it = resolveAlternativesHelper(type, name);
  var res = it.next();
  while (!res.done) {
    res = it.next(tryRequireResolve(res.value, dirname));
  }
  return {
    loader: "require",
    filepath: res.value
  };
}
function resolveStandardizedNameForImport(type, name, dirname) {
  var parentUrl = (0, _url().pathToFileURL)(_path().join(dirname, "./babel-virtual-resolve-base.js")).href;
  var it = resolveAlternativesHelper(type, name);
  var res = it.next();
  while (!res.done) {
    res = it.next(tryImportMetaResolve(res.value, parentUrl));
  }
  return {
    loader: "auto",
    filepath: (0, _url().fileURLToPath)(res.value)
  };
}
function resolveStandardizedName(type, name, dirname, allowAsync) {
  if (!_moduleTypes.supportsESM || !allowAsync) {
    return resolveStandardizedNameForRequire(type, name, dirname);
  }
  try {
    var resolved = resolveStandardizedNameForImport(type, name, dirname);
    if (!(0, _fs().existsSync)(resolved.filepath)) {
      throw Object.assign(new Error("Could not resolve \"" + name + "\" in file " + dirname + "."), {
        type: "MODULE_NOT_FOUND"
      });
    }
    return resolved;
  } catch (e) {
    try {
      return resolveStandardizedNameForRequire(type, name, dirname);
    } catch (e2) {
      if (e.type === "MODULE_NOT_FOUND") throw e;
      if (e2.type === "MODULE_NOT_FOUND") throw e2;
      throw e;
    }
  }
}
{
  var LOADING_MODULES = new Set();
}
function requireModule(type, loader, name) {
  return regeneratorRuntime.wrap(function requireModule$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.delegateYield((0, _async.isAsync)(), "t1", 1);

      case 1:
        context$1$0.t0 = !context$1$0.t1;

        if (!context$1$0.t0) {
          context$1$0.next = 4;
          break;
        }

        context$1$0.t0 = LOADING_MODULES.has(name);

      case 4:
        if (!context$1$0.t0) {
          context$1$0.next = 6;
          break;
        }

        throw new Error("Reentrant " + type + " detected trying to load \"" + name + "\". This module is not ignored " + "and is trying to load itself while compiling itself, leading to a dependency cycle. " + 'We recommend adding it to your "ignore" list in your babelrc, or to a .babelignore.');

      case 6:
        context$1$0.prev = 6;

        LOADING_MODULES.add(name);
        return context$1$0.delegateYield((0, _moduleTypes["default"])(name, loader, "You appear to be using a native ECMAScript module " + type + ", " + "which is only supported when running Babel asynchronously " + "or when using the Node.js `--experimental-require-module` flag.", "You appear to be using a " + type + " that contains top-level await, " + "which is only supported when running Babel asynchronously.", true), "t2", 9);

      case 9:
        return context$1$0.abrupt("return", context$1$0.t2);

      case 12:
        context$1$0.prev = 12;
        context$1$0.t3 = context$1$0["catch"](6);

        context$1$0.t3.message = "[BABEL]: " + context$1$0.t3.message + " (While processing: " + name + ")";
        throw context$1$0.t3;

      case 16:
        context$1$0.prev = 16;

        LOADING_MODULES["delete"](name);
        return context$1$0.finish(16);

      case 19:
      case "end":
        return context$1$0.stop();
    }
  }, marked0$0[3], this, [[6, 12, 16, 19]]);
}
0 && 0;
